---
title: "Data cleaning"
author: "Daoming Wang"
date: "2021/7/18"
output:
  html_document: 
    theme: flatly
    highlight: espresso
    toc: true
    toc_depth: 4
    toc_float: true
  word_document: default
  pdf_document:
    includes:
      in_header: header.tex
      keep_tex: yes
      latex_engine: xelatex
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## 1 Preparation
### 1.1 Import

```{r 1.1, message=FALSE, warning=FALSE, paged.print=FALSE}
source("functions.R")
```

### 1.2 Inputs

```{r 1.2, message=FALSE, warning=FALSE}
# Read 300OB files
ob_ba<-read.table("00.rawData/300OB_phenotype/ob_39ba_entities_301samples_plasma.txt")
ob_phen<-read.table("00.rawData/300OB_phenotype/ob_130pheno_302samples.txt")
ob_id <- read.table('00.rawData/300OB_phenotype/key_300OB.txt',
                    header = T, sep = '\t', check.names = F, stringsAsFactors = F)

# Read LLD files
lld_ba<-read.table("00.rawData/LLD_phenotype/20200525_43BA_1135patients.txt")
lld_phen<-read.table("00.rawData/LLD_phenotype/20200226_207allPhenotypes_1135patients.txt")
lld_id<-read.table("00.rawData/LLD_phenotype/LLD_GoNL_1659samples.txt", header = T)

# Read SV files
dsgv <- read.delim("00.rawData/SV/lld_300OB.dsgv.csv",
                 header = T,sep = ",",check.names = F,stringsAsFactors = F,row.names = "")
vsgv <- read.delim("00.rawData/SV/lld_300OB.vsgv.csv",
                 header = T,sep = ",",check.names = F,stringsAsFactors = F,row.names = "")
dsgv_anno<-read.delim("00.rawData/SV/s02.dSVs_anno.tsv",
                      sep = "\t",header = T,quote = '',stringsAsFactors = F)
vsgv_anno<-read.delim("00.rawData/SV/s03.vSVs_anno.tsv",
                      sep = "\t",header = T,quote = '',stringsAsFactors = F)

# Read database files
taxa_length <- read.table("00.rawdata/database/Species_genome_size.tsv",
                        sep = "\t", header = T,check.names = F,stringsAsFactors = F)
taxonomy    <- read.csv("00.rawdata/database/representatives.genomes.taxonomy.csv",
                   sep = ",", header = T,check.names = F,stringsAsFactors = F)
taxonomy[taxonomy == ""]<-"Unknown"
colnames(taxonomy)[1]<-'X'
ncbi<-read.csv("00.rawData/database/NCBI_accession.txt", sep = "\t",header = T)
tax_relationship<-read.csv("00.rawData/database/progenome1_species_relationship.tsv",sep = "\t",header = F)

# Read species abundance files
lld_SV_abun_s<-read.csv("00.rawData/mbio/LLD.s1135.SV_spe_abun.S.tsv",header = T,sep = "\t",check.names = F)
ob_SV_abun_s<-read.csv("00.rawData/mbio/300OB.s298.SV_spe_abun.S.tsv",header = T,sep = "\t",check.names = F)

```


## 2 Clean phenotypical data

Clean phenotypical data, extract covariates. '0' represents female, '1' represents male.

```{r 2}
## lld
lld_basic <- lld_phen[, c(1, 2, 3, 4)]
colnames(lld_basic) <- c('Gender', 'Age', 'Height', 'BMI') 
lld_basic$Obesity <- rep('Obesity', nrow(lld_basic))
lld_basic$Obesity[lld_basic$BMI < 27] <- 'Non-obesity'
lld_basic$Reads_number <-
  lld_id$LLD_reads[match(rownames(lld_basic), lld_id$LLD_GoNL_all_id)]

## 300OB
ob_basic <- ob_phen[, c(2, 3, 4, 6)]
colnames(ob_basic) <- c('Gender', 'Age', 'Height', 'BMI')
ob_basic$Gender[ob_basic$Gender == 1] <- 0 # Female
ob_basic$Gender[ob_basic$Gender == 2] <- 1 # Male
ob_basic$Obesity <- rep('Obesity', nrow(ob_basic))
ob_basic$Reads_number <-
  ob_phen$reads[match(rownames(ob_basic), rownames(ob_phen))]

## all
all_basic<-rbind(lld_basic, ob_basic)
all_basic<-data.frame(Cohort = c(rep('LLD', nrow(lld_basic)), rep('300OB', nrow(ob_basic))) ,all_basic)
all_basic$Group<-rep('LLD non-obesity',nrow(all_basic))
all_basic$Group[all_basic$Cohort=='LLD' & all_basic$Obesity=='Obesity']<-'LLD obesity'
all_basic$Group[all_basic$Cohort=='300OB']<-'300OB'

all_basic_covar <- all_basic
all_basic_covar$Cohort<-as.factor(all_basic_covar$Cohort) %>% as.numeric
all_basic_covar$Cohort <- all_basic_covar$Cohort-1

## outputs
if(!dir.exists("01.cleanData/")){dir.create("01.cleanData/")}

if(!dir.exists("01.cleanData/phen_lld")){dir.create("01.cleanData/phen_lld")}
write.table(lld_basic, "01.cleanData/phen_lld/20200801_LLD_basic_1135samples.tsv",sep = '\t')

if(!dir.exists("01.cleanData/phen_300OB")){dir.create("01.cleanData/phen_300OB")}
write.table(ob_basic,  "01.cleanData/phen_300OB/20200801_300OB_basic_302samples.tsv",sep = '\t')

if(!dir.exists("01.cleanData/phen_all")){dir.create("01.cleanData/phen_all")}
write.table(all_basic, "01.cleanData/phen_all/20200801_LLD_basic_1437samples.tsv",sep = '\t')
write.table(all_basic_covar, "01.cleanData/phen_all/20200801_LLD_covar_1437samples.tsv",sep = '\t')
```

## 3 Clean BA data

```{r 3}
# align BA profiles of 300OB and LLD
ob_na_col<-colnames(lld_ba)[is.na(match(colnames(lld_ba), colnames(ob_ba)))]
ob_na_df<-matrix(NA, ncol = length(ob_na_col), nrow = nrow(ob_ba)) %>% as.data.frame
colnames(ob_na_df)<-ob_na_col
ob_ba<-cbind(ob_ba, ob_na_df)
ob_ba <- ob_ba[match(rownames(ob_phen), rownames(ob_ba)),match(colnames(lld_ba), colnames(ob_ba))]
rownames(ob_ba)<-rownames(ob_phen)

all_ba <- rbind(lld_ba, ob_ba)[,c('C4',concentration.group)]

### calculation formula 
all_ba$Total_primary_BAs   <- rowSums(all_ba[,primary.group])
all_ba$Total_secondary_BAs <- rowSums(all_ba[,second.group])
all_ba$Total_BAs <- rowSums(all_ba[,concentration.group])

all_ba_p <- all_ba[,concentration.group]/all_ba$Total_BAs
colnames(all_ba_p) <- paste(colnames(all_ba_p),"p",sep = "_")
all_ba<-cbind(all_ba,all_ba_p)

all_ba$Secondary_primary_ratio       <- all_ba$Total_secondary_BAs/all_ba$Total_primary_BAs
all_ba$CA_CDCA_ratio                 <- rowSums(all_ba[,ca.group])/rowSums(all_ba[,cdca.group])
all_ba$Unconjugated_conjugated_ratio <- rowSums(all_ba[,unconjugated])/rowSums(all_ba[,conjugated])
all_ba$CA_dehydro_deconju_ratio      <- rowSums(all_ba[,ca.dehydroxylation])/rowSums(all_ba[,ca.deconjugated])
all_ba$Taurine_glycine_ratio         <- rowSums(all_ba[,taurine.group])/rowSums(all_ba[,glycine.group])

all_ba <- as.matrix(all_ba)
all_ba[is.infinite(all_ba)] <- NA

## Split dataset
lld_ba <- all_ba[c(1:nrow(lld_ba)),]
ob_ba  <- all_ba[(nrow(lld_ba)+1):nrow(all_ba),]

## outputs
if(!dir.exists("01.cleanData/phen_lld")){dir.create("01.cleanData/phen_lld")}
write.table(lld_ba,"01.cleanData/phen_lld/20200801_LLD_39BA_1135samples.tsv",sep = '\t')

if(!dir.exists("01.cleanData/phen_300OB")){dir.create("01.cleanData/phen_300OB/")}
write.table(ob_ba,"01.cleanData/phen_300OB/20200801_300OB_39BA_302samples.tsv",sep = '\t')

if(!dir.exists("01.cleanData/phen_all")){dir.create("01.cleanData/phen_all")}
write.table(all_ba,"01.cleanData/phen_all/20200801_LLD_300OB_39BA_1437samples.tsv",sep = '\t')
```


## 4 Clean SV data
### 4.1 Get clean profiles

```{r 4.1}
# Change SV names
colnames(dsgv) <- changeSVname(colnames(dsgv))
colnames(vsgv) <- changeSVname(colnames(vsgv))

# Split dataset
dsgv_lld <- dsgv[grep('LLD', rownames(dsgv)),]
dsgv_ob  <- dsgv[grep('^G', rownames(dsgv)),]
vsgv_lld <- vsgv[grep('LLD', rownames(vsgv)),]
vsgv_ob  <- vsgv[grep('^G', rownames(vsgv)),]

# Clean 300OB profile
rownames(dsgv_ob) <- ob_id$ID[match(rownames(dsgv_ob), ob_id$G_id)]
rownames(vsgv_ob) <- ob_id$ID[match(rownames(vsgv_ob), ob_id$G_id)]

dsgv_ob<-dsgv_ob[grep('^C',rownames(dsgv_ob)),]
vsgv_ob<-vsgv_ob[grep('^C',rownames(vsgv_ob)),]

dsgv_ob<-dsgv_ob[match(rownames(ob_phen),rownames(dsgv_ob)),]
rownames(dsgv_ob)<-rownames(ob_phen)
vsgv_ob<-vsgv_ob[match(rownames(ob_phen),rownames(vsgv_ob)),]
rownames(vsgv_ob)<-rownames(ob_phen)

# Clean lld profile
dsgv_lld<-dsgv_lld[match(rownames(lld_basic), rownames(dsgv_lld)),]
rownames(dsgv_lld)<-rownames(lld_basic)
vsgv_lld<-vsgv_lld[match(rownames(lld_basic), rownames(vsgv_lld)),]
rownames(vsgv_lld)<-rownames(lld_basic)

# merge dataset
dsgv <- rbind(dsgv_lld, dsgv_ob)
vsgv <- rbind(vsgv_lld, vsgv_ob)

## Outputs
if(!dir.exists("01.cleanData")){dir.create("01.cleanData")}

if(!dir.exists("01.cleanData/SV_lld")){dir.create("01.cleanData/SV_lld")}
write.table(dsgv_lld,"01.cleanData/SV_lld/20200801_LLD_deletionStructuralVariation_1135samples.tsv",sep = '\t')
write.table(vsgv_lld,"01.cleanData/SV_lld/20200801_LLD_variableStructuralVariation_1135samples.tsv",sep = '\t')
save(dsgv_lld, file = "01.cleanData/SV_lld/dsgv_lld.RData")
save(vsgv_lld, file = "01.cleanData/SV_lld/vsgv_lld.RData")

if(!dir.exists("01.cleanData/SV_300OB")){dir.create("01.cleanData/SV_300OB")}
write.table(dsgv_ob,"01.cleanData/SV_300OB/20200801_300OB_deletionStructuralVariation_302samples.tsv",sep = '\t')
write.table(vsgv_ob,"01.cleanData/SV_300OB/20200801_300OB_variableStructuralVariation_302samples.tsv",sep = '\t')
save(dsgv_ob, file = "01.cleanData/SV_300OB/dsgv_ob.RData")
save(vsgv_ob, file = "01.cleanData/SV_300OB/vsgv_ob.RData")

if(!dir.exists("01.cleanData/SV_all")){dir.create("01.cleanData/SV_all")}
write.table(dsgv,"01.cleanData/SV_all/20200801_LLD_300OB_deletionStructuralVariation_1437samples.tsv",sep = '\t')
write.table(vsgv,"01.cleanData/SV_all/20200801_LLD_300OB_variableStructuralVariation_1437samples.tsv",sep = '\t')
save(dsgv, file = "01.cleanData/SV_all/dsgv.RData")
save(vsgv, file = "01.cleanData/SV_all/vsgv.RData")

```

### 4.2 Get name conversion table

```{r 4.2}
# Name conversion
organism<-str_replace_all(colnames(dsgv),"\\:\\d+_\\d+.*","") %>%
  .[!duplicated(.)]

Short_name<- organism %>% 
  str_replace_all('\\[','') %>%
  str_replace_all('\\]', '') %>%
  str_replace_all(' cf\\.','')

Short_name[grep(' sp\\.', organism, invert = F)] <- Short_name[grep(' sp\\.', organism, invert = F)] %>%
  str_replace_all('sp\\..*','sp')

Fst_letter<-Short_name[grep(' sp\\.', organism, invert = T)] %>%
  str_replace_all(' .*','') %>%
  str_sub(start = 1,end = 1)

Spe_name<-Short_name[grep(' sp\\.', organism, invert = T)] %>%
  str_extract_all(' .*') %>%
  str_replace_all('^ ', '') %>%
  str_replace_all(' .*', '')

Short_name[grep(' sp\\.', organism, invert = T)] <-paste(Fst_letter,'.', Spe_name, sep = '')

taxa_name<-data.frame(NCBI_taxonomy_id = taxonomy$X[match(organism,taxonomy$organism)],
                      organism = as.character(organism), 
                      Short_name = as.character(Short_name), stringsAsFactors = F)

taxa_name$Short_name[match('bacterium LF-3',taxa_name$organism)]<-'bacterium LF-3'

taxa_name<-left_join(taxa_name, ncbi, by = "NCBI_taxonomy_id")

if(!dir.exists("01.cleanData/SV_info")){dir.create("01.cleanData/SV_info")}
write.table(taxa_name, "01.cleanData/SV_info/20200801_LLD_300OB_Species_name.tsv", sep = "\t", col.names = T, row.names = F, quote = F)
```

### 4.3 Get SV annotation tables

```{r 4.3}
# SV annotation tables
dsgv_info_anno<-data.frame(dsgv_anno,
                           SV_ID=dsgv_anno$SV_id,
                           Taxonomy_Name = taxa_name$organism[match(str_replace_all(dsgv_anno$Taxonomy_id, '\\..*', ''),
                                                                    taxa_name$NCBI_taxonomy_id)],
                           SV_Name = changeSVname(dsgv_anno$SV_id),
                           Taxonomy_ID = dsgv_anno$Taxonomy_id,
                           SV_size = calcSVSize(dsgv_anno$SV_id))[,c(9,7,6,8,3,10,4,5)]

vsgv_info_anno<-data.frame(vsgv_anno,
                           SV_ID=vsgv_anno$SV_id,
                           Taxonomy_Name = taxa_name$organism[match(str_replace_all(vsgv_anno$Taxonomy_id, '\\..*', ''),
                                                                    taxa_name$NCBI_taxonomy_id)],
                           SV_Name = changeSVname(vsgv_anno$SV_id),
                           Taxonomy_ID = vsgv_anno$Taxonomy_id,
                           SV_size = calcSVSize(vsgv_anno$SV_id))[,c(9,7,6,8,3,10,4,5)]

write.table(dsgv_info_anno, "01.cleanData/SV_info/20200801_LLD_300OB_dsgv_info_anno.tsv", 
            sep = "\t", quote = F, col.names = T, row.names = F)
write.table(vsgv_info_anno, "01.cleanData/SV_info/20200801_LLD_300OB_vsgv_info_anno.tsv", 
            sep = "\t", quote = F, col.names = T, row.names = F)
```


### 4.4 Get species information table

```{r 4.4}
## Get SV number per species
species_dsgv_n<-str_replace_all(colnames(dsgv),"\\:\\d+_\\d+.*","") %>%
  table(.) %>%
  as.data.frame(.)
colnames(species_dsgv_n)<-c("Species","Deletion SVs number")
species_vsgv_n<-str_replace_all(colnames(vsgv),"\\:\\d+_\\d+.*","") %>%
  table(.) %>%
  as.data.frame(.)
colnames(species_vsgv_n)<-c("Species","Variable SVs number")

species_sgv_n<-full_join(species_dsgv_n, species_vsgv_n, by = "Species")
species_sgv_n[is.na(species_sgv_n)]<-0

NCBI_taxonomy_id<-species_sgv_n$Species %>%
  match(.,taxonomy$organism) %>%
  taxonomy$X[.]
species_sgv_n<-data.frame(NCBI_taxonomy_id, species_sgv_n)

## Get sample size per species
dsgv_infor_sample_n<-str_replace_all(colnames(dsgv),"\\:\\d+_\\d+.*","") %>%
  duplicated(.) %>%
  `!`%>%
  dsgv[,.] %>%
  is.na(.) %>%
  `!`%>%
  colSums(.) %>%
  as.data.frame(.)
colnames(dsgv_infor_sample_n) <- "Sample_number"
rownames(dsgv_infor_sample_n) <- rownames(dsgv_infor_sample_n) %>%
  str_replace_all(.,"\\:\\d+_\\d+.*", "")
dsgv_infor_sample_n<-data.frame(Species = rownames(dsgv_infor_sample_n),dsgv_infor_sample_n)

Taxonomy_name <- match(dsgv_infor_sample_n$Species,taxa_name$organism) %>%
  taxa_name$Short_name[.]
sample_n<-data.frame(Short_name=Taxonomy_name, dsgv_infor_sample_n)

## LLD sample size per species
lld_infor_sample_n<-str_replace_all(colnames(dsgv),"\\:\\d+_\\d+.*","") %>%
  duplicated(.) %>%
  `!`%>%
  dsgv[grep('^LLD',rownames(dsgv)),.] %>%
  is.na(.) %>%
  `!`%>%
  colSums(.) %>%
  as.data.frame(.)
colnames(lld_infor_sample_n) <- "LLD"
rownames(lld_infor_sample_n) <- rownames(lld_infor_sample_n) %>%
  str_replace_all(.,"\\:\\d+_\\d+.*", "")
lld_infor_sample_n<-data.frame(Species = rownames(lld_infor_sample_n),lld_infor_sample_n)

## 300OB sample size per species
OB_infor_sample_n<-str_replace_all(colnames(dsgv),"\\:\\d+_\\d+.*","") %>%
  duplicated(.) %>%
  `!`%>%
  dsgv[grep('^LLD',rownames(dsgv),invert = T),.] %>%
  is.na(.) %>%
  `!`%>%
  colSums(.) %>%
  as.data.frame(.)
colnames(OB_infor_sample_n) <- "300OB"
rownames(OB_infor_sample_n) <- rownames(OB_infor_sample_n) %>%
  str_replace_all(.,"\\:\\d+_\\d+.*", "")
OB_infor_sample_n<-data.frame(Species = rownames(OB_infor_sample_n),OB_infor_sample_n)

## merge LLD and 300OB sample size
infor_sample_n <- cbind(lld_infor_sample_n,OB_infor_sample_n,sample_n)[,-c(3,6)] %>% .[,c(1,4,2,3,5)]

## Merge sample size and SV number information
species_sample_n<-dplyr::full_join(species_sgv_n,infor_sample_n, by = "Species")
taxa_length$Species<-str_replace_all(taxa_length$Species, '\\..*', '')
species_sample_n$NCBI_taxonomy_id<-as.character(species_sample_n$NCBI_taxonomy_id)
species_sample_n<-dplyr::left_join(species_sample_n, taxa_length, by = c("NCBI_taxonomy_id"="Species"))
species_sample_n<-data.frame(species_sample_n,
                             SVs.number = species_sample_n[,3]+species_sample_n[,4])

## Merge all information
Informative_species_information <- match(species_sample_n$NCBI_taxonomy_id, taxonomy$X)%>%
  taxonomy[.,] %>%
  cbind(.,species_sample_n)

info <- full_join(Informative_species_information[,-11],
                                             taxa_name,
                                             by = 'organism')[,c(1:9,20,21,11,12,18,14,15,16,17)]

colnames(info)[c(1,10:18)]<-c("NCBI_taxonomy_id","Short_name","NCBI_bioproject_accession",   "Deletion_SVs_number", "Variable_SVs_number","SVs_number","LLD_sample_number","300OB_sample_number","Total_samples_number", "Length")

write.table(info, "01.cleanData/SV_info/20200801_LLD_300OB_Informative_species_information.tsv", col.names = T, row.names = F, sep = "\t", quote = F)
```

### 4.5 Get distance matrices
#### 4.5.1 All samples

```{r 4.5.1, eval=FALSE}
## msv (vsv+dsv) distance
sgv<-cbind(vsgv, dsgv)
all_shared_sv_dis<-shared_sv_dis(sgv)
save(all_shared_sv_dis, file = "01.cleanData/SV_all/all_shared_sv_dis.RData")

## SV distance matrices of all species
all_msv_dist<-NULL

for (i in c(1:nrow(info))){
  #i<-16
  file_name<-str_replace_all(info$organism[i], "\\/", "\\.")
  spe_name<-str_replace_all(info$organism[i],"\\[","\\\\\\[") %>% str_replace_all("\\]","\\\\\\]")
  vsgv_i<-vsgv[,grep(spe_name,colnames(vsgv))]
  dsgv_i<-dsgv[,grep(spe_name,colnames(dsgv))]
  all_msv_i<-cbind(vsgv_i,dsgv_i) %>%
    na.omit(.)

  all_msv_dist_i <- as.matrix(vegdist(as.data.frame(all_msv_i),method = "canberra"))
  all_msv_dist[[i]]<-all_msv_dist_i
}

names(all_msv_dist)<-paste('msv_',info$organism, sep = '')
all_msv_dist_std <- lapply(all_msv_dist, myfun<-function(x){x/max(x,na.rm = T)})

if(!dir.exists("01.cleanData/SV_all/distMat")){dir.create("01.cleanData/SV_all/distMat")}
save(all_msv_dist, file = "01.cleanData/SV_all/all_msv_dist.RData")
save(all_msv_dist_std, file = "01.cleanData/SV_all/all_msv_dist_std.RData")
```

#### 4.5.2 LLD samples

```{r 4.5.2, eval=FALSE}
## msv (vsv+dsv) distance
lld_msv_dist<-NULL

for (i in c(1:nrow(Informative_species_information))){
  #i<-16
  file_name<-str_replace_all(Informative_species_information$organism[i], "\\/", "\\.")
  spe_name<-str_replace_all(Informative_species_information$organism[i],"\\[","\\\\\\[") %>% str_replace_all("\\]","\\\\\\]")
  vsgv_i<-vsgv_lld[,grep(spe_name,colnames(vsgv_lld))]
  dsgv_i<-dsgv_lld[,grep(spe_name,colnames(dsgv_lld))]
  lld_msv_i<-cbind(vsgv_i,dsgv_i) %>%
    na.omit(.)
  
  lld_msv_dist_i <- as.matrix(vegdist(as.data.frame(lld_msv_i),method = "canberra"))
  lld_msv_dist[[i]]<-lld_msv_dist_i
}

names(lld_msv_dist)<-paste('msv_',Informative_species_information$organism, sep = '')
lld_msv_dist_std <- lapply(lld_msv_dist, myfun<-function(x){x/max(x,na.rm = T)})

save(lld_msv_dist, file = "01.cleanData/SV_lld/lld_msv_dist.RData")
save(lld_msv_dist_std, file = "01.cleanData/SV_lld/lld_msv_dist_std.RData")
```

#### 4.5.3 300OB samples

```{r 4.5.3, eval=FALSE}
## msv (vsv+dsv) distance
ob_msv_dist<-NULL

for (i in c(1:nrow(Informative_species_information))){
  #i<-16
  file_name<-str_replace_all(Informative_species_information$organism[i], "\\/", "\\.")
  spe_name<-str_replace_all(Informative_species_information$organism[i],"\\[","\\\\\\[") %>% str_replace_all("\\]","\\\\\\]")
  vsgv_i<-vsgv_ob[,grep(spe_name,colnames(vsgv_ob))]
  dsgv_i<-dsgv_ob[,grep(spe_name,colnames(dsgv_ob))]
  ob_msv_i<-cbind(vsgv_i,dsgv_i) %>%
    na.omit(.)
  
  ob_msv_dist_i <- as.matrix(vegdist(as.data.frame(ob_msv_i),method = "canberra"))
  ob_msv_dist[[i]]<-ob_msv_dist_i
}

names(ob_msv_dist)<-paste('msv_',Informative_species_information$organism, sep = '')
ob_msv_dist_std <- lapply(ob_msv_dist, myfun<-function(x){x/max(x,na.rm = T)})

save(ob_msv_dist, file = "01.cleanData/SV_300OB/ob_msv_dist.RData")
save(ob_msv_dist_std, file = "01.cleanData/SV_300OB/ob_msv_dist_std.RData")
```

## 5 Clean species abundance data

```{r 5}
sv_spe_taxid<-info$NCBI_taxonomy_id
sv_spe_taxid[sv_spe_taxid==245018]<-649756

# relative abundance of species detected with SVs in LLD
lld_SV_abun_s[,-c(1:2)]<-apply(lld_SV_abun_s[,-c(1:2)], 2, myfun<-function(x){x/sum(x)})
lld_s_abun <- sv_spe_taxid %>% 
  match(., tax_relationship$V2) %>% 
  tax_relationship$V1[.] %>% 
  match(., lld_SV_abun_s$NCBI_tax_id) %>% 
  lld_SV_abun_s[., -c(1:2)] %>% 
  t %>%
  as.data.frame

colnames(lld_s_abun)<-info$organism
rownames(lld_s_abun)<-str_replace_all(rownames(lld_s_abun), ".S.bracken", "")
lld_s_abun<-lld_s_abun[match(intersect(str_replace_all(lld_id$LLD_bam_id, "fece_", ""),
                                        rownames(lld_s_abun)),
                              rownames(lld_s_abun)),]
rownames(lld_s_abun)<-lld_id$LLD_GoNL_all_id[match(rownames(lld_s_abun), str_replace_all(lld_id$LLD_bam_id, "fece_", ""))]
lld_s_abun<-lld_s_abun[match(rownames(lld_basic),rownames(lld_s_abun)),]
rownames(lld_s_abun)<-rownames(lld_basic)

# relative abundance of species detected with SVs in 300-OB
ob_SV_abun_s[,-c(1:2)]<-apply(ob_SV_abun_s[,-c(1:2)], 2, myfun<-function(x){x/sum(x)})
ob_s_abun <- sv_spe_taxid %>% 
  match(., tax_relationship$V2) %>% 
  tax_relationship$V1[.] %>% 
  match(., ob_SV_abun_s$NCBI_tax_id) %>% 
  ob_SV_abun_s[., -c(1:2)] %>% 
  t %>%
  as.data.frame

colnames(ob_s_abun)<-info$organism
rownames(ob_s_abun)<-str_replace_all(rownames(ob_s_abun), ".S.bracken", "")

ob_s_abun<-ob_s_abun[match(intersect(ob_id$G_id, rownames(ob_s_abun)), rownames(ob_s_abun)),]
rownames(ob_s_abun)<-ob_id$ID[match(rownames(ob_s_abun), ob_id$G_id)]

ob_s_abun<-ob_s_abun[match(rownames(ob_phen),rownames(ob_s_abun)),]
rownames(ob_s_abun)<-rownames(ob_phen)

# merge dataset
all_s_abun <- rbind(lld_s_abun, ob_s_abun)

# outputs
if(!dir.exists("01.cleanData/mbio_lld")){dir.create("01.cleanData/mbio_lld")}
write.table(lld_s_abun, "01.cleanData/mbio_lld/LLD_SV_55species_abun_1135samples.tsv",sep = '\t')
if(!dir.exists("01.cleanData/mbio_300OB")){dir.create("01.cleanData/mbio_300OB")}
write.table(ob_s_abun, "01.cleanData/mbio_300OB/300OB_SV_55species_abun_302samples.tsv",sep = '\t')
if(!dir.exists("01.cleanData/mbio_all")){dir.create("01.cleanData/mbio_all")}
write.table(all_s_abun, "01.cleanData/mbio_all/LLD_300OB_SV_55species_abun_1437samples.tsv",sep = '\t')

mean(rowSums(all_s_abun,na.rm = T),na.rm = T) # Mean of total relative abundance of species detected with SVs: 0.825154
rowSums(all_s_abun,na.rm = T)[rowSums(all_s_abun,na.rm = T)!=0] %>% min # Minimum of total relative abundance of species detected with SVs: 0.43729
max(rowSums(all_s_abun,na.rm = T),na.rm = T) # Maximum of total relative abundance of species detected with SVs: 0.94705
```
