---
title: "Metabolites_analysis_v5"
author: "Arnau Vich"
date: "6/27/2020"
output: html_document
---


Introduction
===

Untargeted metabolomics
---

In this project we aim to investigate the relation between fecal metabolites and inflammatory bowel diseases phenotypes. The cohort consist of 750 samples of different patients, devided in 255 controls from the population cohort LifeLines (NL), 270 patients with Crohns disease and 211 patients with ulcerative colitis from the 1000IBD cohort (NL). 

Metabolites were measured using METABOLON platform combining untargeted metabolite detection with short chain fatty acids measurments. Untargeted metabolite detection consisted on an ultrahigh performance liquid chromatography tandem mass spectrocopy (UPLC-MS/MS): 2 acidic positive ion condition (C18 columns) + 1 Basic negative ion optimazed (C18 column) + 1 Basic negative ionization (Hilic column).

Metabolites will be combined with the following data layers: matching shot-gun sequencing metagenomics (from the same sample), host phenotypes at time of sampling (including disease phenotypes), diet (food frequency questionnaires) & host genetics (GSA and WES data). 


SCFA quantification [METABOLON description]
---

Human feces samples are analyzed for eight short chain fatty acids: acetic acid (C2), propionic acid (C3), isobutyric acid (C4), butyric acid (C4), 2-methyl-butyric acid (C5), isovaleric acid (C5), valeric acid (C5) and caproic acid (hexanoic acid, C6) by LC-MS/MS.


Human feces samples are spiked with stable labelled internal standards and are homogenized and subjected to protein precipitation with an organic solvent. 

The peak area of the individual analyte product ions is measured against the peak area of the product ions of the corresponding internal standards. Quantitation is performed using a weighted linear least squares regression analysis generated from fortified calibration standards prepared immediately prior to each run.


The QC Endogenous was prepared from a single lot of human stool. All analytes are at the endogenous level except hexanoic acid, which was spiked in due to low levels in this lot. The other three levels of QC were prepared by spiking phosphate buffered saline (PBS) with all analytes to achieve the appropriate concentrations for each level.


Sample analysis was carried out in a 96-well plate format containing two calibration curves and eight QC samples (per plate) to monitor assay performance. Twelve batches were prepared and analyzed.


Precision was evaluated using the corresponding QC replicates in the sample runs. QCs met acceptance criteria at all levels for all analytes (Targeted acceptance criteria: at least 50% of QC samples at each concentration level per analyte should be within ±20.0% of the running mean, and at least 2/3 of all QC samples per analyte should fall within ±20.0% of the corresponding running mean.).


A total of 750 human stool samples were analyzed.
Analyte concentrations that fell below and above the limit of quantitation are extrapolated, and the reported values given a comment of BLOQ or ALOQ, respectively. 

Analytes that cannot be extrapolated below the limit of quantitation are considered not quantifiable and are noted as N/Q. Samples that did not have sufficient quantity are noted as QNS.


Load libraries
---

```{r, warning=F,message=FALSE, echo=TRUE}
library(corrplot)
library(plotly)
library(pheatmap)
library(NbClust)
library(reshape2)
library(ggrepel)
library(dplyr)
library(eulerr)
library(vegan)
library(ape)
library(ggridges)
library (UpSetR)
library(ggplot2)
library(psych)
library(ggord)
library(data.table)
library(mice)
library(caret)
library(factoextra)
library(Rtsne)
library(heatmaply)
library(ggpubr)
library(ggbiplot)
library(coin)
library(DT)
library(pvclust)
library(glmnet)
library(compositions)
library(RColorBrewer)
library(iClusterPlus)
library (ggvegan)
library (VIM)
```

Create functions
---

```{r}
####################################
#Inverse rank transformation function.
####################################

invrank= function(x) {qnorm((rank(x,na.last="keep")-0.5)/sum(!is.na(x)))}

###############################
#negate in (to exclude columns)
###############################

`%ni%` <- Negate(`%in%`)

#####################################
#Imputation and filtering metabolites
#####################################

impute_missing_values=function(x, samples_row=T, method="knn", missing_filter=0){
  #if samples are in columns transpose
  if (!samples_row){
    x=as.data.frame(t(x))
  } 
  #Exclude/keep columns that pass the missigness threshold
  if (missing_filter>1){
    stop("\n Hey! \n Values should be a proportion of missing values allowed per column: a value from 0 to 1") 
  }
  col_ori=ncol(x)
  col_keep=colnames(x)[colSums(!is.na(x))/nrow(x)>missing_filter]
  x=x[,col_keep]
  my_num_removed=col_ori-ncol(x)
  warning (paste(my_num_removed, "columns removed due to many missing values"))
  if (method=="zero"){
    x[is.na(x)]=0
  }
  else if (method=="min"){
    x[sapply(x, is.numeric)] <- lapply(x[sapply(x, is.numeric)], function(a) ifelse(is.na(a), min(a, na.rm = TRUE), a))
  }
  else if (method=="hfmin"){
    x[sapply(x, is.numeric)] <- lapply(x[sapply(x, is.numeric)], function(a) ifelse(is.na(a), min(a, na.rm = TRUE)/2, a))
  }
  else if (method=="mean"){
    x[sapply(x, is.numeric)] <- lapply(x[sapply(x, is.numeric)], function(a) ifelse(is.na(a), mean(a, na.rm = TRUE), a))
  }
  else if (method=="median"){
    x[sapply(x, is.numeric)] <- lapply(x[sapply(x, is.numeric)], function(a) ifelse(is.na(a), median(a, na.rm = TRUE), a))
  }
  else if (method=="knn"){
    x=kNN(x,k=10,metric = "euclidean", imp_var = F)
  }
  else if (method=="none"){
    x=x
  }
  else{
    stop("\n Hey! \n Method parameters for imputing missing values per column are: \n min -> min (col), hfmin -> min(col)/2, mean -> mean (col), median -> median(col), mice -> using mice package [Multivariate Imputation By Chained Equations] defaults \n")
  }
  return(as.data.frame(x))
}


##############################
# Summary stats
##############################

summary_stats=function(feature_matrix,pheno, missing_vals=0){
  a=1
  num_var=length(levels(pheno[,1]))
  print (paste("Number of phenotypes detected for summary statistics:", num_var))
  summary_met=matrix(nrow=ncol(feature_matrix),ncol=(9*(num_var+1))+1)
  for (i in 1:ncol(feature_matrix)){
    #print (colnames(feature_matrix)[i])
    if (missing_vals==0){
      pos=2
      summary_met[i,pos]=sum(feature_matrix[,i]!=0)
      pos=pos+1
      summary_met[i,pos]=sum(feature_matrix[,i]==0)
    }
    else {  
      pos=2
      summary_met[i,pos]=sum(is.na(feature_matrix[,i]))
      pos=pos+1
      summary_met[i,pos]=sum(!is.na(feature_matrix[,i]))
    }
    if (sum(is.na(feature_matrix[,i]) == nrow(feature_matrix))){
      pos=pos+1
      summary_met[i,pos]="NA"
      pos=pos+1
      summary_met[i,pos]="NA"
      pos=pos+1
      summary_met[i,pos]="NA"
      pos=pos+1
      summary_met[i,pos]="NA"
      pos=pos+1
      summary_met[i,pos]="NA"
      pos=pos+1
      summary_met[i,pos]="NA"
      pos=pos+1
      summary_met[i,pos]="NA"
    }
    else{
      pos=pos+1
      summary_met[i,pos]=min(feature_matrix[,i], na.rm = T)
      pos=pos+1
      summary_met[i,pos]=max(feature_matrix[,i], na.rm = T)
      pos=pos+1
      summary_met[i,pos]=mean(feature_matrix[,i], na.rm = T)
      pos=pos+1
      summary_met[i,pos]=median(feature_matrix[,i], na.rm = T)
      pos=pos+1
      summary_met[i,pos]=sd(feature_matrix[,i], na.rm = T)
      pos=pos+1
      summary_met[i,pos]=length(boxplot(feature_matrix[,i], plot=FALSE, range = 3)$out)
      if (sum(!is.na(feature_matrix[,i])) < 10){
        pos=pos+1
        summary_met[i,pos]="NA"
        } else {
        nor=shapiro.test(feature_matrix[,i])
        pos=pos+1
        summary_met[i,pos]=nor$p.value>=0.05
      }
    }
    for (x in levels (pheno[,1])){
      id_list=rownames(pheno)[pheno[,1]==x]
      tmp_all=feature_matrix[rownames(feature_matrix)%in%id_list,]
      pos=pos+1
      summary_met[i,pos]=sum(is.na(tmp_all[,i]))
      pos=pos+1
      summary_met[i,pos]=sum(!is.na(tmp_all[,i]))
      if (sum(is.na(tmp_all[,i]) == nrow(tmp_all))){
        pos=pos+1
        summary_met[i,pos]="NA"
        pos=pos+1
        summary_met[i,pos]="NA"
        pos=pos+1
        summary_met[i,pos]="NA"
        pos=pos+1
        summary_met[i,pos]="NA"
        pos=pos+1
        summary_met[i,pos]="NA"
        pos=pos+1
        summary_met[i,pos]="NA"
        pos=pos+1
        summary_met[i,pos]="NA"
      }else{
        pos=pos+1
        summary_met[i,pos]=min(tmp_all[,i], na.rm = T)
        pos=pos+1 
        summary_met[i,pos]=max(tmp_all[,i], na.rm = T)
        pos=pos+1 
        summary_met[i,pos]=mean(tmp_all[,i], na.rm = T)
        pos=pos+1
        summary_met[i,pos]=median(tmp_all[,i], na.rm = T)
        pos=pos+1 
        summary_met[i,pos]=sd(tmp_all[,i], na.rm = T)
        pos=pos+1 
        summary_met[i,pos]=length(boxplot(tmp_all[,i], plot=FALSE, range = 3)$out)
        if (sum(!is.na(tmp_all[,i])) < 10){
          pos=pos+1 
          summary_met[i,pos]="NA"
          } else {
          #nor=NA
          nor=shapiro.test(tmp_all[,i])
          pos=pos+1 
          summary_met[i,pos]=nor$p.value>=0.05
          }
        }
      }
    }
  summary_met[,1]=colnames(feature_matrix)
  my_colnames=c("Metabolite", "NAs","Non_NAs", "Min", "Max", "Mean", "Median","SD", "Outliers_x3IQR","Normal_distrib")
  for (o in levels (pheno[,1])){
    my_colnames=c(my_colnames,paste("NAs",o, sep="_"))
    my_colnames=c(my_colnames,paste("Non_NAs",o,sep="_"))
    my_colnames=c(my_colnames,paste("Min",o,sep="_"))
    my_colnames=c(my_colnames,paste("Max",o,sep="_"))
    my_colnames=c(my_colnames,paste("Mean",o,sep="_"))
    my_colnames=c(my_colnames,paste("Median",o,sep="_"))
    my_colnames=c(my_colnames,paste("SD",o,sep="_"))
    my_colnames=c(my_colnames,paste("Outliers_x3IQR",o,sep="_"))
    my_colnames=c(my_colnames,paste("Normal_distrib",o,sep="_"))
  }
  colnames(summary_met)=my_colnames
  return(summary_met)
}  



###############################################
# Filtering and transforming taxa             #
###############################################

transform_and_filter_taxa=function(x, samples_row=T, method="asin", missing_filter=0){
  x[x=="NA"]=0
  x[x=="NaN"]=0
  #if samples are in columns transpose
  if (!samples_row){
  
    x=as.data.frame(t(x))
  
  } 
  #Exclude/keep columns that pass the missigness threshold
  if (missing_filter>100){
  
    stop("\n Hey! \n Values should be a proportion of missing values allowed per column: a value from 0 to 100") 
  
  }
  
  x_filt=x[,((colSums(x !=0) / nrow(x)) *100 )>missing_filter]
  my_num_removed=ncol(x)-ncol(x_filt)
  print (paste(my_num_removed, "species removed due to many missing values"))
  
  if (method=="asin"){

    x_filt=x_filt/100
    x_filt=asin(sqrt(x_filt))

  } else if (method=="log"){
    
    #replace 0 by the half of the smallest value observed
    my_min=min(x_filt[x_filt>0])/2
    x_filt=x_filt+my_min
    x_filt=log10(x_filt)

  }else if (method=="clr"){
  
    #x_filt=x_filt/100
    #replace 0 by the half of the smallest value observed
    my_min=min(x_filt[x_filt>0])/2
    x_filt=x_filt+my_min
    d <- t(apply(x_filt, 1, function(b) {
      my_mean=exp(sum(log(b))/length(b))
      log(b/my_mean)
    }))
    x_filt=d
  
  }
  return(as.data.frame(x_filt))
}



###############################################
# Calculate ratios                           #
###############################################


calculate_all_ratios=function(x,is_log=T,feature_in_column=T){
  print (paste("Number of features detected:",ncol(x)))
  if(feature_in_column==F){
    x=as.data.frame(t(x))
    print ("transposing")
  }
  flag=1
  n=1
  l=c()
  for (a in 1:(ncol(x)-1)){
    for (b in (a+1):ncol(x)){
      my_ratio_names=paste(colnames(x)[a],colnames(x)[b], sep = "_vs_")
      l[[n]]=my_ratio_names
      n=n+1
      if (is_log==T){
        my_ratios=x[,a]-x[,b]
        #colnames(my_ratios)=my_ratio_names
      }else{
        my_ratios=log10(x[,a]/x[,b])
        #colnames(my_ratios)=my_ratio_names
      }
      if (flag==1){
        my_out=my_ratios
        #colnames(my_out)=my_ratio_names
        flag=5
      }else{
        my_out=cbind(my_out,my_ratios)
      }
    }
  }
  my_out=as.data.frame(my_out)
  colnames(my_out)=l
  rownames(my_out)=row.names(x)
  return(my_out)
}





```


1. Import data
===

a) Metabolites: Here we use both raw values provided by Metabolon (metabolites AUC).


b) Short Chain Fatty Acids (SFCA): Together with the untargetted metabolite assessment, SCFA concentrations were measured in the same fecal samples (μg/g)


c) Host phenotypes: Information about the host (age,sex,BMI, diet, etc.). This consist of 3 files: phenotypes shared and used for the case-control analysis, phenotypes specific for the population controls & phenotypes specific for cases (disease location, activity, etc.)


d) Experiment variables: Different technical variables from the untargetted metabolic experiments. We will use this information as a potential counfounding factors. 


e) Genetics. 


f) Metagenomics taxa: MetaPhlan's profiles (v2.9)


g) Pathway annotation provided by Metabolon


h) ID's for each data modality: Allows to combine different datasets / information 

```{r, warning, message=F, echo=T}
#transformed metabolites (raw data transformed to adjust distrubution, medians = 1)
#all_metabolites <- read.delim("~/Documents/Project_Metabolome/1.Input_files/all_metabolites2.txt", row.names=1)

# a) Raw values (AUC of peaks)

all_metabolites_raw <- read.delim("~/Documents/Project_Metabolome/1.Input_files/all_metabolites_raw2.txt", row.names=1)

# b) SCFA

SCFA <- read.delim("~/Documents/Project_Metabolome/1.Input_files/old/SCFA.txt")

# c) Phenotypes

phenos_ibd=read.delim("~/Documents/Project_Metabolome/1.Input_files/Merged_IBD_phenos_clean_v2.txt",row.names = 1)
phenos_lld=read.delim("~/Documents/Project_Metabolome/1.Input_files/Merged_LLD_phenos_clean_v2.txt", row.names=1)
cc_pheno=read.delim("~/Documents/Project_Metabolome/1.Input_files/case_control_meta_v1.txt", row.names=1)
cc_stoma=cc_pheno[,c("ibd_CurrentStomaOrPouch"),drop=F]

# d) Experiment variables 

my_batch=read.delim("~/Documents/Project_Metabolome/1.Input_files/Metabolon_preparation_v2.txt", row.names=1)


# e) Genetics


# f) Metagenomic taxa (add metagenomic genes/pathways)

#MetaPhlAn 2.7
#IBD_taxa <- read.delim("~/Documents/Project_Metabolome/2.taxa/IBD_taxonomy_unstrat_clean.txt", row.names=1)
#LLD_taxa <- read.delim("~/Documents/Project_Metabolome/2.taxa/LLD_taxonomy_unstrat_clean.txt", row.names=1)

#mOTUs2
#my_taxa_motus = read.delim("~/Documents/Project_Metabolome/2.taxa/mOTUs_prediction_v2.txt", row.names=1, check.names = F)

#MetaPhlAn v3
my_taxa = read.delim("~/Documents/Project_Metabolome/2.taxa/taxa_metabolomics_samples_mpa3_3.txt", row.names=1, check.names = F)

# g) Metabolites annotation
annot <- read.delim("~/Documents/Project_Metabolome/1.Input_files/info_metabolites_v2.txt", row.names=1)
annot$SUPER.PATHWAY.1=NULL


# h) ID's index (translate from different data modalities metabolomics <-> metagenomic <-> phenotypes)

IDs_LLD <- read.delim("~/Documents/Project_Metabolome/1.Input_files/IDs_LLD.txt", header=FALSE)
IDs_IBD <- read.delim("~/Documents/Project_Metabolome/1.Input_files/IDs_IBD.txt", header=FALSE)
```

Check phenotypes files (count NA's)
---

```{r}
na_count = sapply(cc_pheno, function(y) sum(length(which(is.na(y)))))
na_count=data.frame(na_count)
cc_pheno$host_BMI[is.na(cc_pheno$host_BMI)]=median(cc_pheno$host_BM, na.rm = T)
cc_pheno$med_ACE_inhibitor[is.na(cc_pheno$med_ACE_inhibitor)]="Non_user"
cc_pheno$med_PPI[is.na(cc_pheno$med_PPI)]="Non_user"

diet=colnames(cc_pheno)[grep("diet",colnames(cc_pheno))]

for (a in diet){
  cc_pheno[,a][is.na(cc_pheno[,a])]=median(cc_pheno[,a], na.rm = T)
}

cc_pheno$host_smk_current[is.na(cc_pheno$host_smk_current)]="no"

na_count = sapply(cc_pheno, function(y) sum(length(which(is.na(y)))))
na_count=data.frame(na_count)

```


Adjust all ids
---

__Translate all id's to the same id (easier to deal with multiple tables)__


```{r, warning=F,message=FALSE, echo=TRUE}

#Change Metabolon ids to UMCG ids to connect later to phenotypes 
all_raw=as.data.frame(t(all_metabolites_raw))

# Select IDs
IDs_IBD=IDs_IBD[,c(1,5)]
IDs_LLD$V2=NULL

# [IMPORTANT] Table with matching ids between UMCG and Metabolon
IDs=data.frame(rbind(as.matrix(IDs_IBD), as.matrix(IDs_LLD)))
rownames(IDs)=IDs$V1
IDs$V1=NULL

# Merge to replace the ids Metabolon => UMCG

all_new_ID_raw=merge(IDs,all_raw, by="row.names")
rownames(all_new_ID_raw)=all_new_ID_raw$V5
all_new_ID_raw$Row.names=NULL
all_new_ID_raw$V5=NULL

#Repeat for the batch information: Metabolon ID => UMCG ID

my_batch_id=merge(IDs,my_batch, by="row.names")
rownames(my_batch_id)=my_batch_id$V5
my_batch_id$Row.names=NULL
my_batch_id$V5=NULL

#In case that we use the transformed data provided by Metabolon

#all=as.data.frame(t(all_metabolites))
#all_new_ID=merge(IDs,all, by="row.names")
#row.names(all_new_ID)=all_new_ID$V5
#all_new_ID$V5=NULL
#all_new_ID$Row.names=NULL
#all_new_ID=all_new_ID[row.names(all_new_ID)%ni%remove_pouch_stoma,]
#all_new_ID_with_stoma=all_new_ID

tract_id=data.frame(row.names(all_new_ID_raw))
rownames(tract_id)=tract_id$row.names.all_new_ID_raw.
tract_id$In_metabolomics="Yes"
tract_id$row.names.all_new_ID_raw.=NULL
tract_id=merge(tract_id,cc_stoma, by="row.names")
rownames(tract_id)=tract_id$Row.names
tract_id$Row.names=NULL

```


2. Summary statistics
===


Annotation by METABOLON
---

Describe the categories of metabolites
---

```{r, warning=F,message=FALSE, echo=F}
met_2=as.data.frame(table(annot$SUPER.PATHWAY))
met_1=as.data.frame(table(annot$SUB.PATHWAY))

yy=ggplot(met_2, aes(x = reorder(Var1, -Freq),Freq)) + geom_bar(stat = "identity", color="black") + theme_classic() + theme(axis.title.x = element_blank(),axis.text.y = element_text(size = 12)) + coord_flip() + xlab("") + ylab("Number of metabolites")

ggplotly(yy)
```

```{r, warning=F,message=FALSE, echo=F,fig.height=20}
xx=ggplot(met_1, aes(x = reorder(Var1, -Freq),Freq)) + geom_bar(stat = "identity", color="black") + theme_classic() + theme(axis.title.x = element_blank(),axis.text.y = element_text(size = 5)) + coord_flip() + xlab("") + ylab("Number of metabolites")

ggplotly(xx)
```


Remove participants with a pouch or stoma
---

__Rational: Patients with pouch or stoma do not capture the colonic metabolite/microbiota profile__

```{r}
remove_pouch_stoma=rownames(phenos_ibd)[phenos_ibd$ibd_CurrentStomaOrPouch=="yes"]
#Keep data of the participants with stoma
all_new_ID_raw_with_stoma=all_new_ID_raw
#Remove
all_new_ID_raw=all_new_ID_raw[row.names(all_new_ID_raw)%ni%remove_pouch_stoma,]
cc_pheno=cc_pheno[row.names(cc_pheno)%ni%remove_pouch_stoma,]
print (paste("Number of samples excluded due to pouch or stoma:", length(remove_pouch_stoma)))
print (paste("Number of samples left:", length(rownames(all_new_ID_raw))))
```


Summary stats per metabolite per cohort
---

```{r, echo=T, message=F, warning=F}

#Raw data
select=cc_pheno[,1, drop=F]
summary_met=summary_stats(all_new_ID_raw,select, missing_vals = NA)
summary_met=as.data.frame(summary_met)
#write.table(summary_met, "~/Desktop/Metabolomics_for_cluster/3.Preliminary_results_v2/metabolites_summary_2020.txt", sep = "\t", quote = F)
```


__Check prevalence per phenotype__

```{r, echo=T, message=F, warning=F}
summary_short=summary_met[,c("Metabolite", "Non_NAs")]
summary_short2=summary_met[,c("Metabolite","Non_NAs","Non_NAs_Control","Non_NAs_CD", "Non_NAs_UC")]
summary_short2$prevalence_all=((as.numeric(as.character(summary_short2$Non_NAs))/nrow(all_new_ID_raw))*100)
summary_short2$prevalence_Control=((as.numeric(as.character(summary_short2$Non_NAs_Control))/length(select[select$ibd_Diagnosis=="Control",])*100))
summary_short2$prevalence_CD=((as.numeric(as.character(summary_short2$Non_NAs_CD))/length(select[select$ibd_Diagnosis=="CD",])*100))
summary_short2$prevalence_UC=((as.numeric(as.character(summary_short2$Non_NAs_UC))/length(select[select$ibd_Diagnosis=="UC",])*100))
summary_short2$Non_NAs=NULL
summary_short2$Non_NAs_Control=NULL
summary_short2$Non_NAs_CD=NULL
summary_short2$Non_NAs_UC=NULL
summary_short2[,-1] <-round(summary_short2[,-1],0)
summary_short3=melt(summary_short2)
```

__Presence of each metabolite per sample__

```{r, echo=F}
aa=ggplot(summary_short, aes(reorder(Metabolite,-as.numeric(as.character(Non_NAs))),as.numeric(as.character(Non_NAs))))+ geom_bar(color="darkblue",stat = "identity") + theme_classic() + ylab("Number of samples (n=682)") + xlab("Metabolites") +  theme(axis.text.x=element_blank(), axis.ticks.x=element_blank())

ggplotly(aa)
```

__Percentage of prevalence of metabolites per cohort__

```{r, echo=F}
bb=ggplot(summary_short3, aes(value, fill=variable)) + geom_histogram(position = "dodge", bins = 10, binwidth = 7, color="black") + theme_bw() + scale_fill_manual(values = c("grey79","white","blue2", "firebrick2")) + ylab("Number of metabolites") + xlab("% presence in samples")

ggplotly(bb)
```

__Prevalence/Missigness per metabolite per cohort with annotations__

```{r, echo=F}
summary_short4=summary_short2
rownames(summary_short4)=summary_short4$Metabolite
summary_short4$Metabolite=NULL
summary_short4=merge(annot,summary_short4, by="row.names")
rownames(summary_short4)=summary_short4$Row.names
summary_short4$Row.names=NULL
colnames(summary_short4)=c("Super pathway","Annotation", "Detected in cohort (%)", "Detected in controls (%)", "Detected in CD (%)", "Dectected in UC (%)")
datatable(summary_short4)
#write.table(summary_short4, "~/Documents/Project_Metabolome/4.Summary_stats/Metabolites_prevalence_short_2020.txt", sep = "\t", quote = F)
```


__Number of metabolites per samples__

```{r, warning=F,message=FALSE, echo=TRUE}

all_new_ID_raw2=as.data.frame(t(all_new_ID_raw))
summary_sample=matrix(nrow=ncol(all_new_ID_raw2), ncol=3)

for (i in 1:ncol(all_new_ID_raw2)){
  summary_sample[i,2]=sum(is.na(all_new_ID_raw2[,i]))
  summary_sample[i,3]=sum(!is.na(all_new_ID_raw2[,i]))
}

summary_sample[,1]=colnames(all_new_ID_raw2)
summary_sample=as.data.frame(summary_sample)
colnames(summary_sample)=c("Metabolite", "NAs","Non_NAs")
summary_sample$perc_detected=((as.numeric(as.character(summary_sample$Non_NAs))/nrow(all_new_ID_raw2))*100)
rownames(summary_sample)=summary_sample$Metabolite
summary_sample$Metabolite=NULL
summary_sample2=merge(summary_sample,cc_pheno,by="row.names")
#write.table(summary_sample, "~/Documents/Project_Metabolome/4.Summary_stats/Metabolites_per_sample_2020.txt", sep = "\t", quote = F)

```


```{r, echo=F}
cc=ggplot(summary_sample2, aes(perc_detected, fill=ibd_Diagnosis)) + geom_histogram(position = "dodge", bins = 100, binwidth = 7, color="black") + theme_bw() + scale_fill_manual(values = c("firebrick2","white","grey42","blue2")) + ylab("Number of samples") + xlab("% detected metabolites")

ggplotly(cc)
```




__Summary statistics microbiome__

1. Focus only at species taxa level


2. Filter out taxa present in less than 20% of all samples


```{r, warning=F,message=FALSE, echo=T}
######################################
##Taxa predicted from MetaPhlAn 3 ####
######################################
#Remove the percentage of unaligned reads
taxa=my_taxa[-1,]
preFilt=data.frame(colnames(taxa))
preFilt$In_MGS="Yes"

#Remove samples that failed in the microbiome prediction: IBDFEC0535, IBDFEC1024,IBDFEC0515
taxa=taxa[, colSums(taxa != 0) > 0]

#Keep track of the samples that failed microbiome prediction 
preFilt[preFilt$colnames.taxa.%ni%colnames(taxa),]$In_MGS="Failed"
rownames(preFilt)=preFilt$colnames.taxa.
preFilt$colnames.taxa.=NULL
tract_id=merge(tract_id,preFilt, by="row.names", all=T)
tract_id=tract_id[complete.cases(tract_id$In_metabolomics),]
row.names(tract_id)=tract_id$Row.names


###################################
### Taxa predicted from mOTUS 2 ###
###################################


#Select species level
#taxa=taxa[unlist(lapply(rownames(taxa),function(x) length(unlist(strsplit(x, "|", fixed=TRUE)))==7)),]
#Remove other taxa levels from names
#rownames(taxa)=sapply(strsplit(rownames(taxa), "|", fixed=TRUE), tail, 1)
#Remove non-alpha-numeric characters from the row.names
#rownames(taxa)=make.names(rownames(taxa))

#################################

######################################
##Taxa predicted from MetaPhlAn2  ####
######################################
#Clean names
#LLD_sp=LLD_taxa[grep("s__", rownames(LLD_taxa)), ]
#IBD_sp=IBD_taxa[grep("s__", rownames(IBD_taxa)), ]
#LLD_sp=LLD_sp[!grepl("t__", rownames(LLD_sp)), ]
#IBD_sp=IBD_sp[!grepl("t__", rownames(IBD_sp)), ]

#row.names(IBD_sp)=gsub(".*s__","",row.names(IBD_sp))
#Merge
#taxa=as.data.frame(t(merge(IBD_sp,LLD_sp, by="row.names")))
#taxa=merge(IBD_sp,LLD_sp, by="row.names")
#row.names(taxa)=taxa$Row.names
#taxa$Row.names=NULL

#Select genus level

taxa_gn=taxa[unlist(lapply(rownames(taxa),function(x) length(unlist(strsplit(x, "|", fixed=TRUE)))==6)),]
#Remove other taxa levels from names
rownames(taxa_gn)=sapply(strsplit(rownames(taxa_gn), "|", fixed=TRUE), tail, 1)

#Remove samples that miss microbiome data: IBDFEC0535, IBDFEC1024,IBDFEC0515
taxa_sp=taxa[grep("s__", rownames(taxa)), ]
taxa_sp=taxa_sp[!grepl("t__", rownames(taxa_sp)), ]
row.names(taxa_sp)=gsub(".*s__","",row.names(taxa_sp))
##########################################


taxa=as.data.frame(t(taxa_sp))

my_shannon=as.data.frame(diversity(taxa, index="shannon"))
colnames(my_shannon)=c("Shannon_Index")
my_simpson=as.data.frame(diversity(taxa, index="simpson"))
colnames(my_simpson)=c("Simpson_Index")
richness_microbiota=merge(my_shannon,my_simpson,by="row.names")
rownames(richness_microbiota)=richness_microbiota$Row.names
richness_microbiota$Row.names=NULL

#Remove samples with low richeness
norich=rownames(richness_microbiota)[richness_microbiota$Shannon_Index==0] 
tract_id[rownames(tract_id)==norich,]$In_MGS="Failed"

richness_microbiota2=subset(richness_microbiota,rownames(richness_microbiota) %in% rownames(summary_sample))
richness_microbiota2=richness_microbiota2[richness_microbiota2$Shannon_Index>0,]
cc_pheno2=merge(richness_microbiota2,cc_pheno,by="row.names")
rownames(cc_pheno2)=cc_pheno2$Row.names
cc_pheno2$Row.names=NULL

print (paste("Number of samples excluded due failed metagenomics:", length(tract_id[tract_id$In_MGS!="yes",])))
print (paste("Number of samples left:", length(rownames(cc_pheno2))))
```

__Microbial richness indexes (Shannon & Simpson)__

```{r, echo=F}
ggplot(cc_pheno2, aes(ibd_Diagnosis,Shannon_Index, fill=ibd_Diagnosis)) + theme_bw() + geom_violin() + geom_violin() + geom_boxplot(width=0.1,position=position_dodge(1), alpha=0.4) + scale_fill_manual(values = c("firebrick2","white","grey42","blue2"))


ggplot(cc_pheno2, aes(ibd_Diagnosis,Simpson_Index, fill=ibd_Diagnosis)) + theme_bw() + geom_violin() + geom_violin() + geom_boxplot(width=0.1,position=position_dodge(1), alpha=0.4) + scale_fill_manual(values = c("firebrick2","white","grey42","blue2"))
```


__Filter bacteria (detection in >20% of the samples) and normalization__


```{r}

#Subset samples from the taxonomic table for which we also have metabolomics 
taxa2=subset(taxa,rownames(taxa) %in% rownames(cc_pheno2))
taxa_gn2=taxa_gn[,colnames(taxa_gn) %in% rownames(cc_pheno2)]

print (paste("Number of taxa detected:",ncol(taxa2)))

#Filter 20% and to clr transformation
taxa_filt=transform_and_filter_taxa(taxa2,samples_row = T,method = "clr",missing_filter = 20)
taxa_gn_filt=transform_and_filter_taxa(t(taxa_gn2),samples_row = T,method = "clr",missing_filter = 20)

print (paste("Number of taxa after filtering",ncol(taxa_filt)))

#Merge taxa and phenotypes
cc_pheno3=merge(cc_pheno2,taxa_filt,by="row.names")
rownames(cc_pheno3)=cc_pheno3$Row.names
cc_pheno3$Row.names=NULL

```



3. Filter metabolites by missigness
===

__We remove metabolites that are present in less than 20% of the samples in the 3 cohorts (CD,UC,Controls)__


```{r,echo=F}

to_remove=subset(summary_short2,prevalence_Control<20 & prevalence_CD<20 & prevalence_UC<20)
q_mtb_v2=all_new_ID_raw[,colnames(all_new_ID_raw) %ni%to_remove$Metabolite]
print(paste("Number of metabolites removed:",length(to_remove$Metabolite)))
```

__Metabolites present in more than 70% of the samples will be analyzed quantitatively__

```{r,echo=F}

#to_keep=subset(summary_short2,prevalence_Control>70 & prevalence_CD>70 & prevalence_UC>70)
to_keep=subset(summary_short2,prevalence_all>70)
q_mtb=all_new_ID_raw[,colnames(all_new_ID_raw) %in%to_keep$Metabolite]
print(paste("Number of metabolites for quantitative analysis:",length(to_keep$Metabolite)))
```


__Metabolites that are present between 20% to 70% of the samples will be subset and transform to binary feature (present/absent)__

```{r,echo=F}
rest=c(list(to_remove$Metabolite),list(to_keep$Metabolite))
#mtb_logistic=subset(summary_short2,prevalence_Control>20 & prevalence_Control<70 & prevalence_UC>20 & prevalence_UC<70 & prevalence_CD>20 & prevalence_CD<70 )

a=droplevels(unlist(to_remove$Metabolite))
b=droplevels(unlist(to_keep$Metabolite))
c=c(levels(a),levels(b))

bi_mtb=all_new_ID_raw[,colnames(all_new_ID_raw) %ni% c]
bi_mtb[!is.na(bi_mtb)]=1
bi_mtb[is.na(bi_mtb)]=0

print(paste("Number of metabolites for presence/absence analysis:",length(colnames(bi_mtb))))
```




Normalize and impute missing metabolites
---

```{r, echo=T, message=F, warning=F}

#Normalize by experiment day: divide each metabolite value by the median of its 
my_flag=1
for (z in unique(my_batch_id$run_day_cat)){
  my_names=rownames(my_batch_id)[my_batch_id$run_day_cat==z]
  d1=subset(q_mtb, row.names(q_mtb) %in% my_names)
  d1.1=apply(d1, 2, function (x) {x/median(x,na.rm = T)})

  if (my_flag==1){
    q_mtb_norm=d1.1
    my_flag=5

  } else{
    q_mtb_norm=rbind(q_mtb_norm,d1.1)

  }
}


my_flag=1
for (z in unique(my_batch_id$run_day_cat)){
  my_names=rownames(my_batch_id)[my_batch_id$run_day_cat==z]
  d1=subset(q_mtb_v2, row.names(q_mtb_v2) %in% my_names)
  d1.1=apply(d1, 2, function (x) {x/median(x,na.rm = T)})

  if (my_flag==1){
    q_mtb_norm_v2=d1.1
    my_flag=5

  } else{
    q_mtb_norm_v2=rbind(q_mtb_norm_v2,d1.1)

  }
}


q_mtb_t=as.data.frame(q_mtb_norm)

q_mtb_t_v2=as.data.frame(q_mtb_norm_v2)

#apply log10 transformation + scale
q_mtb_t=log10(q_mtb_t)
q_mtb_t_v2=log10(q_mtb_t_v2)
#scale date, mean =0 , sd =1
#q_mtb_t <- as.data.frame(scale(q_mtb_t))

#CLR transformation of metabolites

#q_mtb_t= t(apply(q_mtb_t, 1, function(b) {
#  my_mean=exp(sum(log(b), na.rm = T)/length(b))
#  log(b/my_mean)
#}))

#Impute na's 
### WARNING: KNN imputation takes a while, take a nap or a coffee (in case of re-run save and load results of this stage)
q_mtb=impute_missing_values(q_mtb_t,samples_row = T, method = "knn", missing_filter = 0)
rownames(q_mtb)=row.names(q_mtb_t)
write.table(q_mtb, "~/Desktop/Metabolomics_for_cluster/2.Input/knn_Imputed_metabolites_median_log10.txt", sep = "\t",quote = F)

q_mtb=as.data.frame(q_mtb)
```

__Merge taxa and metabolites__

```{r, echo=F}

print("Missing metagenomic data:")
setdiff( rownames(q_mtb), rownames(taxa_filt))
taxa_and_mb=merge(taxa_filt,q_mtb,by="row.names")
rownames(taxa_and_mb)=taxa_and_mb$Row.names
taxa_and_mb$Row.names=NULL
```


__Summary statistics on transformed & filtered metabolites__

```{r, echo=F}
#Transformed data
summary_met_trans=summary_stats(q_mtb, select, missing_vals = "NA")
summary_met_trans=as.data.frame(summary_met_trans)
write.table(summary_met_trans, "~/Documents/Project_Metabolome/4.Summary_stats/Metabolites_summary_post_filtering_2020_v3", sep = "\t", quote = F)
```


4.Summary of SCFA
===

__Show SCFA names__

```{r, warning=F,message=FALSE, echo=TRUE}
unique(levels(SCFA$Analyte))
```

__Load data and replace values under the detection levels for missing values and restructure the data__

Original report contain 1 row per sample per measurment, we want to restructure in a way that all samples are in rows and each column is a SCFA.  


```{r, warning=F,message=FALSE, echo=TRUE}
#Duplicate original values (before remplacing for NA)

SCFA$ori_Result=SCFA$Result

#Replace values below level quantification (BLOQ) to NA
SCFA$Result[SCFA$Comment.1=="BLOQ"] = "N/Q"
SCFA$Result[SCFA$Result=="N/Q"] = NA
```

__Summary of the values under the detection levels (TRUE column)__

```{r}
table(SCFA$Analyte,is.na(SCFA$Result))
```


```{r, warning=F,message=FALSE, echo=TRUE}
#Remove unecessary columns (sample type, group/project name, detection limits and comments)

SCFAsub=SCFA[,c("Unique.Sample.ID...as.labeled.on.tube.","Sample.Amount..gram.","Comment","Analyte","Result")]

#Reshape the SCFA table
SCFA_table <- dcast(SCFAsub, ...~Analyte)

#Chance sample names to match metadata
IDs=data.frame(rbind(as.matrix(IDs_IBD), as.matrix(IDs_LLD)))
rownames(SCFA_table)=SCFA_table$Unique.Sample.ID...as.labeled.on.tube.
SCFA_table$Unique.Sample.ID...as.labeled.on.tube.=NULL
rownames(IDs)=IDs$V1
IDs$V1=NULL
SCFA_new_ID=merge(IDs,SCFA_table, by="row.names")
SCFA_new_ID$Row.names=NULL
row.names(SCFA_new_ID)=SCFA_new_ID$V5
SCFA_new_ID$V5=NULL
colnames(SCFA_new_ID)=c("Amount_sample_gram", "Box", "Methylbutyric_acid", "acetic_acid", "butyric_acid", "hexanoic_acid", "isobutyric_acid","isovaleric_acid" ,"propionic_acid", "valeric_acid")
```

__Merge phenotypes to SCFA__

```{r, warning=F,message=FALSE, echo=TRUE}
SCFA2=SCFA_new_ID
SCFA2[,3:10]=sapply(SCFA2[,c(3:10)], function(x) log10(as.numeric(as.character(x))) )
SCFA_with_phenos=merge(SCFA_new_ID,cc_pheno3, by="row.names")
cc_pheno4=merge(SCFA2,cc_pheno3, by="row.names")

```


__Plot SCFA per diagnosis__

```{r, warning=F,message=FALSE, echo=T}
plot_SCFA=SCFA_with_phenos[,c("Row.names","ibd_Diagnosis","Methylbutyric_acid", "acetic_acid", "butyric_acid", "hexanoic_acid", "isobutyric_acid","isovaleric_acid" ,"propionic_acid", "valeric_acid")]

plot_SCFA_m=melt(plot_SCFA, id.vars = c("Row.names","ibd_Diagnosis"))
plot_SCFA_m=subset(plot_SCFA_m, plot_SCFA_m$ibd_Diagnosis!="IBDU")
plot_SCFA_m$diag=as.character(plot_SCFA_m$ibd_Diagnosis)
plot_SCFA_m$diag[plot_SCFA_m$ibd_Diagnosis=="Control"]="A_Control"
my_comparisions=list( c("CD","A_Control"), c("A_Control","UC"), c("CD","UC"))
```

```{r, warning=F,message=FALSE, echo=F, fig.height=15}
ggplot (plot_SCFA_m, aes (diag,log10(as.numeric(as.character(value))), fill=diag)) + geom_violin() + geom_boxplot(width=0.1,position=position_dodge(1), alpha=0.2) + facet_wrap(~ variable, nrow=2) + theme_bw() +scale_fill_manual(values=c("snow","firebrick2","blue2"))  + stat_compare_means(comparisons = my_comparisions, method = "wilcox.test") + ylab ("log10(SCFA µg/g)") + xlab ("Cohort") + scale_x_discrete(labels=c("(n=255)","(n=239)", "(n=176)"))
```

```{r, warning=F,message=FALSE, echo=T}

montreal= read.delim("~/Documents/Project_Metabolome/1.Input_files/only_montreal.txt", row.names=1)
rownames(plot_SCFA)=plot_SCFA$Row.names
plot_SCFA$Row.names=NULL
tkd_SCFA=merge(montreal,plot_SCFA,by="row.names")
tkd_SCFA_m=melt(tkd_SCFA, id.vars = c("Row.names","ibd_montreal_B","ibd_montreal_Bperianal","ibd_montreal_S","ibd_montreal_A","ibd_montreal_Bperianal.1" ,"ibd_Montreal_L","ibd_MontrealE","ibd_Diagnosis"))
tkd_SCFA_m=subset(tkd_SCFA_m, tkd_SCFA_m$ibd_Diagnosis!="IBDU")

tkd_SCFA_b=subset(tkd_SCFA_m, tkd_SCFA_m$ibd_Diagnosis!="UC")
tkd_SCFA_b=tkd_SCFA_b[complete.cases(tkd_SCFA_b$ibd_montreal_B), ]


my_comparisions=list(  c("B1","B2"), c("B2","B3"), c("Control","B3"),c("B1","B3"),c("Control","B2"))

ggplot (tkd_SCFA_b, aes (ibd_montreal_B,log10(as.numeric(as.character(value))), fill=ibd_montreal_B)) + geom_violin(na.rm = T) + geom_boxplot(width=0.1,position=position_dodge(1), alpha=0.2,na.rm = T) + facet_wrap(~ variable, nrow=2) + theme_bw() + stat_compare_means(comparisons = my_comparisions, method = "wilcox.test") + ylab ("log10(SCFA µg/g)") + xlab ("Cohort") + scale_x_discrete(labels=c("(n=129)","(n=76)", "(n=29)", "(n=255)")) +  scale_fill_manual(values=c("lightsalmon","tomato","red4","snow"))


tkd_SCFA_s=subset(tkd_SCFA_m, tkd_SCFA_m$ibd_Diagnosis!="CD")
tkd_SCFA_s=tkd_SCFA_s[complete.cases(tkd_SCFA_s$ibd_montreal_S), ]

my_comparisions=list(  c("Control","S0"), c("S0","S1"), c("S1","S2"), c("S2","S3"), c("Control","S1"),c("Control","S2"),c("Control","S3"))

ggplot (tkd_SCFA_s, aes (ibd_montreal_S,log10(as.numeric(as.character(value))), fill=ibd_montreal_S)) + geom_violin(na.rm = T) + geom_boxplot(width=0.1,position=position_dodge(1), alpha=0.2,na.rm = T) + facet_wrap(~ variable, nrow=2) + theme_bw() + stat_compare_means(comparisons = my_comparisions, method = "wilcox.test") + ylab ("log10(SCFA µg/g)") + xlab ("Cohort")  + scale_fill_manual(values=c("snow","lightcyan2","lightblue2","steelblue1","blue3"))


plot_SCFA_cal=SCFA_with_phenos[,c("Row.names","ibd_Diagnosis","ibd_FecalCalprotectinOver200yesno","Methylbutyric_acid", "acetic_acid", "butyric_acid", "hexanoic_acid", "isobutyric_acid","isovaleric_acid" ,"propionic_acid", "valeric_acid")]
plot_SCFA_cal=melt(plot_SCFA_cal, id.vars = c("Row.names","ibd_Diagnosis", "ibd_FecalCalprotectinOver200yesno"))
plot_SCFA_cal=subset(plot_SCFA_cal, plot_SCFA_cal$ibd_Diagnosis!="IBDU")
plot_SCFA_cal=plot_SCFA_cal[complete.cases(plot_SCFA_cal$ibd_FecalCalprotectinOver200yesno), ]
my_comparisions=list( c("no","yes"))

ggplot (plot_SCFA_cal, aes (ibd_FecalCalprotectinOver200yesno,log10(as.numeric(as.character(value))), fill=ibd_FecalCalprotectinOver200yesno)) + geom_violin() + geom_boxplot(width=0.1,position=position_dodge(1), alpha=0.2) + facet_wrap(ibd_Diagnosis~ variable , nrow=3) + theme_bw() +scale_fill_brewer(palette ="RdBu")  + stat_compare_means(comparisons = my_comparisions, method = "wilcox.test") + ylab ("log10(SCFA µg/g)")  

```



5.PCA analysis on metabolites present in more than 70% of the samples
===

```{r, warning=F,message=FALSE, echo=TRUE}
#filt_1=subset(summary_short2,prevalence_Control>20 | prevalence_UC>20 | prevalence_CD>20  )

#norm_filt=all_new_ID[,colnames(all_new_ID) %in%filt_1$Metabolite]
#norm_filt=q_mtb[,colnames(q_mtb) %in%filt_1$Metabolite]
#norm_filt=as.data.frame(scale(q_mtb))
norm_filt=q_mtb
pca_log= prcomp(norm_filt,scale. = T, center = T)
cmp_log=as.data.frame(pca_log$x[,1:5])
#cc_pheno2=read.delim("~/Documents/Project_Metabolome/1.Input_files/mini_pheno2.txt", row.names=1)
cmp_log2=merge(cc_pheno2,cmp_log,by="row.names")
v_exp = pca_log$sdev^2
prop_v_exp= data.frame( PC=1:length(v_exp),perc= 100*(v_exp / sum(v_exp)))

```



Percentage of variance explained per component
---

```{r}
prop_v_exp2=prop_v_exp[1:20,]
ggplot(prop_v_exp2, aes (PC,perc)) + geom_bar (stat="identity") + geom_text(aes(label=round(perc, 2)), size=3, vjust=-.5) + theme_bw() + xlab ("Principal components") + ylab ("% explained variance")
```


PC1 & PC2: Diagnosis
---

```{r}
ggplot(cmp_log2, aes(PC1,PC2, color=ibd_Diagnosis)) + geom_point(size=2) + theme_bw() + scale_colour_manual(values =c("firebrick3","black","gold3" ,"steelblue2"))
```

PC2 & PC3: Diagnosis
---

```{r}
ggplot(cmp_log2, aes(PC2,PC3, color=ibd_Diagnosis)) + geom_point() + theme_bw() + scale_colour_manual(values =c("firebrick3","black","gold3" ,"steelblue2"))
```


Exploring different phenotypes
---

__PC1 & PC2: Bowel Movements a day__


```{r}
ggplot(cmp_log2, aes(PC1,PC2, color=clinical_BowelMovementADayDef)) + geom_point() + theme_bw()
```


__PC1 & PC2: Bristol stool scale binary (soft vs hard)__


```{r}
ggplot(cmp_log2, aes(PC1,PC2, color=clinical_BinaryBSS)) + geom_point() + theme_bw()
```



__PC1 & PC2: Number of months that the sample spend in the -80 freezer__


```{r}
ggplot(cmp_log2, aes(PC1,PC2, color=metabolon_Month_in_freezer)) + geom_point() + theme_bw()
```



__PC1 & PC2:Activity__


```{r}
ggplot(cmp_log2, aes(PC1,PC2, color=ibd_ActiveDisease)) + geom_point() + theme_bw() + scale_colour_manual(values=c("red2","gray32","blue2","snow"))
```




__PC1 & PC2: Food frequency: Sum of Carhydrates__


```{r}
ggplot(cmp_log2, aes(PC1,PC2, color=diet_SUMOFKHTOT.Res)) + geom_point() + theme_bw() + scale_colour_viridis()
```


__PC1 & PC2: Food frequency: Sum of fat__


```{r}
ggplot(cmp_log2, aes(PC1,PC2, color=diet_SUMOFVETTOT.Res)) + geom_point() + theme_bw() + scale_colour_viridis()
```


__PC1 & PC2: Food frequency: plant protein vs animal proten__


```{r}
ggplot(cmp_log2, aes(PC1,PC2, color=diet_PlanttoAnimal)) + geom_point() + theme_bw() + scale_colour_viridis()
```



__PC1 & PC2: Montreal classifications__



```{r, warning=F, echo=F}
rownames(cmp_log2)=cmp_log2$Row.names
cmp_log2$Row.names=NULL
montreal=read.table("~/Documents/Project_Metabolome/1.Input_files/only_montreal.txt", sep = "\t", header = T, row.names = 1)
cmp_log3=merge(montreal,cmp_log2, by="row.names")
cmp_log4=cmp_log3[,c("Row.names","PC1","PC2","ibd_montreal_B","ibd_montreal_Bperianal","ibd_montreal_S","ibd_montreal_A","ibd_Montreal_L","ibd_MontrealE")]
cmp_log4=melt(cmp_log4, id.vars=c("Row.names","PC1","PC2"))

ggplot(cmp_log4, aes(PC1,PC2, color=value)) + geom_point() + theme_bw() + facet_wrap(.~ variable , ncol=2) + scale_color_manual(values=c("lightsalmon","tomato","red4","lightsalmon","tomato","red4","gray88","lightcyan2","lightblue2","steelblue1","lightcyan2","lightblue2","steelblue1","blue3", "black", "lightsalmon","tomato","red", "red4", "red1","snow" ))

table(cmp_log3$ibd_Diagnosis,cmp_log3$ibd_montreal_Bperianal)
table(cmp_log3$ibd_Diagnosis,cmp_log3$ibd_montreal_B)
table(cmp_log3$ibd_Diagnosis,cmp_log3$ibd_montreal_S)

```


5.PCA & PCoA analysis on bacteria present in more than 20% of the samples
===

__Calculate PCA based on filtered clr relative abundaces__

```{r, echo=F}
print(paste("Number of species used for PCA analysis:", ncol(taxa_filt)))
```

```{r, echo=T}
pca_clr= prcomp(taxa_filt,scale. = T, center = T)
cmp_clr=as.data.frame(pca_clr$x[,1:5])
cmp_clr2=merge(cc_pheno2,cmp_clr,by="row.names")

```

```{r, echo=F}

ggplot(cmp_clr2, aes(PC1,PC2, color=ibd_Diagnosis)) + geom_point() + theme_bw() + scale_colour_manual(values =c("firebrick3","black","gold3" ,"steelblue2"))

```


```{r, echo=F}

ggplot(cmp_clr2, aes(PC1,PC2, color=clinical_BowelMovementADayDef)) + geom_point() + theme_bw()

```


__Repeat for disease scores Monteral__

```{r, message=F, warning=F}

rownames(cmp_clr2)=cmp_clr2$Row.names
cmp_clr2$Row.names=NULL
cmp_clr3=merge(montreal,cmp_clr2, by="row.names")

cmp_clr4=cmp_clr3[,c("Row.names","PC1","PC2","ibd_montreal_B","ibd_montreal_Bperianal","ibd_montreal_S","ibd_montreal_A","ibd_Montreal_L","ibd_MontrealE")]
cmp_clr4=melt(cmp_clr4, id.vars=c("Row.names","PC1","PC2"))

ggplot(cmp_clr4, aes(PC1,PC2, color=value)) + geom_point() + theme_bw() + facet_wrap(.~ variable , ncol=2) + scale_color_manual(values=c("lightsalmon","tomato","red4","lightsalmon","tomato","red4","gray88","lightcyan2","lightblue2","steelblue1","lightcyan2","lightblue2","steelblue1","blue3", "black", "lightsalmon","tomato","red", "red4", "red1","snow" ))

```

__Calculate PCoA based on Bray-Curtis distances__


```{r, warning=F,message=FALSE, echo=TRUE}
#Filter present in at least 20% of the samples
taxa2=taxa2/100
taxa2[taxa2=="NA"]=0
taxa2[taxa2=="NaN"]=0
taxa_filt=taxa2[,((colSums(taxa2 !=0) / nrow(taxa2)) *100 )>20]
#Data transformation
#Choose square root arcsine transformation
taxa_asin=asin(sqrt(taxa_filt))

#Calculate bray curtis dissimilarity
bc_dis=vegdist(taxa_asin,method = "bray")
tax_mds <- metaMDS(comm=bc_dis,k=5, autotransform = FALSE)
data.scores <- as.data.frame(scores(tax_mds))
data.scores$site <- rownames(tax_mds) 
bc_mds=merge(cc_pheno2,data.scores,by="row.names")

```


```{r, echo=F}

ggplot(bc_mds, aes(NMDS1,NMDS2, color=ibd_Diagnosis)) + geom_point() + theme_bw() + scale_colour_manual(values =c("firebrick3","black","gold3" ,"steelblue2"))

```


6. Clustering metabolites
===


Method 1: Regress disease and (potential) technical confouders and perform spearman correlation
---

__Regress batch & diseases effects__

Here we use all metabolites present in more than 20% of the samples. 

We extract residuals from the following linear model (lm) : 

__lm(Metabolite ~ Day_of_measurment  + Amount_of_input_material + LC Column + Months_sample_in_freezer + ibd_Diagnosis (Control/CD/UC) )__

```{r}
rownames(cc_pheno4)=cc_pheno4$Row.names
cc_pheno4$Row.names=NULL
cc_pheno5=merge(my_batch_id,cc_pheno4, by="row.names")
#my_regress_phenos=cc_pheno5[,c("Row.names","run_day_cat", "Amount_sample_gram", "LC.COLUMN" , "metabolon_Month_in_freezer" , "ibd_Diagnosis")]
#Since we normalized each metabolite by its experiment day median we don't need to regress this effect anymore  
my_regress_phenos=cc_pheno5[,c("Row.names", "Amount_sample_gram", "LC.COLUMN" , "metabolon_Month_in_freezer" , "ibd_Diagnosis")]
rownames(my_regress_phenos)=my_regress_phenos$Row.names
my_regress_phenos$Row.names=NULL
rgs_mb=merge(my_regress_phenos,norm_filt, by="row.names")

rgs_mb2=matrix(nrow=nrow(my_regress_phenos), ncol = ncol(norm_filt))
 n=1
for (i in 6:ncol(rgs_mb)){
 #my_lm=lm(rgs_mb[,i] ~ run_day_cat+Amount_sample_gram+LC.COLUMN+metabolon_Month_in_freezer+ibd_Diagnosis,data = rgs_mb)
my_lm=lm(rgs_mb[,i] ~ Amount_sample_gram+LC.COLUMN+metabolon_Month_in_freezer+ibd_Diagnosis,data = rgs_mb)
 rgs_mb2[,n] = my_lm$residuals
 n=n+1
}
rgs_mb2=as.data.frame(rgs_mb2)
rownames(rgs_mb2)=rgs_mb$Row.names
colnames(rgs_mb2)=colnames(rgs_mb)[6:ncol(rgs_mb)]
```

```{r}
#Select metabolite annotation
library ("Hmisc")
annot_v2=annot
annot_v2$SUB.PATHWAY=NULL
annot_v2=subset(annot_v2, rownames(annot_v2) %in% colnames(rgs_mb2))
#cor_metabolites=rcorr(as.matrix(q_mtb), type="spearman")
#cor_metabolites_coef=cor_metabolites$r
#cor_metabolites_pval=cor_metabolites$P
#cor_coef=melt(cor_metabolites_coef)
#cor_pval=melt(cor_metabolites_pval)
#cor_coef$pvalue=cor_pval$value
#write.table(cor_coef, "~/Documents/Project_Metabolome/6.Results/072020/Correlation_corrected_metabolites.txt", sep = "\t", quote = F)

#cor_mtbs=cor_coef

my_IBD=row.names(cc_pheno)[cc_pheno$ibd_Diagnosis!="Control"]
my_CNT=row.names(cc_pheno)[cc_pheno$ibd_Diagnosis=="Control"]

IBDqmtb=subset(q_mtb, rownames(q_mtb) %in% my_IBD)
cor_metabolites=rcorr(as.matrix(IBDqmtb), type="spearman")
cor_metabolites_coef=cor_metabolites$r
cor_metabolites_pval=cor_metabolites$P
cor_coef=melt(cor_metabolites_coef)
cor_pval=melt(cor_metabolites_pval)
cor_coef$pvalue=cor_pval$value
write.table(cor_coef, "~/Desktop/Metabolomics_for_cluster/3.Preliminary_results_v2/Correlation_metabolites_IBD.txt", sep = "\t", quote = F)

CNTqmtb=subset(q_mtb, rownames(q_mtb) %in% my_CNT)
cor_metabolites=rcorr(as.matrix(CNTqmtb), type="spearman")
cor_metabolites_coef=cor_metabolites$r
cor_metabolites_pval=cor_metabolites$P
cor_coef=melt(cor_metabolites_coef)
cor_pval=melt(cor_metabolites_pval)
cor_coef$pvalue=cor_pval$value
write.table(cor_coef, "~/Desktop/Metabolomics_for_cluster/3.Preliminary_results_v2/Correlation_metabolites_CNT.txt", sep = "\t", quote = F)


#TO IMPLEMENT: Network analysis
mynet=graph.adjacency(as.matrix(as.dist(cor(CNTqmtb, method = "spearman"))), mode = "undirected", weighted = T, diag = F)
mynet=simplify(mynet, remove.multiple = T, remove.loops = T)
# Colour negative correlation edges as blue
E(mynet)[which(E(mynet)$weight<0)]$color <- "darkblue"

# Colour positive correlation edges as red
E(mynet)[which(E(mynet)$weight>0)]$color <- "darkred"

# Convert edge weights to absolute values
E(mynet)$weight <- abs(E(mynet)$weight)

mynet <- delete_edges(mynet, E(mynet)[which(E(mynet)$weight<0.6)])

mynet <- delete.vertices(mynet, degree(mynet)==0)

```


Method 2:iClust
----

```{R}
library (iClusterPlus)

#Features to use: q_mtb (quantitative metabolies), bacterial species clr-transformed ()
clus_taxa=cc_pheno3[,206:314]
clus_pheno=cc_pheno3[,c("ibd_Diagnosis","ibd_DiseaseLocation")]

clus_pheno$IBD=0
#clus_pheno$IBD[clus_pheno$ibd_Diagnosis=="Control",]=0
clus_pheno$IBD[clus_pheno$ibd_Diagnosis!="Control"]=1

#clus_pheno$Location=0
#clus_pheno$Location[clus_pheno$ibd_DiseaseLocation=="colon"]=1
#clus_pheno$Location[clus_pheno$ibd_DiseaseLocation=="ileum"]=2
#clus_pheno$Location[clus_pheno$ibd_DiseaseLocation=="both"]=2

clus_pheno$ibd_Diagnosis=NULL
clus_pheno$ibd_DiseaseLocation=NULL

#clus_pheno$Location=as.numeric(unlist(clus_pheno$Location))
clus_pheno$IBD=as.numeric(unlist(clus_pheno$IBD))

clus_mtb=q_mtb[row.names(q_mtb) %in% row.names(clus_pheno),]
clus_mtb=clus_mtb[order(row.names(clus_mtb)),]
table(row.names(clus_mtb)==row.names(clus_pheno))

write.table(clus_mtb, "~/Desktop/Metabolomics_for_cluster/2.Input/clst_mtb3", sep="\t", quote = F)
write.table(clus_pheno, "~/Desktop/Metabolomics_for_cluster/2.Input/clst_pheno3", sep="\t", quote = F)
write.table(clus_taxa, "~/Desktop/Metabolomics_for_cluster/2.Input/clst_taxa3", sep="\t", quote = F)


#Run in Boxy (hpc cluster)

#ml Rplust
#library (iClusterPlus)

clus_mtb=read.table("./clst_mtb3", sep = "\t", header = T, row.names = 1)
clus_taxa=read.table("./clst_taxa3", sep = "\t", header = T, row.names = 1)
clus_pheno=read.table("./clst_pheno3", sep = "\t", header = T, row.names = 1)

#test 1 to 5 clusters

for(k in 1:10){
  cv.fit = tune.iClusterPlus(cpus=8,dt1 = data.matrix(clus_mtb),dt2 = data.matrix(clus_taxa),dt3 = data.matrix(clus_pheno),type=c("gaussian","gaussian", "binomial"),K=k,n.lambda=NULL,scale.lambda=c(1,1,1),maxiter=20)
  save(cv.fit, file=paste("cv.fit.k",k,".Rdata",sep=""))
  }



#Check BIC for each K (number of clusters)


setwd("/Users/arnauvich/Desktop/Metabolomics_for_cluster/2.Input/iClust_out/")
output=alist() 
files=grep("cv.fit",dir())
for(i in 1:length(files)){
  load(dir()[files[i]])
  output[[i]]=cv.fit
} 

#From the manual
#To  select  the  best  k,  we  compute  the  deviance  ratio  which  is  the  ratio  of  the log-likelihood(fitted) - log-likelihood(null model) divided by the log-likelihood (full model)- log-likelihood(null model).  The deviance ratio can be interpreted as the percentEV. We choose the k where the curve of percentEV levels off. 
nLambda = nrow(output[[1]]$lambda)
nK = length(output)
BIC = getBIC(output)
devR = getDevR(output)

minBICid = apply(BIC,2,which.min)
devRatMinBIC = rep(NA,nK)
for(i in 1:nK){
  devRatMinBIC[i] = devR[minBICid[i],i]
}

clusters=getClusters(output)
rownames(clusters)=rownames(clus_mtb)
colnames(clusters)=paste("K=",2:(length(output)+1),sep="")
write.table(clusters, file="/Users/arnauvich/Desktop/Metabolomics_for_cluster/3.Preliminary_results_v2/clusterMembership_iCluster.txt",sep='\t',quote=F)

#Inspect visualy the number of clusters to use

plot(1:(nK+1),c(0,devRatMinBIC),type="b",xlab="Number of clusters (K+1)",ylab="%Explained Variation")

# Warning k=k+1
#plot shows that 4 is the best cluster solution (k=3)
#Show number of samples per cluster: 
print (table(clusters[,3]))
k=3 
best.cluster=clusters[,k]
best.fit=output[[k]]$fit[[which.min(BIC[,k])]]

#Check top-features

features = alist()
features[[1]] = colnames(clus_mtb)
features[[2]] = colnames(clus_taxa)
features[[3]] = colnames(clus_pheno)
sigfeatures=alist()
for(i in 1:3){   
  rowsum=apply(abs(best.fit$beta[[i]]),1, sum)
  upper=quantile(rowsum,prob=0.75)
  sigfeatures[[i]]=(features[[i]])[which(rowsum>upper)]
}
names(sigfeatures)=c("metabolite","taxa","IBD_pheno")
head(sigfeatures[[1]])
head(sigfeatures[[2]])
head(sigfeatures[[3]])

#capture.output(sigfeatures, file= "/Users/arnauvich/Desktop/Metabolomics_for_cluster/3.Preliminary_results_v2/features_icluster.txt")


#Plot heatmap 
library (gplots)
bw.col = colorpanel(2,low="white",high="black")
col.scheme = alist()
col.scheme[[1]] = bluered(800)
col.scheme[[2]] = bluered(800)
col.scheme[[3]] = bw.col 
plotHeatmap(fit=best.fit,datasets=list(clus_mtb,clus_taxa,clus_pheno),
            type=c("gaussian","gaussian", "binomial"), col.scheme = col.scheme)

plotHeatmap(fit=best.fit,datasets=list(clus_mtb,clus_taxa),
            type=c("gaussian","gaussian"), col.scheme = col.scheme)


#View solution in bacteria PCA
new_pc_bac=merge(clusters,cmp_clr2, by="row.names")
ggplot (new_pc_bac, aes(PC1,PC2, color=as.factor(new_pc_bac$`K=4`))) + geom_point() + theme_bw() + scale_color_manual(values=c("black","firebrick2", "lightblue2", "gold2","purple"))

#View solution in metabolite PCA
#rownames(cmp_log2)=(cmp_log2$Row.names)
#cmp_log2$Row.names=NULL
new_pc_mtb=merge(clusters,cmp_log2, by="row.names")

ggplot (new_pc_mtb, aes(PC1,PC2, color=as.factor(new_pc_mtb$`K=4`))) + geom_point() + theme_bw() + scale_color_manual(values=c("black","firebrick2", "lightblue2", "gold2","purple"))
#bla=tune.iClusterPlus(cpus = 2,dt1 = data.matrix(clus_mtb),dt2 = data.matrix(clus_taxa),dt3 = data.matrix(clus_pheno), type = c("gaussian","gaussian","multinomial"))
plot_ly(x=new_pc_mtb$PC1,y=new_pc_mtb$PC2,z=new_pc_mtb$PC3, size=10 ,color = as.factor(new_pc_mtb$`K=4`), colors = c("black","firebrick2", "lightblue2", "gold2"))
plot_ly(x=new_pc_mtb$PC2,y=new_pc_mtb$PC3,z=new_pc_mtb$PC4, size=10 ,color = as.factor(new_pc_mtb$`K=4`), colors = c("black","firebrick2", "lightblue2", "gold2"))


####################
#Play with heatmaps#
####################


ph_heat=new_pc_mtb[,c("Row.names","ibd_Diagnosis","ibd_DiseaseLocation","K=4")]
rownames(ph_heat)=ph_heat$Row.names
ph_heat$Row.names=NULL
htmtb=merge(ph_heat,clus_mtb, by="row.names")
htbac=merge(ph_heat,clus_taxa,  by="row.names")

htmtb=melt(htmtb,id.vars=c("Row.names","ibd_Diagnosis","ibd_DiseaseLocation","K=4"))
htbac=melt(htbac,id.vars=c("Row.names","ibd_Diagnosis","ibd_DiseaseLocation","K=4"))

colnames(htbac)=c("ID","IBD","Location","Clusters","Metabolite","Abundance")
htbac$type="Microbiome"

colnames(htmtb)=c("ID","IBD","Location","Clusters","Metabolite","Abundance")
htmtb$type="Metabolome"
ht_test=rbind(htbac,htmtb)

ggplot(ht_test,aes(reorder(as.factor(ID),Clusters),Metabolite,fill=as.numeric(Abundance))) + geom_tile(stat = "identity") + scale_fill_viridis(discrete=FALSE) + theme(axis.title.x=element_blank(), axis.text.x=element_blank(), axis.ticks.x=element_blank()) + facet_grid(type~Clusters, scales = "free", space="free_x")

ggplot(ht_test,aes(reorder(as.factor(ID),Clusters),Metabolite,fill=as.numeric(Abundance))) + geom_tile(stat = "identity") + scale_fill_gradient2(low="navy", mid="white", high="red", midpoint=0) + theme(axis.title.x=element_blank(), axis.text.x=element_blank(), axis.ticks.x=element_blank(), axis.text.y = element_text(size=3)) + facet_grid(type~Clusters, scales = "free", space="free_x")


htmtb_2=htmtb[ htmtb$Metabolite %in% sigfeatures[[1]], ]
ht_test2=rbind(htbac,htmtb_2)

ggplot(ht_test2,aes(reorder(as.factor(ID),Clusters),Metabolite,fill=as.numeric(Abundance))) + geom_tile(stat = "identity") + scale_fill_gradient2(low="navy", mid="white", high="red", midpoint=0) + theme(axis.title.x=element_blank(), axis.text.x=element_blank(), axis.ticks.x=element_blank(), axis.text.y = element_text(size=3)) + facet_grid(type~Clusters, scales = "free", space="free_x")

ggplot(ht_test2,aes(reorder(as.factor(ID),Clusters),Metabolite,fill=as.numeric(Abundance))) + geom_tile(stat = "identity") + scale_fill_viridis(discrete=FALSE) + theme(axis.title.x=element_blank(), axis.text.x=element_blank(), axis.ticks.x=element_blank()) + facet_grid(type~Clusters, scales = "free", space="free_x")




s1=new_pc_mtb[,c("Row.names","K=4")]
rownames(s1)=s1$Row.names
s1$Row.names=NULL
s1=merge(s1,clus_mtb, by="row.names")
s1=melt(s1, id.vars=c("Row.names","K=4"))
s2=new_pc_mtb[,c("Row.names","K=4","ibd_Diagnosis","ibd_DiseaseLocation", "ibd_ActiveDisease","clinical_BowelMovementADayDef", "Shannon_Index")]
s2=melt(s2, id.var=c("Row.names","K=4"))

s1$type="Metabolite"
s2$type="Phenotype"
colnames(s1)=c("ID","Clusters","Metabolite","Abundance","type")
colnames(s2)=c("ID","Clusters","Metabolite","Abundance","type")
s1=s1[ s1$Metabolite %in% sigfeatures[[1]], ]
s3=rbind(s1,s2)

ggplot(s3,aes(reorder(as.factor(ID),Clusters),Metabolite,fill=as.numeric(Abundance))) + geom_tile(stat = "identity") + scale_fill_viridis(discrete=FALSE) + theme(axis.title.x=element_blank(), axis.text.x=element_blank(), axis.ticks.x=element_blank()) + facet_grid(type~Clusters, scales = "free", space="free")


new_pc_mtb2=new_pc_mtb[,c("Row.names","K=4","ibd_Diagnosis","ibd_DiseaseLocation", "ibd_ActiveDisease","clinical_BowelMovementADayDef", "Shannon_Index")]
rownames(new_pc_mtb2)=new_pc_mtb2$Row.names
new_pc_mtb2$Row.names=NULL
new_pc_mtb2$cluster=paste("c", new_pc_mtb2$`K=4`,sep="_")
new_pc_mtb2$`K=4`=NULL

new_pc_mtb2=new_pc_mtb2[order (new_pc_mtb2$cluster),]
new_clus_mtb2=clus_mtb[(row.names(new_pc_mtb2)),]
new_clus_mtb3=new_clus_mtb2[ ,colnames(new_clus_mtb2) %in% sigfeatures[[1]] ]

#pheatmap(t(new_clus_mtb2), cluster_cols = F, cluster_rows = T,show_rownames = T, breaks = seq(-10,10,length.out=20),gaps_col= c(head(as.numeric(cumsum(table(new_pc_mtb2$cluster)))), -1),show_colnames = F, annotation_col = new_pc_mtb2, scale = "none", color = inferno(20),fontsize = 5, annotation_colors = list ( ibd_Diagnosis=c("CD"="firebrick2", "UC"="blue2", "Control"="white", "IBDU"="black"),cluster=c("c_1"="black","c_2"="firebrick2","c_3"="lightblue2","c_4"="gold2" ), ibd_ActiveDisease=c("Control"="white", "Active"="red3", "NotActive"="lightblue1"), ibd_DiseaseLocation=c("Control"="white","colon"="green","ileum"="darkred","both"="gold3")))

#pheatmap(t(new_clus_mtb3), cluster_cols = F, cluster_rows = F,show_rownames = F, gaps_col= c(head(as.numeric(cumsum(table(new_pc_mtb2$cluster)))), -1),show_colnames = F, annotation_col = new_pc_mtb2, scale = "none", color = inferno(20),fontsize = 5, annotation_colors = list ( ibd_Diagnosis=c("CD"="firebrick2", "UC"="blue2", "Control"="white", "IBDU"="black"),cluster=c("c_1"="black","c_2"="firebrick2","c_3"="lightblue2","c_4"="gold2" ), ibd_ActiveDisease=c("Control"="white", "Active"="red3", "NotActive"="lightblue1"), ibd_DiseaseLocation=c("Control"="white","colon"="green","ileum"="darkred","both"="gold3")))

new_clus_tx2=clus_taxa[(row.names(new_pc_mtb2)),]
pheatmap(t(new_clus_tx2), cluster_cols = F, cluster_rows = T,show_rownames = T,breaks = seq(-10,10,length.out=10), gaps_col= c(head(as.numeric(cumsum(table(new_pc_mtb2$cluster)))), -1),show_colnames = F, annotation_col = new_pc_mtb2, scale = "none", color = inferno(10),fontsize = 5, annotation_colors = list ( ibd_Diagnosis=c("CD"="firebrick2", "UC"="blue2", "Control"="white", "IBDU"="black"),cluster=c("c_1"="black","c_2"="firebrick2","c_3"="lightblue2","c_4"="gold2" ), ibd_ActiveDisease=c("Control"="white", "Active"="red3", "NotActive"="lightblue1"), ibd_DiseaseLocation=c("Control"="white","colon"="green","ileum"="darkred","both"="gold3")))

pheatmap(t(new_clus_mtb3), cluster_cols = F, cluster_rows = T,show_rownames = T,breaks = seq(-3,3,length.out=10), gaps_col= c(head(as.numeric(cumsum(table(new_pc_mtb2$cluster)))), -1),show_colnames = F, annotation_col = new_pc_mtb2, scale = "none", color = inferno(10),fontsize = 5, annotation_colors = list ( ibd_Diagnosis=c("CD"="firebrick2", "UC"="blue2", "Control"="white", "IBDU"="black"),cluster=c("c_1"="black","c_2"="firebrick2","c_3"="lightblue2","c_4"="gold2" ), ibd_ActiveDisease=c("Control"="white", "Active"="red3", "NotActive"="lightblue1"), ibd_DiseaseLocation=c("Control"="white","colon"="green","ileum"="darkred","both"="gold3")))

```


6a: Compare metabolomics clusters
===

```{r}
clus_sel=clusters[,3, drop=F]
clus_sel$Clus="A"
clus_sel$Clus[clus_sel$`K=4`==3]="B"
clus_sel$Clus[clus_sel$`K=4`==2]="C"
clus_sel$Clus[clus_sel$`K=4`==4]="D"
clus_sel$`K=4`=NULL
clus_sel=as.data.frame(clus_sel)
clus_sel$Clus=as.factor(clus_sel$Clus)
clus_stats_mtb=summary_stats(all_new_ID_raw, clus_sel, missing_vals = "1")

taxa3=subset(taxa,rownames(taxa) %in% rownames(cc_pheno2))
taxa3[taxa3=="NA"]=0
taxa3[taxa3=="NaN"]=0

taxa3=taxa3[,((colSums(taxa3 !=0) / nrow(taxa3)) *100 )>20]
taxa3[taxa3==0]=NA


clus_stats_taxa=summary_stats(taxa3,clus_sel,missing_vals = "1")

#bgc_fliter2=as.data.frame(t(bgc_filter))
#bgc_fliter2[bgc_fliter2==0]=NA

#clus_stats_bgc=summary_stats(bgc_fliter2,clus_sel,missing_vals = "1")

##################
####METABOLITES###
##################


cm=merge(clus_sel,clus_mtb, by="row.names")

flag=1
for (a in 3:ncol(cm)){
  my_name=colnames(cm)[a]
  yy= pairwise.wilcox.test(cm[,a],cm$Clus, p.adjust.method = "none")$p.value
  cc=data.frame("Feature"=my_name,"CA_CB"=yy[1,1], "CA_CC"=yy[2,1],"CA_CD"=yy[3,1],"CB_CC"=yy[2,2], "CB_CD"=yy[3,2],"CC_CD"=yy[3,3] )
  if(flag==1){
    clus_res_mtb=cc
    flag=5
  }else{
    clus_res_mtb=rbind(clus_res_mtb,cc)
  }
}

m_clus_res_mtb=melt(clus_res_mtb)
m_clus_res_mtb$FDR=p.adjust(m_clus_res_mtb$value, method = "BH")
m_clus_res_mtb=dcast(m_clus_res_mtb, formula=Feature~variable, value.var="FDR")
a_report_mtb=merge(clus_stats_mtb,m_clus_res_mtb, by.x="Metabolite", by.y="Feature")

write.table(a_report_mtb, "~/Desktop/Metabolomics_for_cluster/3.Preliminary_results_v2/iCluster_enriched_metabolites.txt", quote = F, sep = "\t")

##################
#### TAXA ########
##################


ct=merge(clus_sel,clus_taxa, by="row.names")

flag=1
for (a in 3:ncol(ct)){
  my_name=colnames(ct)[a]
  yy= pairwise.wilcox.test(ct[,a],cm$`K=4`, p.adjust.method = "none")$p.value
  cct=data.frame("Feature"=my_name,"CA_CB"=yy[1,1], "CA_CC"=yy[2,1],"CA_CD"=yy[3,1],"CB_CC"=yy[2,2], "CB_CD"=yy[3,2],"CC_CD"=yy[3,3] )
  if(flag==1){
    clus_res_taxa=cct
    flag=5
  }else{
    clus_res_taxa=rbind(clus_res_taxa,cct)
  }
}


m_clus_res_taxa=melt(clus_res_taxa)
m_clus_res_taxa$FDR=p.adjust(m_clus_res_taxa$value, method = "BH")
m_clus_res_taxa=dcast(m_clus_res_taxa, formula=Feature~variable, value.var="FDR")
a_report_taxa=merge(clus_stats_taxa,m_clus_res_taxa, by.x="Metabolite", by.y="Feature")


write.table(a_report_taxa, "~/Desktop/Metabolomics_for_cluster/3.Preliminary_results_v2/iCluster_enriched_taxa.txt", quote = F, sep = "\t")

############################
####Biosynthetic cluster####
############################


cb=merge(clus_sel,bcg_fil_trans, by="row.names")

flag=1
for (a in 3:ncol(cb)){
  my_name=colnames(cb)[a]
  yy= pairwise.wilcox.test(cb[,a],cm$`K=4`, p.adjust.method = "none")$p.value
  ccb=data.frame("Feature"=my_name,"C1_C2"=yy[1,1], "C1_C3"=yy[2,1],"C1_C4"=yy[3,1],"C2_C3"=yy[2,2], "C2_C4"=yy[3,2],"C3_C4"=yy[3,3] )
  if(flag==1){
    clus_res_bgc=ccb
    flag=5
  }else{
    clus_res_bgc=rbind(clus_res_bgc,ccb)
  }
}

m_clus_res_bgc=melt(clus_res_bgc)
m_clus_res_bgc$FDR=p.adjust(m_clus_res_bgc$value, method = "BH")
m_clus_res_bgc=dcast(m_clus_res_bgc, formula=Feature~variable, value.var="FDR")
a_report_bgc=merge(clus_stats_bgc,m_clus_res_bgc, by.x="Metabolite", by.y="Feature")

##################
####PHENOTYPES####
##################
library("lsmeans")

phenos_ibdcc=read.table("~/Documents/Project_Metabolome/6.Results/072020/input/phenos_IBD_cc.txt", sep = "\t", row.names = 1, header = T)

#colnames(clus_sel)="Clusters"
cp=merge(clus_sel,phenos_ibdcc, by="row.names")

flag=1
for (a in 3:ncol(cp)){
  my_name=colnames(cp)[a]
  if (is.numeric(cp[,a])){
    yy= pairwise.wilcox.test(cp[,a],cm$Clus, p.adjust.method = "none")$p.value
    ccb=data.frame("Feature"=my_name,"CA_CB"=yy[1,1], "CA_CC"=yy[2,1],"CA_CD"=yy[3,1],"CB_CC"=yy[2,2], "CB_CD"=yy[3,2],"CC_CD"=yy[3,3] )
    if(flag==1){
      clus_res_phe=ccb
      flag=5
    }else{
      clus_res_phe=rbind(clus_res_phe,ccb)
    }
  } else {
    tmpdf=cp[,c(2,a)]
    zzz=glm(tmpdf[,2] ~ Clus, family = binomial(), data = tmpdf)
    yyy=as.data.frame(pairs(regrid(lsmeans(zzz,"Clus"), p.adjust.methods="none")))
    ccb=data.frame("Feature"=my_name,"CA_CB"=yyy[1,6], "CA_CC"=yyy[2,6],"CA_CD"=yyy[3,6],"CB_CC"=yyy[4,6], "CB_CD"=yyy[5,6],"CC_CD"=yyy[6,6] )
    if(flag==1){
      clus_res_phe=ccb
      flag=5
    }else{
      clus_res_phe=rbind(clus_res_phe,ccb)
    }
  }
}

m_clus_res_phe=melt(clus_res_phe)
m_clus_res_phe$FDR=p.adjust(m_clus_res_phe$value, method = "BH")
m_clus_res_phe=dcast(m_clus_res_phe, formula=Feature~variable, value.var="FDR")
row.names(cp)=cp$Row.names
cp$Row.names=NULL
stats_phenos_clus=describe.by(cp, cp$Clus)
stats_phenos=data.frame("Phenotype"=colnames(cp), CA_N= stats_phenos_clus$A$n, CA_Min=stats_phenos_clus$A$min, CA_Max=stats_phenos_clus$A$max, CA_Mean=stats_phenos_clus$A$mean, CA_Median=stats_phenos_clus$A$median, CA_SD=stats_phenos_clus$A$sd,CB_N= stats_phenos_clus$B$n, CB_Min=stats_phenos_clus$B$min, CB_Max=stats_phenos_clus$B$max, CB_Mean=stats_phenos_clus$B$mean, CB_Median=stats_phenos_clus$B$median, CB_SD=stats_phenos_clus$B$sd , CC_N= stats_phenos_clus$C$n, CC_Min=stats_phenos_clus$C$min, CC_Max=stats_phenos_clus$C$max, CC_Mean=stats_phenos_clus$C$mean, CC_Median=stats_phenos_clus$C$median, CC_SD=stats_phenos_clus$C$sd, CD_N= stats_phenos_clus$D$n, CD_Min=stats_phenos_clus$D$min, CD_Max=stats_phenos_clus$D$max, CD_Mean=stats_phenos_clus$D$mean, CD_Median=stats_phenos_clus$D$median, CD_SD=stats_phenos_clus$D$sd)

a_report_phe=merge(stats_phenos,m_clus_res_phe, by.x="Phenotype", by.y="Feature")

write.table(a_report_phe,"~/Documents/Project_Metabolome/6.Results/072020/summary_stats_clusters_phe.txt", sep = "\t", quote = F)
#write.table(a_report_bgc,"~/Documents/Project_Metabolome/6.Results/iClust/new/summary_stats_clusters_BGC.txt", sep = "\t", quote = F)
#write.table(a_report_mtb,"~/Documents/Project_Metabolome/6.Results/iClust/new/summary_stats_clusters_MTB.txt", sep = "\t", quote = F)
#write.table(a_report_taxa,"~/Documents/Project_Metabolome/6.Results/iClust/new/summary_stats_clusters_TAXA.txt", sep = "\t", quote = F)


ggplot(ct, aes(as.factor(Clus),Bacteroides_fragilis, fill=as.factor(Clus))) + geom_violin() + geom_jitter(height = 0, width = 0.1, alpha=0.4) + theme_bw() + xlab("Clusters")  + scale_fill_manual(values =c("gray80","steelblue2","firebrick3","gold3" ))


ggplot(cm, aes(as.factor(cm$Clus),tryptophan, fill=as.factor(cm$Clus))) + geom_violin() + geom_jitter(height = 0, width = 0.1, alpha=0.4) + theme_bw() + xlab("Clusters")  + scale_fill_manual(values =c("gray80","steelblue2","firebrick3","gold3" ))

```




7. Evaluating the effect of technical factors and host age, sex, BMI, Bowel movements per day
===

```{r, message=F, warning=F}

rownames(cc_pheno5)=cc_pheno5$Row.names
cc_pheno5$Row.names=NULL
cc_pheno6=cc_pheno5
table(cc_pheno6$run_day_cat, cc_pheno6$COMMENT)
cc_pheno6=subset(cc_pheno6,select = -c(Preparation_batch,Preparation_day,Preparation_order,CLIENT.IDENTIFIER,Box,Preparation_day_cat, RUN.DAY,SPL_AMNT_GRAM))

#Correction factors: day measurment was done, LC column, ammount sample per gram, month sample in a freezer, sex and age. 
my_correction_factors=cc_pheno6[,c("run_day_cat","LC.COLUMN","Amount_sample_gram","metabolon_Month_in_freezer","host_Sex","host_Age_sampling", "host_BMI", "clinical_BowelMovementADayDef")]
my_correction_factors_test=merge(my_correction_factors,q_mtb,by="row.names")
rownames(my_correction_factors_test)=my_correction_factors_test$Row.names
my_correction_factors_test$Row.names=NULL

#Box and day are correlated (i.e Box1 was analyzed between day 1 and 3, box10 only on day21)
#Therefore I will only use day for correction
#table(my_correction_factors_test$run_day_cat, test_mtb_quantitative$COMMENT)

my_corr_factors_results=matrix(ncol = 5, nrow = 6382)
b=1
for ( a in 1:ncol(my_correction_factors)){
  for (x in (ncol(my_correction_factors)+1):ncol(my_correction_factors_test)){
    if(is.numeric(my_correction_factors[,a])){
      my_corr_factors_results[b,1]=colnames(my_correction_factors_test)[a]
      my_corr_factors_results[b,2]=colnames(my_correction_factors_test)[x]
      my_test=cor.test(my_correction_factors_test[,a],my_correction_factors_test[,x], method = "spearman")
      my_corr_factors_results[b,3]=my_test$estimate
      my_corr_factors_results[b,4]=my_test$p.value
      my_corr_factors_results[b,5]="spearman"
      b=b+1
    } else{
      my_corr_factors_results[b,1]=colnames(my_correction_factors_test)[a]
      my_corr_factors_results[b,2]=colnames(my_correction_factors_test)[x]
      my_test=kruskal.test(my_correction_factors_test[,x]~my_correction_factors_test[,a])
      my_corr_factors_results[b,3]=my_test$statistic
      my_corr_factors_results[b,4]=my_test$p.value
      my_corr_factors_results[b,5]="krustal"
      b=b+1
    }
  }
}

my_corr_factors_results=as.data.frame(my_corr_factors_results)
colnames(my_corr_factors_results)=c("Factor", "Metabolite","Estimate","p.value", "test")
my_corr_factors_results$p.value=as.numeric(as.character(my_corr_factors_results$p.value))
my_corr_factors_results$FDR=p.adjust(my_corr_factors_results$p.value,method = "fdr")
my_corr_factors_results$Bonferroni=p.adjust(my_corr_factors_results$p.value,method = "bonferroni")

x=as.data.frame(table(my_corr_factors_results$Factor[my_corr_factors_results$Bonferroni<0.05]))
y=as.data.frame(table(my_corr_factors_results$Factor[my_corr_factors_results$FDR<0.05]))
associations_factors=as.data.frame(cbind(x,y))
associations_factors[,3]=NULL
colnames(associations_factors)=c("Name","Bonferroni","FDR")
plot_asso=melt(associations_factors)


write.table(my_corr_factors_results, "~/Desktop/Metabolomics_for_cluster/3.Preliminary_results_v2/technical_confouders_quantitative.txt", sep = "\t", quote = F)

ggplot(plot_asso,aes(Name,value, fill=variable)) + geom_bar(stat="identity", position =position_dodge(), colour="black") + theme_bw() + theme(axis.text.x = element_text(angle = 90, hjust = 1)) + xlab("") +ylab ("Number of associations")+ scale_x_discrete(labels=c("Amount_sample_gram" = "Amount_input_sample","clinical_BowelMovementADayDef" ="Bowel Movements", "run_day_cat" = "Run day","host_Age_sampling" = "Age", "host_Sex"="Sex", "LC.COLUMN"="LC-Column","metabolon_Month_in_freezer"="storage_time", "host_BMI"="BMI"))

test_plot=merge(cc_pheno6,q_mtb,by="row.names")
ggplot(test_plot, aes(run_day_cat,alpha_ketoglutarate)) + geom_violin(fill="gray88") + geom_jitter(aes(color=ibd_Diagnosis), position=position_jitter(0.2))+ theme_bw() + scale_color_manual(values = c("firebrick2", "black", "gray88","blue2")) 


my_correction_factors_test_bi=merge(my_correction_factors,bi_mtb,by="row.names")
rownames(my_correction_factors_test_bi)=my_correction_factors_test_bi$Row.names
my_correction_factors_test_bi$Row.names=NULL
colnames(my_correction_factors_test_bi)=make.names(colnames(my_correction_factors_test_bi))

my_corr_factors_results_bi=matrix(ncol = 5, nrow = 4112)
b=1
for ( a in 1:ncol(my_correction_factors)){
  for (x in (ncol(my_correction_factors)+1):ncol(my_correction_factors_test_bi)){
    if(is.numeric(my_correction_factors[,a])){
      my_corr_factors_results_bi[b,1]=colnames(my_correction_factors_test_bi)[a]
      my_corr_factors_results_bi[b,2]=colnames(my_correction_factors_test_bi)[x]
      my_frm=as.formula(paste(colnames(my_correction_factors_test_bi)[x], "~ ", colnames(my_correction_factors_test_bi)[a]))
      my_test=summary(glm(my_frm, family = binomial(link="logit"), data =my_correction_factors_test_bi))
      my_corr_factors_results_bi[b,3]=my_test$coefficients[2,1]
      my_corr_factors_results_bi[b,4]=my_test$coefficients[2,4]
      my_corr_factors_results_bi[b,5]="logit"
      b=b+1
    } else{
      my_corr_factors_results_bi[b,1]=colnames(my_correction_factors_test_bi)[a]
      my_corr_factors_results_bi[b,2]=colnames(my_correction_factors_test_bi)[x]
      my_test=chisq.test(my_correction_factors_test_bi[,x],my_correction_factors_test_bi[,a])
      my_corr_factors_results_bi[b,3]=my_test$statistic
      my_corr_factors_results_bi[b,4]=my_test$p.value
      my_corr_factors_results_bi[b,5]="chisq"
      b=b+1
    }
  }
}


my_corr_factors_results_bi=as.data.frame(my_corr_factors_results_bi)
colnames(my_corr_factors_results_bi)=c("Factor", "Metabolite","Estimate","p.value")
my_corr_factors_results_bi$p.value=as.numeric(as.character(my_corr_factors_results_bi$p.value))
my_corr_factors_results_bi$FDR=p.adjust(my_corr_factors_results_bi$p.value,method = "fdr")
my_corr_factors_results_bi$Bonferroni=p.adjust(my_corr_factors_results_bi$p.value,method = "bonferroni")


write.table(my_corr_factors_results_bi, "~/Desktop/Metabolomics_for_cluster/3.Preliminary_results_v2/technical_confouders_metabolites_prevalence.txt", sep = "\t", quote = F)

x=as.data.frame(table(my_corr_factors_results_bi$Factor[my_corr_factors_results_bi$Bonferroni<0.05]))
y=as.data.frame(table(my_corr_factors_results_bi$Factor[my_corr_factors_results_bi$FDR<0.05]))
association_corr_bi=as.data.frame(cbind(x,y))
association_corr_bi[,3]=NULL
colnames(association_corr_bi)=c("Name","Bonferroni","FDR")
plot_asso_bi=melt(association_corr_bi)

ggplot(plot_asso_bi,aes(Name,value, fill=variable)) + geom_bar(stat="identity", position =position_dodge(), colour="black") + theme_bw() + theme(axis.text.x = element_text(angle = 90, hjust = 1)) + xlab("") +ylab ("Number of associations")+ scale_x_discrete(labels=c("Amount_sample_gram" = "Amount_input_sample", "COMMENT" = "Box","host_Age_sampling" = "Age", "host_Sex"="Sex", "LC.COLUMN"="LC-Column","metabolon_Month_in_freezer"="storage_time"))


test_plot2=merge(cc_pheno6,bi_mtb,by="row.names")
test_pv=as.data.frame(table(test_plot2$run_day_cat, test_plot2$`cerotoylcarnitine__C26*`))
ggplot(test_pv, aes(Var1,Freq, fill=Var2)) + geom_bar(stat = "identity") +coord_flip()+ theme_bw() + ylab ("Presence of Cerotoylcarnitine") + xlab ("Day of the experiment")
```


8. Association analyses with covariates
===

Testing each metabolite~phenotype association using the several techinicals and host charactersitic covariates( age, sex, BMI, bowel movements a day,day measuring (run day), amount of sample, month sample in a freezer). 

3 different tests: 

a) Case-control to assess IBD associated metabolites
b) Quantitative metabolites (>70%) association analysis between phenotype & metabolite in IBD & Controls separate
c) Presence/absence metabolites (>20 MTB <70%) association analysis between phenotype & metabolite in IBD & Controls separate 


8a. Metabolites associated to IBD
---

```{r}

#TEST QUANTITATIVE METABOLITES >70%

test_mtb_quantitative=merge(cc_pheno6,q_mtb,by="row.names")
test_mtb_quantitative=as.data.frame(lapply(test_mtb_quantitative, gsub, pattern="Control", replacement="A_Control", fixed=T))
rownames(test_mtb_quantitative)=test_mtb_quantitative$Row.names
test_mtb_quantitative$Row.names=NULL
test_mtb_quantitative$ibd_FecalCalprotectin=NULL
test_mtb_quantitative$COMMENT=NULL

#write.table(test_mtb_quantitative,"~/Desktop/Metabolomics_for_cluster/2.Input/cc_v4.txt",sep="\t", quote = F, row.names = T)
#my_correction_factors=cc_pheno6[,c("run_day_cat","LC.COLUMN","Amount_sample_gram","metabolon_Month_in_freezer","host_Sex","host_Age_sampling", "host_BMI", "clinical_BowelMovementADayDef")]

case_control_quantitative=read.table("~/Desktop/Metabolomics_for_cluster/2.Input/cc_v4_in.txt",sep="\t", header = T, row.names = 1)
case_control_quantitative$Shannon_Index=NULL
flag=1
for ( i in 1:13) {
  my_pheno=colnames(case_control_quantitative)[i]
  if (my_pheno=="clinical_BowelMovementADayDef" | my_pheno=="LC.COLUMN" | my_pheno=="host_BMI"| my_pheno=="Amount_sample_gram"| my_pheno=="metabolon_Month_in_freezer" | my_pheno=="host_Sex" | my_pheno=="host_Age") {
    print (paste("Skipping phenotype:",my_pheno))
  } else {
    print (paste("Testing:",my_pheno))
    for (a in 14:ncol(case_control_quantitative)){
      my_trait=colnames(case_control_quantitative)[a]
      my_uni_test=case_control_quantitative[,c("host_Sex","host_Age","clinical_BowelMovementADayDef","host_BMI","LC.COLUMN","Amount_sample_gram","metabolon_Month_in_freezer",my_pheno,my_trait)]
      my_preds=c("host_Sex","host_Age","clinical_BowelMovementADayDef", "host_BMI","LC.COLUMN","Amount_sample_gram","metabolon_Month_in_freezer",my_pheno)
      my_f=as.formula(paste(my_trait, paste(my_preds, collapse = " + "), sep = " ~ "))
      #my_uni_test[,8]=as.numeric(as.character(my_uni_test[,8]))
      my_samples=nrow(my_uni_test)
      my_lm=summary(lm(my_f,data = my_uni_test ))
      my_lm_coef=as.data.frame(my_lm$coefficients)
      my_lm_pheno=try(my_lm_coef[grep(my_pheno, rownames(my_lm_coef)), ])
      my_lm_pheno$metabolite=my_trait
      my_lm_pheno$phenotype=my_pheno
      my_lm_pheno$factor=rownames(my_lm_pheno)
      if (flag!=1){
          my_univariate_results=rbind(my_univariate_results,my_lm_pheno)
      }else{
          my_univariate_results=my_lm_pheno
          flag=5
      }
    }
  }
}

associations_ibd=my_univariate_results
associations_ibd$FDR=p.adjust(associations_ibd$`Pr(>|t|)`,method = "BH")
associations_ibd$Bonferroni=p.adjust(associations_ibd$`Pr(>|t|)`,method = "bonferroni")
write.table(associations_ibd, "~/Desktop/Metabolomics_for_cluster/3.Preliminary_results_v2/IBD_associations.txt", sep = "\t", quote = F)


# TEST PREVALENCE <20% MTB <70%
# LOGISTIC REGRESSION #

test_mtb_presence=merge(cc_pheno6,bi_mtb,by="row.names")
rownames(test_mtb_presence)=test_mtb_presence$Row.names
test_mtb_presence$Row.names=NULL
test_mtb_presence$ibd_FecalCalprotectin=NULL
test_mtb_presence$COMMENT=NULL

#write.table(test_mtb_presence,"~/Desktop/Metabolomics_for_cluster/2.Input/cc_v4_prev.txt",sep="\t", quote = F, row.names = T)
case_control_prevalence=read.table("~/Desktop/Metabolomics_for_cluster/2.Input/cc_v4_in_prev.txt",sep="\t", header = T, row.names = 1)


flag=1
for ( i in 1:14) {
  my_pheno=colnames(case_control_prevalence)[i]
  if (my_pheno=="run_day_cat" | my_pheno=="clinical_BowelMovementADayDef" | my_pheno=="LC.COLUMN"| my_pheno=="Amount_sample_gram"| my_pheno=="metabolon_Month_in_freezer" | my_pheno=="host_Sex" | my_pheno=="host_Age" | my_pheno=="host_BMI") {
    print (paste("Skipping phenotype:",my_pheno))
  } else {
    print (paste("Testing:",my_pheno))
    for (a in 15:ncol(case_control_prevalence)){
      my_trait=colnames(case_control_prevalence)[a]
      my_uni_test=case_control_prevalence[,c("host_Sex","host_Age","run_day_cat","LC.COLUMN","Amount_sample_gram","metabolon_Month_in_freezer","clinical_BowelMovementADayDef", "host_BMI",my_pheno,my_trait)]
      my_preds=c("host_Sex","host_Age","host_BMI", "run_day_cat","LC.COLUMN","Amount_sample_gram","metabolon_Month_in_freezer","clinical_BowelMovementADayDef",my_pheno)
      my_f=as.formula(paste(my_trait, paste(my_preds, collapse = " + "), sep = " ~ "))
      my_uni_test[,10]=as.numeric(as.character(my_uni_test[,10]))
      my_samples=nrow(my_uni_test)
      #my_lm=summary(lm(my_f,data = my_uni_test ))
      my_lm=summary(glm(my_f, family = binomial(link="logit"), data =my_uni_test))
      my_lm_coef=as.data.frame(my_lm$coefficients)
      my_lm_pheno=try(my_lm_coef[grep(my_pheno, rownames(my_lm_coef)), ])
      my_lm_pheno$metabolite=my_trait
      my_lm_pheno$phenotype=my_pheno
      my_lm_pheno$factor=rownames(my_lm_pheno)
      if (flag!=1){
          my_univariate_results=rbind(my_univariate_results,my_lm_pheno)
      }else{
          my_univariate_results=my_lm_pheno
          flag=5
      }
    }
  }
}

prev_univariate_ibd=my_univariate_results
prev_univariate_ibd$FDR=p.adjust(prev_univariate_ibd$`Pr(>|z|)`,method = "BH")
prev_univariate_ibd$Bonferroni=p.adjust(prev_univariate_ibd$`Pr(>|z|)`,method = "bonferroni")
write.table(prev_univariate_ibd, "~/Desktop/Metabolomics_for_cluster/3.Preliminary_results_v2/IBD_associations_prevalence.txt", sep = "\t", quote = F)




#TEST QUANTITATIVE METABOLITES >20% MTB <70% => non-NAs
q_mtb_t_v3=q_mtb_t_v2[,colnames(q_mtb_t_v2) %in% colnames(bi_mtb)]
test_mtb_quantitative2=merge(cc_pheno6,q_mtb_t_v3,by="row.names")
test_mtb_quantitative2=as.data.frame(lapply(test_mtb_quantitative2, gsub, pattern="Control", replacement="A_Control", fixed=T))
rownames(test_mtb_quantitative2)=test_mtb_quantitative2$Row.names
test_mtb_quantitative2$Row.names=NULL
test_mtb_quantitative2$ibd_FecalCalprotectin=NULL
test_mtb_quantitative2$COMMENT=NULL

write.table(test_mtb_quantitative2,"~/Desktop/Metabolomics_for_cluster/2.Input/cc_v4_prev2.txt",sep="\t", quote = F, row.names = T)


case_control_quantitative2=read.table("~/Desktop/Metabolomics_for_cluster/2.Input/cc_v4_in_2.txt",sep="\t", header = T, row.names = 1)
case_control_quantitative2$Shannon_Index=NULL
flag=1
for ( i in 1:13) {
  my_pheno=colnames(case_control_quantitative2)[i]
  if (my_pheno=="clinical_BowelMovementADayDef" | my_pheno=="LC.COLUMN" | my_pheno=="host_BMI"| my_pheno=="Amount_sample_gram"| my_pheno=="metabolon_Month_in_freezer" | my_pheno=="host_Sex" | my_pheno=="host_Age") {
    print (paste("Skipping phenotype:",my_pheno))
  } else {
    print (paste("Testing:",my_pheno))
    for (a in 14:ncol(case_control_quantitative2)){
      my_trait=colnames(case_control_quantitative2)[a]
      my_uni_test=case_control_quantitative2[,c("host_Sex","host_Age","clinical_BowelMovementADayDef","host_BMI","LC.COLUMN","Amount_sample_gram","metabolon_Month_in_freezer",my_pheno,my_trait)]
      my_preds=c("host_Sex","host_Age","clinical_BowelMovementADayDef", "host_BMI","LC.COLUMN","Amount_sample_gram","metabolon_Month_in_freezer",my_pheno)
      my_f=as.formula(paste(my_trait, paste(my_preds, collapse = " + "), sep = " ~ "))
      my_uni_test=my_uni_test[complete.cases(my_uni_test[,9]),]
      my_uni_test[,10]=as.numeric(as.character(my_uni_test[,9]))
      my_samples=nrow(my_uni_test)
      my_lm=summary(lm(my_f,data = my_uni_test ))
      my_lm_coef=as.data.frame(my_lm$coefficients)
      my_lm_pheno=try(my_lm_coef[grep(my_pheno, rownames(my_lm_coef)), ])
      my_lm_pheno$metabolite=my_trait
      my_lm_pheno$phenotype=my_pheno
      my_lm_pheno$factor=rownames(my_lm_pheno)
      if (flag!=1){
          my_univariate_results=rbind(my_univariate_results,my_lm_pheno)
      }else{
          my_univariate_results=my_lm_pheno
          flag=5
      }
    }
  }
}

associations_ibd2=my_univariate_results
associations_ibd2$FDR=p.adjust(associations_ibd2$`Pr(>|t|)`,method = "BH")
associations_ibd2$Bonferroni=p.adjust(associations_ibd2$`Pr(>|t|)`,method = "bonferroni")
write.table(associations_ibd2, "~/Desktop/Metabolomics_for_cluster/3.Preliminary_results_v2/IBD_associations_low_prevalent_quantitative.txt", sep = "\t", quote = F)
```


8b. Quantitative metabolites vs phenotypes
----


Within controls


```{r}

controls_phenos=read.table("~/Desktop/Metabolomics_for_cluster/2.Input/intermed/controls_phenos2.txt", sep = "\t", header = T, row.names = 1)


controls_phenos$ibd_Diagnosis=NULL
controls_phenos$ibd_FecalCalprotectinOver200yesno.1=NULL
controls_phenos$run_day_cat=NULL
controls_phenos$run_day_cat.1=NULL
controls_phenos$LC.COLUMN.1=NULL
controls_phenos$Amount_sample_gram.1=NULL
controls_phenos$clinical_BowelMovementADayDef.1=NULL
controls_phenos$clinical_Strict_IBS.1=NULL
controls_phenos$metabolon_Month_in_freezer.1=NULL
controls_phenos$host_Age.1=NULL
controls_phenos$host_BMI.1=NULL
controls_phenos$host_Sex.1=NULL
controls_phenos$metabolon_Month_in_freezer.1=NULL
controls_phenos$BioMK_Calprotectin=NULL
controls_phenos$med_anti_epileptics=NULL
controls_phenos$med_insulin=NULL
controls_phenos$med_opiat=NULL
controls_phenos$med_oral_anti_diabetics=NULL
controls_phenos$med_other_antidepressant=NULL

flag=1
for ( i in 1:355) {
  my_pheno=colnames(controls_phenos)[i]
  if (my_pheno=="clinical_BowelMovementADayDef" | my_pheno=="LC.COLUMN" | my_pheno=="host_BMI"| my_pheno=="Amount_sample_gram"| my_pheno=="metabolon_Month_in_freezer" | my_pheno=="host_Sex" | my_pheno=="host_Age") {
    print (paste("Skipping phenotype:",my_pheno))
  } else {
    print (paste("Testing:",my_pheno))
    for (a in 356:ncol(controls_phenos)){
      my_trait=colnames(controls_phenos)[a]
      my_uni_test=controls_phenos[,c("host_Sex","host_Age","clinical_BowelMovementADayDef","host_BMI","LC.COLUMN","Amount_sample_gram","metabolon_Month_in_freezer",my_pheno,my_trait)]
      my_preds=c("host_Sex","host_Age","clinical_BowelMovementADayDef", "host_BMI","LC.COLUMN","Amount_sample_gram","metabolon_Month_in_freezer",my_pheno)
      my_f=as.formula(paste(my_trait, paste(my_preds, collapse = " + "), sep = " ~ "))
      #my_uni_test[,8]=as.numeric(as.character(my_uni_test[,8]))
      my_samples=nrow(my_uni_test)
      my_lm=summary(lm(my_f,data = my_uni_test ))
      my_lm_coef=as.data.frame(my_lm$coefficients)
      my_lm_pheno=try(my_lm_coef[grep(my_pheno, rownames(my_lm_coef)), ])
      my_lm_pheno$metabolite=my_trait
      my_lm_pheno$phenotype=my_pheno
      my_lm_pheno$factor=rownames(my_lm_pheno)
      if (flag!=1){
          my_univariate_results=rbind(my_univariate_results,my_lm_pheno)
      }else{
          my_univariate_results=my_lm_pheno
          flag=5
      }
    }
  }
}

associations_controls=my_univariate_results
associations_controls$FDR=p.adjust(associations_controls$`Pr(>|t|)`,method = "BH")
associations_controls$Bonferroni=p.adjust(associations_controls$`Pr(>|t|)`,method = "bonferroni")

write.table(associations_controls, "~/Desktop/Metabolomics_for_cluster/3.Preliminary_results_v2/Controls_associations_quantitative.txt", sep = "\t", quote = F)


##TEST PREVALENCE ###


controls_phenos_prev=read.table("~/Desktop/Metabolomics_for_cluster/2.Input/intermed/controls_phenos_prev2.txt", sep = "\t", header = T, row.names = 1)


controls_phenos_prev$ibd_Diagnosis=NULL
controls_phenos_prev$ibd_FecalCalprotectinOver200yesno.1=NULL
#controls_phenos$run_day_cat=NULL
controls_phenos_prev$run_day_cat.1=NULL
controls_phenos_prev$LC.COLUMN.1=NULL
controls_phenos_prev$Amount_sample_gram.1=NULL
controls_phenos_prev$clinical_BowelMovementADayDef.1=NULL
controls_phenos_prev$clinical_Strict_IBS.1=NULL
controls_phenos_prev$metabolon_Month_in_freezer.1=NULL
controls_phenos_prev$host_Age.1=NULL
controls_phenos_prev$host_BMI.1=NULL
controls_phenos_prev$host_Sex.1=NULL
controls_phenos_prev$metabolon_Month_in_freezer.1=NULL
controls_phenos_prev$BioMK_Calprotectin=NULL
controls_phenos_prev$med_anti_epileptics=NULL
controls_phenos_prev$med_insulin=NULL
controls_phenos_prev$med_opiat=NULL
controls_phenos_prev$med_oral_anti_diabetics=NULL
controls_phenos_prev$med_other_antidepressant=NULL


flag=1
for ( i in 1:355) {
  my_pheno=colnames(controls_phenos_prev)[i]
  if (my_pheno=="run_day_cat" | my_pheno=="clinical_BowelMovementADayDef" | my_pheno=="LC.COLUMN"| my_pheno=="Amount_sample_gram"| my_pheno=="metabolon_Month_in_freezer" | my_pheno=="host_Sex" | my_pheno=="host_Age" | my_pheno=="host_BMI") {
    print (paste("Skipping phenotype:",my_pheno))
  } else {
    print (paste("Testing:",my_pheno))
    for (a in 356:ncol(controls_phenos_prev)){
      my_trait=colnames(controls_phenos_prev)[a]
      my_uni_test=controls_phenos_prev[,c("host_Sex","host_Age","run_day_cat","LC.COLUMN","Amount_sample_gram","metabolon_Month_in_freezer","clinical_BowelMovementADayDef", "host_BMI",my_pheno,my_trait)]
      my_preds=c("host_Sex","host_Age","host_BMI", "run_day_cat","LC.COLUMN","Amount_sample_gram","metabolon_Month_in_freezer","clinical_BowelMovementADayDef",my_pheno)
      my_f=as.formula(paste(my_trait, paste(my_preds, collapse = " + "), sep = " ~ "))
      my_uni_test[,10]=as.numeric(as.character(my_uni_test[,10]))
      my_samples=nrow(my_uni_test)
      #my_lm=summary(lm(my_f,data = my_uni_test ))
      my_lm=summary(glm(my_f, family = binomial(link="logit"), data =my_uni_test))
      my_lm_coef=as.data.frame(my_lm$coefficients)
      my_lm_pheno=try(my_lm_coef[grep(my_pheno, rownames(my_lm_coef)), ])
      my_lm_pheno$metabolite=my_trait
      my_lm_pheno$phenotype=my_pheno
      my_lm_pheno$factor=rownames(my_lm_pheno)
      if (flag!=1){
          my_univariate_results=rbind(my_univariate_results,my_lm_pheno)
      }else{
          my_univariate_results=my_lm_pheno
          flag=5
      }
    }
  }
}

#mtb=controls_phenos_prev[,357:ncol(controls_phenos_prev)]
#my_remove=remove=mtb[,colSums(mtb)<nrow(mtb)*0.2]
my_univariate_results2=subset(my_univariate_results, my_univariate_results$metabolite%ni%colnames(my_remove))
prev_univariate_control=my_univariate_results2
prev_univariate_control=prev_univariate_control[prev_univariate_control$phenotype!="Shannon_Index.1",]
prev_univariate_control$FDR=p.adjust(prev_univariate_control$`Pr(>|z|)`,method = "BH")
prev_univariate_control$Bonferroni=p.adjust(prev_univariate_control$`Pr(>|z|)`,method = "bonferroni")
write.table(prev_univariate_control, "~/Desktop/Metabolomics_for_cluster/3.Preliminary_results_v2/Controls_associations_prevalence.txt", sep = "\t", quote = F)


```





Within IBD




```{r}

ibd_test_phenos=read.table("~/Desktop/Metabolomics_for_cluster/2.Input/intermed/ibd_phenos2.txt", sep = "\t", header = T, row.names = 1)

ibd_test_phenos$ibd_CurrentStomaOrPouch=NULL
ibd_test_phenos$ibd_FecalCalprotectinOver200yesno.x=NULL
ibd_test_phenos$host_Sex.1=NULL
ibd_test_phenos$med_beta_sympathomimetic_inhaler=NULL
ibd_test_phenos$med_parasympathicolytic_inhaler=NULL

flag=1
for ( i in 1:342) {
  my_pheno=colnames(ibd_test_phenos)[i]
  if (my_pheno=="clinical_BowelMovementADayDef" | my_pheno=="LC.COLUMN" | my_pheno=="host_BMI"| my_pheno=="Amount_sample_gram"| my_pheno=="metabolon_Month_in_freezer" | my_pheno=="host_Sex" | my_pheno=="host_Age") {
    print (paste("Skipping phenotype:",my_pheno))
  } else {
    print (paste("Testing:",my_pheno))
    for (a in 343:ncol(ibd_test_phenos)){
      my_trait=colnames(ibd_test_phenos)[a]
      my_uni_test=ibd_test_phenos[,c("host_Sex","host_Age","clinical_BowelMovementADayDef","host_BMI","LC.COLUMN","Amount_sample_gram","metabolon_Month_in_freezer",my_pheno,my_trait)]
      my_preds=c("host_Sex","host_Age","clinical_BowelMovementADayDef", "host_BMI","LC.COLUMN","Amount_sample_gram","metabolon_Month_in_freezer",my_pheno)
      my_f=as.formula(paste(my_trait, paste(my_preds, collapse = " + "), sep = " ~ "))
      #my_uni_test[,8]=as.numeric(as.character(my_uni_test[,8]))
      my_samples=nrow(my_uni_test)
      my_lm=summary(lm(my_f,data = my_uni_test ))
      my_lm_coef=as.data.frame(my_lm$coefficients)
      my_lm_pheno=try(my_lm_coef[grep(my_pheno, rownames(my_lm_coef)), ])
      my_lm_pheno$metabolite=my_trait
      my_lm_pheno$phenotype=my_pheno
      my_lm_pheno$factor=rownames(my_lm_pheno)
      if (flag!=1){
          my_univariate_results=rbind(my_univariate_results,my_lm_pheno)
      }else{
          my_univariate_results=my_lm_pheno
          flag=5
      }
    }
  }
}

associations_wIBD=my_univariate_results
#associations_controls=subset(associations_controls, associations_controls$phenotype!="BioMK_ChromograninA")
associations_wIBD$FDR=p.adjust(associations_wIBD$`Pr(>|t|)`,method = "BH")
associations_wIBD$Bonferroni=p.adjust(associations_wIBD$`Pr(>|t|)`,method = "bonferroni")

write.table(associations_ibd, "~/Desktop/Metabolomics_for_cluster/3.Preliminary_results_v2/within_IBD_associations_quantitative.txt", sep = "\t", quote = F)


##TEST PREVALENCE ###


ibd_test_phenos2=read.table("~/Desktop/Metabolomics_for_cluster/2.Input/intermed/ibd_phenos_prev.txt", sep = "\t", header = T, row.names = 1)

ibd_test_phenos2$ibd_CurrentStomaOrPouch=NULL
ibd_test_phenos2$ibd_FecalCalprotectinOver200yesno.x=NULL
ibd_test_phenos2$host_Sex.1=NULL
ibd_test_phenos2$med_beta_sympathomimetic_inhaler=NULL
ibd_test_phenos2$med_parasympathicolytic_inhaler=NULL

flag=1
for ( i in 1:344) {
  my_pheno=colnames(ibd_test_phenos2)[i]
  if (my_pheno=="run_day_cat" | my_pheno=="clinical_BowelMovementADayDef" | my_pheno=="LC.COLUMN"| my_pheno=="Amount_sample_gram"| my_pheno=="metabolon_Month_in_freezer" | my_pheno=="host_Sex" | my_pheno=="host_Age" | my_pheno=="host_BMI") {
    print (paste("Skipping phenotype:",my_pheno))
  } else {
    print (paste("Testing:",my_pheno))
    for (a in 345:ncol(ibd_test_phenos2)){
      my_trait=colnames(ibd_test_phenos2)[a]
      my_uni_test=ibd_test_phenos2[,c("host_Sex","host_Age","run_day_cat","LC.COLUMN","Amount_sample_gram","metabolon_Month_in_freezer","clinical_BowelMovementADayDef", "host_BMI",my_pheno,my_trait)]
      my_preds=c("host_Sex","host_Age","host_BMI", "run_day_cat","LC.COLUMN","Amount_sample_gram","metabolon_Month_in_freezer","clinical_BowelMovementADayDef",my_pheno)
      my_f=as.formula(paste(my_trait, paste(my_preds, collapse = " + "), sep = " ~ "))
      my_uni_test[,10]=as.numeric(as.character(my_uni_test[,10]))
      my_samples=nrow(my_uni_test)
      #my_lm=summary(lm(my_f,data = my_uni_test ))
      my_lm=summary(glm(my_f, family = binomial(link="logit"), data =my_uni_test))
      my_lm_coef=as.data.frame(my_lm$coefficients)
      my_lm_pheno=try(my_lm_coef[grep(my_pheno, rownames(my_lm_coef)), ])
      my_lm_pheno$metabolite=my_trait
      my_lm_pheno$phenotype=my_pheno
      my_lm_pheno$factor=rownames(my_lm_pheno)
      if (flag!=1){
        my_univariate_results=rbind(my_univariate_results,my_lm_pheno)
      }else{
        my_univariate_results=my_lm_pheno
        flag=5
      }
    }
  }
}

prev_univariate_control=my_univariate_results
prev_univariate_control=prev_univariate_control[prev_univariate_control$phenotype!="Shannon_Index.1",]
prev_univariate_control$FDR=p.adjust(prev_univariate_control$`Pr(>|z|)`,method = "BH")
prev_univariate_control$Bonferroni=p.adjust(prev_univariate_control$`Pr(>|z|)`,method = "bonferroni")
write.table(prev_univariate_control, "~/Desktop/Metabolomics_for_cluster/3.Preliminary_results_v2/Controls_associations_prevalence.txt", sep = "\t", quote = F)


```





11. Calculate ratios
===

```{r}
#Needs improvement now it takes forever!
#my_ratios=calculate_all_ratios(q_mtb)
```


12. LASSO
===

```{r}
#cc_pheno7=read.table("~/Documents/Project_Metabolome/1.Input_files/Models_032020/cc_phenos_tested.txt", sep = "\t", header = T, row.names = 1)
#tot=merge(cc_pheno7,q_mtb,by="row.names")
tot=test_mtb_quant

#Check NA's
#colnames(tot)[colSums(is.na(tot)) > 0]

tot$Methylbutyric_acid=NULL
tot$acetic_acid=NULL
tot$butyric_acid=NULL
tot$hexanoic_acid=NULL
tot$isobutyric_acid=NULL
tot$isovaleric_acid=NULL
tot$propionic_acid=NULL
tot$valeric_acid=NULL

tot$ibd_FecalCalprotectinOver200yesno[is.na(tot$ibd_FecalCalprotectinOver200yesno)]="no"
tot$clinical_BowelMovementADayDef[is.na(tot$clinical_BowelMovementADayDef)]=median(as.numeric(as.character(tot$clinical_BowelMovementADayDef)), na.rm = T)
tot$ibd_ActiveDisease[is.na(tot$ibd_ActiveDisease)]="NotActive"
tot$clinical_BinaryBSS[is.na(tot$clinical_BinaryBSS)]="Soft"
tot$ibd_DiseaseLocation[is.na(tot$ibd_DiseaseLocation)]="colon"

write.table(tot, "~/Documents/Project_Metabolome/1.Input_files/Models_052020/temp_tot.txt")
tot=read.table("~/Documents/Project_Metabolome/1.Input_files/Models_052020/temp_tot.txt")

my_results_for_lasso=all_uni_quant
my_results_for_lasso=subset(my_results_for_lasso,my_results_for_lasso$metabolite!="Methylbutyric_acid")
my_results_for_lasso=subset(my_results_for_lasso,my_results_for_lasso$metabolite!="acetic_acid")
my_results_for_lasso=subset(my_results_for_lasso,my_results_for_lasso$metabolite!="butyric_acid")
my_results_for_lasso=subset(my_results_for_lasso,my_results_for_lasso$metabolite!="hexanoic_acid")
my_results_for_lasso=subset(my_results_for_lasso,my_results_for_lasso$metabolite!="isobutyric_acid")
my_results_for_lasso=subset(my_results_for_lasso,my_results_for_lasso$metabolite!="isovaleric_acid")
my_results_for_lasso=subset(my_results_for_lasso,my_results_for_lasso$metabolite!="propionic_acid")
my_results_for_lasso=subset(my_results_for_lasso,my_results_for_lasso$metabolite!="valeric_acid")
#
#cor_diet_bact=cor.test(tot[,64:316], method = "spearman")
#
#highlyCorrelated <- findCorrelation(cor_diet_bact, cutoff=0.75)
#
q_mtb_names=q_mtb
colnames(q_mtb_names)=make.names(colnames(q_mtb_names))
q_mtb_names=subset(q_mtb_names,row.names(q_mtb_names) %in% row.names(tot))
my_models=matrix(nrow = 900,ncol = 6)
flag=1
mycount=0
for ( i in unique (my_results_for_lasso$metabolite)){
  mycount=mycount+1
  #Select phenotypes associated to metabolite in the previous analysis
  my_mtb_selection= subset (my_results_for_lasso, my_results_for_lasso$metabolite==i & my_results_for_lasso$FDR<0.05)
  
  #Subset them from the phenotype table
  my_list_of_phenos= my_mtb_selection$phenotype
  #my_list_of_phenos= my_list_of_phenos[3:42]
  my_phenos_lasso=tot[, colnames(tot) %in% my_list_of_phenos, drop=F]
  
  #my_phenos_lasso=my_phenos_lasso[,complete.cases(my_phenos_lasso)]
  print (paste("Metabolite test number:", mycount))
  print (paste("Metabolite:",i))
  #Select metabolite to test
  my_feature_lasso=q_mtb_names[,i, drop=F]
  
  #Sample sample order for phenos and metabolites tables 
  my_phenos_lasso2=my_phenos_lasso[order(row.names(my_phenos_lasso)), ]
  my_feature_lasso2=my_feature_lasso[order(row.names(my_feature_lasso)),, drop=F]
  n_samples=nrow(my_phenos_lasso2)
  
  #Use 70% for training (lambda estimation)
  lasso_train_samples=sample(1:n_samples, 0.7*n_samples)
  subtrain=rownames(my_phenos_lasso2)[lasso_train_samples]
  #Subset features 
  lasso_train_feature = my_feature_lasso2[subtrain,, drop=F]
  lasso_test_feature = subset(my_feature_lasso2, row.names(my_feature_lasso2) %ni% subtrain)
  #= my_feature_lasso2[-lasso_train_samples,, drop=F]
  #Subset phenos
  lasso_train_phenos = my_phenos_lasso2[subtrain,]
  lasso_test_phenos  = subset(my_phenos_lasso2, row.names(my_phenos_lasso2) %ni% subtrain)
  #= my_phenos_lasso2[-lasso_train_samples,]
  #Cross validation
  lasso_fit = cv.glmnet( x=data.matrix(lasso_train_phenos),y=as.matrix(lasso_train_feature), type.measure="mse", alpha=1, family="gaussian")
  #Prediction
  lasso_predicted = predict(lasso_fit, s=lasso_fit$lambda.min, newx=data.matrix(lasso_test_phenos), type="coefficients")
  #Select phenotypes with coefficients !=0
  sel_phenos_2=names(data.matrix(lasso_predicted)[data.matrix(lasso_predicted)!=0,][-1])
  #Run lm on the selected features 
  if (length(sel_phenos_2)>=1){
    my_f=as.formula(paste(colnames(my_feature_lasso), paste(sel_phenos_2, collapse = " + "), sep = " ~ ")) 
    my_lm=summary(lm(my_f,data = tot ))
    if (nrow(my_lm$coefficients)<3){
      my_lm_coef=my_lm$coefficients[-1,]
      my_lm_coef=as.data.frame(as.matrix(t(my_lm$coefficients[-1,])))
      my_lm_coef$metabolite=colnames(my_feature_lasso)
      rownames(my_lm_coef)=sel_phenos_2
      my_lm_coef$pheno=sel_phenos_2
    } else{
      my_lm_coef=as.data.frame(my_lm$coefficients[-1,])
      my_lm_coef$metabolite=colnames(my_feature_lasso)
      my_lm_coef$pheno=rownames(my_lm_coef)      
    }
    if (flag!=1){
      my_lasso_results=rbind(my_lasso_results,my_lm_coef)
    }else{
      my_lasso_results=my_lm_coef
      flag=5
    }
  }
  my_models[mycount,1]=i
  my_models[mycount,2]=paste(my_list_of_phenos, collapse = "+")
  my_models[mycount,3]=paste(my_f, collapse = "")
  my_models[mycount,4]=my_lm$adj.r.squared
  comp=data.frame( lasso_test_feature,predict(lasso_fit, s=lasso_fit$lambda.min, newx=data.matrix(lasso_test_phenos), type="response"))
  my_models[mycount,5]=RMSE(comp[,2],comp[,1])
  my_models[mycount,6]=caret::R2(obs = comp[,1], pred = comp[,2])
}
a_results_lasso=as.data.frame(my_lasso_results)
a_results_lasso$bonferroni=p.adjust(as.numeric(as.character(a_results_lasso$`Pr(>|t|)`)),method = "bonferroni")
a_results_lasso$fdr=p.adjust(as.numeric(as.character(a_results_lasso$`Pr(>|t|)`)),method = "BH")

```



13. Prediction Models
===

a) Untargetted metaboites predicting IBD
b) Untargetted metabolites predicting CD vs non-CD
c) Untargetted metabolites predicting UC vs non-UC
d) SCFA predicting IBD
e) SCFA predicting CD
f) SCFA predicting UC
g) Compare models: [basic phenos + biomarkers] vs [phenos + biomarkers + taxa] vs [phenos + biomarkers + SCFA] + vs [phenos + biomarkers + Metabolites]  


```{r}

library (lattice)
library(caret)
library (gdata)


# load the library (download it from github first:
#https://github.com/GRONINGEN-MICROBIOME-CENTRE/Groningen-Microbiome/tree/master/Scripts/RGacesa_ML_Scripts) )
source('~/Desktop/Metabolomics_for_cluster/2.Input/Prediction/R_ML_scripts_V4.R')
setwd('~/Desktop/Metabolomics_for_cluster/2.Input/Prediction/')


# test 1: prediction of cohort
# ==========================================
inDFcoh <- inDF3
# clean NAs
for (c in colnames(inDFcoh)) {
  if (class(inDFcoh[[c]]) == "numeric") {
    inDFcoh[[c]][is.na(inDFcoh[[c]])] <- median(inDFcoh[[c]],na.rm = T)
  }
}

##Prepare data

cc_ibd2=cc_ibd
cc_ibd2$ibd_Diagnosis=as.character(cc_ibd2$ibd_Diagnosis)
cc_ibd2$ibd_Diagnosis[cc_ibd2$ibd_Diagnosis!="Control"]="IBD"


#Remove correlated metabolites

mtb_corr=cor(q_mtb, method = "spearman")
corr_mtb <- findCorrelation(mtb_corr, cutoff=0.85, names = T)
pred_IBD=q_mtb[,colnames(q_mtb) %ni% corr_mtb]
pred_IBD2=merge(cc_ibd2,pred_IBD, by="row.names")
rownames(pred_IBD2)=pred_IBD2$Row.names
pred_IBD2$Row.names=NULL

# run modelling: test 1 prediction of cohort
doMLModelling(outFolderName = 'IBD_Pred',
              allDM = list(pred_IBD2),
              allDMn = c("ibd_Diagnosis"),
              responseVar = "ibd_Diagnosis",
              posC = "IBD",
              inTrainP = 0.75,
              doLC = T,
              doRFE = T,
              doDataPrepPreselection = T,
              doDataPrepPreselectP = 0.25,
              doRFEpreselection = T,
              doRFEpreselectP = 0.25,
              doOptLC = T,
              mtds = c("glm","rf","svmRadial","gamboost"),
              smoothROCs = F,
              rfeR = 5,
              tBut = 5
)

# test 2: prediction of remission in samples combined, only MGH, only UMCG
# ==========================================
inDFPredict4 <- inDF3
for (c in colnames(inDFPredict4)) {
  if (class(inDFPredict4[[c]]) == "numeric") {
    inDFPredict4[[c]][is.na(inDFPredict4[[c]])] <- median(inDFPredict4[[c]],na.rm = T)
  }
}
inDFPredict4 = inDFPredict4[83:129,]
inDFPredict4$SID <- NULL
inDFPredict4$ResponseDoc <- as.character(inDFPredict4$ResponseDoc)
inDFPredict4$ResponseDoc[inDFPredict4$ResponseDoc=="0"] <- "N"
inDFPredict4$ResponseDoc[inDFPredict4$ResponseDoc=="1"] <- "Y"
inDFPredict4$ResponseDoc <- as.factor(inDFPredict4$ResponseDoc)

doMLModelling(outFolderName = 'out_ClinicalOlinkVariables2',
              allDM = list(inDFPredict4),
              allDMn = c("ResponseDoc"),
              responseVar = "ResponseDoc",
              posC = "yes",
              inTrainP = 0.75,
              doLC = T,
              doRFE = T,
              doDataPrepPreselection = T,
              doDataPrepPreselectP = 0.25,
              doRFEpreselection = T,
              doRFEpreselectP = 0.25,
              doOptLC = T,
              mtds = c("glm"),
              smoothROCs = F,
              rfeR = 5,
              tBut = 5
)





```




13. Random forest for IBD prediction
===

IBD vs non-IBD
---

```{r}
library(mlbench)
library(caret)
library(ggplot2)
library(MLeval)
library(randomForest)
library(plyr)
library(MLmetrics)
library(pROC)


#Select uncorrelated metabolites

#mtb_corr=cor(q_mtb, method = "spearman")
corr_mtb <- findCorrelation(mtb_corr, cutoff=0.85, names = T)


#Define what to test
rf_IBD=mtb_quantitative_ibd
rf_IBD=rf_IBD[,c(10,64:ncol(rf_IBD))]
levels(rf_IBD$ibd_Diagnosis) <- c(levels(rf_IBD$ibd_Diagnosis), "IBD", "nonCD", "nonUC") 
rf_IBD[rf_IBD$ibd_Diagnosis!="A_Control",]$ibd_Diagnosis="IBD"
rf_IBD$ibd_Diagnosis=droplevels(rf_IBD$ibd_Diagnosis)

#Remove high correlated metabolites, predictions based on 650 metabolites.  
rf_IBD=rf_IBD[,colnames(rf_IBD) %ni% corr_mtb]
rf_IBD$valeric_acid=NULL
rf_IBD$propionic_acid=NULL
rf_IBD$isovaleric_acid=NULL
rf_IBD$isobutyric_acid=NULL
rf_IBD$hexanoic_acid=NULL
rf_IBD$butyric_acid=NULL
rf_IBD$acetic_acid=NULL
rf_IBD$Methylbutyric_acid=NULL
#Create testing / prediction datasets, splitting 70%
my_Train70 = createDataPartition( y = rf_IBD$ibd_Diagnosis,p = .70,list = FALSE )
training = rf_IBD[ my_Train70,]
testing  = rf_IBD[-my_Train70,]


#Defining resampling

fitControl = trainControl( method = "repeatedcv", number = 10, classProbs = T, savePredictions = T,repeats = 3,summaryFunction = twoClassSummary)

# Training with Random Forest 

model <- train( ibd_Diagnosis ~ ., data = training, method = "rf", trControl = fitControl,tuneLength = 15, metric = "ROC",na.action=na.omit)
#model2 <- train( ibd_Diagnosis ~ ., data = training, method = "rf", trControl = fitControl, metric = "ROC",na.action=na.omit)

print ("Final model training")
print ("Variable importance")
varImpPlot(model$finalModel)
print ("Training model")
model$finalModel

ddply(model$pred, "Resample", summarise, accuracy = Accuracy(pred, obs))

#for_roc=as.data.frame(model$pred[model$pred$mtry==model$bestTune$mtry,])

my_mtry=model$pred$mtry==model$bestTune$mtry

# Check and plot the training model ROC curve
my_model_eval <- evalm(model)
my_model_eval$roc
my_model_eval$stdres


#Prediction time :D
my_pred_class = predict(model,testing)
my_pred = predict(model,testing,type="prob")
confusionMatrix(data=my_pred_class, testing$ibd_Diagnosis)
my_roc=pROC::roc(testing$ibd_Diagnosis,my_pred[["IBD"]], auc=T, ci=T, smooth=F)

my_roc$auc
pROC::ggroc(my_roc, col="firebrick2", size=1.5, ci=T) + theme_bw() + geom_abline(intercept = 1, col="black", size=1, linetype="dashed")

```


CD vs Controls
---

```{r}

#Define what to test
#rf_IBD=mtb_quantitative_ibd
my_CD=rownames(mtb_quantitative_ibd)[mtb_quantitative_ibd$ibd_CD=="CD" | mtb_quantitative_ibd$ibd_Diagnosis=="A_Control" ]
my_UC=rownames(mtb_quantitative_ibd)[mtb_quantitative_ibd$ibd_UC=="UC" | mtb_quantitative_ibd$ibd_Diagnosis=="A_Control"]
#rf_IBD=rf_IBD[,c(10,64:ncol(rf_IBD))]
#rf_IBD=rf_IBD[,c(4,70:ncol(rf_IBD))]
#levels(rf_IBD$ibd_Diagnosis) <- c(levels(rf_IBD$ibd_Diagnosis), "IBD", "a_nonCD", "a_nonUC") 
#rf_IBD[rf_IBD$ibd_Diagnosis!="CD",]$ibd_Diagnosis="a_nonCD"
#rf_IBD$ibd_Diagnosis=droplevels(rf_IBD$ibd_Diagnosis)
#rf_IBD=rf_IBD[,colnames(rf_IBD) %ni% remove_corr_mtb]
#rf_IBD$valeric_acid=NULL
#rf_IBD$propionic_acid=NULL
#rf_IBD$isovaleric_acid=NULL
#rf_IBD$isobutyric_acid=NULL
#rf_IBD$hexanoic_acid=NULL
#rf_IBD$butyric_acid=NULL
#rf_IBD$acetic_acid=NULL
#rf_IBD$Methylbutyric_acid=NULL

rf_IBD2=rf_IBD[rownames(rf_IBD) %in% my_CD, ]

#Create testing / prediction datasets, splitting 70%
my_Train70 = createDataPartition( y = rf_IBD2$ibd_Diagnosis,p = .70,list = FALSE )
training = rf_IBD2[ my_Train70,]
testing  = rf_IBD2[-my_Train70,]


#Defining resampling

fitControl = trainControl( method = "repeatedcv", number = 10, classProbs = T, savePredictions = T,summaryFunction = twoClassSummary)

# Training with Random Forest 

model <- train( ibd_Diagnosis ~ ., data = training, method = "rf", trControl = fitControl,tuneLength = 15, metric = "ROC",na.action=na.omit)
#model2 <- train( ibd_Diagnosis ~ ., data = training, method = "rf", trControl = fitControl, metric = "ROC",na.action=na.omit)

print ("Final model training")
print ("Variable importance")
varImpPlot(model$finalModel)
print ("Training model")
model$finalModel

#ddply(model$pred, "Resample", summarise, accuracy = Accuracy(pred, obs))

#for_roc=as.data.frame(model$pred[model$pred$mtry==model$bestTune$mtry,])

my_mtry=model$pred$mtry==model$bestTune$mtry

# Check and plot the training model ROC curve
my_model_eval <- evalm(model)
my_model_eval$roc
my_model_eval$stdres


#Prediction time :D
my_pred_class = predict(model,testing)
my_pred = predict(model,testing,type="prob")
confusionMatrix(data=my_pred_class, testing$ibd_Diagnosis)
my_roc=pROC::roc(testing$ibd_Diagnosis,my_pred[["IBD"]], auc=T, ci=T, smooth=F)

my_roc$auc
pROC::ggroc(my_roc, col="darkblue", size=1.5, ci=T) + theme_bw() + geom_abline(intercept = 1, col="black", size=1, linetype="dashed")

```


UC vs Controls
---

```{r}

#Define what to test
rf_IBD2=rf_IBD[rownames(rf_IBD) %in% my_UC, ]


#Create testing / prediction datasets, splitting 70%
my_Train70 = createDataPartition( y = rf_IBD2$ibd_Diagnosis,p = .70,list = FALSE )
training = rf_IBD2[ my_Train70,]
testing  = rf_IBD2[-my_Train70,]


#Defining resampling

fitControl = trainControl( method = "repeatedcv", number = 10, classProbs = T, savePredictions = T,summaryFunction = twoClassSummary)

# Training with Random Forest 

model <- train( ibd_Diagnosis ~ ., data = training, method = "rf", trControl = fitControl,tuneLength = 15, metric = "ROC",na.action=na.omit)
#model2 <- train( ibd_Diagnosis ~ ., data = training, method = "rf", trControl = fitControl, metric = "ROC",na.action=na.omit)

print ("Final model training")
print ("Variable importance")
varImpPlot(model$finalModel)
print ("Training model")
model$finalModel

ddply(model$pred, "Resample", summarise, accuracy = Accuracy(pred, obs))

#for_roc=as.data.frame(model$pred[model$pred$mtry==model$bestTune$mtry,])

my_mtry=model$pred$mtry==model$bestTune$mtry

# Check and plot the training model ROC curve
my_model_eval <- evalm(model)
my_model_eval$roc
my_model_eval$stdres


#Prediction time :D
my_pred_class = predict(model,testing)
my_pred = predict(model,testing,type="prob")
confusionMatrix(data=my_pred_class, testing$ibd_Diagnosis)
my_roc=pROC::roc(testing$ibd_Diagnosis,my_pred[["IBD"]], auc=T, ci=T, smooth=F, levels=c("A_Control", "IBD"))

my_roc$auc
pROC::ggroc(my_roc, col="purple", size=1.5, ci=T) + theme_bw() + geom_abline(intercept = 1, col="black", size=1, linetype="dashed")

```


CD vs UC 
---

```{r}

#Define what to test
#rf_IBD2=rf_IBD[rownames(rf_IBD) %in% my_UC, ]
rf_IBD=mtb_quantitative_ibd
rf_IBD=rf_IBD[,c(10,64:ncol(rf_IBD))]
rf_IBD=rf_IBD[,colnames(rf_IBD) %ni% remove_corr_mtb]
rf_IBD$valeric_acid=NULL
rf_IBD$propionic_acid=NULL
rf_IBD$isovaleric_acid=NULL
rf_IBD$isobutyric_acid=NULL
rf_IBD$hexanoic_acid=NULL
rf_IBD$butyric_acid=NULL
rf_IBD$acetic_acid=NULL
rf_IBD$Methylbutyric_acid=NULL




rf_IBD2=subset(rf_IBD, rf_IBD$ibd_Diagnosis!="A_Control")
rf_IBD2=subset(rf_IBD2, rf_IBD2$ibd_Diagnosis!="IBDU")

rf_IBD2$ibd_Diagnosis=droplevels(rf_IBD2$ibd_Diagnosis)

#Create testing / prediction datasets, splitting 70%
my_Train70 = createDataPartition( y = rf_IBD2$ibd_Diagnosis,p = .70,list = FALSE )
training = rf_IBD2[ my_Train70,]
testing  = rf_IBD2[-my_Train70,]


#Defining resampling

fitControl = trainControl( method = "repeatedcv", number = 10, classProbs = T, savePredictions = T,summaryFunction = twoClassSummary)

# Training with Random Forest 

model <- train( ibd_Diagnosis ~ ., data = training, method = "rf", trControl = fitControl,tuneLength = 15, metric = "ROC",na.action=na.omit)
#model2 <- train( ibd_Diagnosis ~ ., data = training, method = "rf", trControl = fitControl, metric = "ROC",na.action=na.omit)

print ("Final model training")
print ("Variable importance")
varImpPlot(model$finalModel)
print ("Training model")
model$finalModel

ddply(model$pred, "Resample", summarise, accuracy = Accuracy(pred, obs))

#for_roc=as.data.frame(model$pred[model$pred$mtry==model$bestTune$mtry,])

my_mtry=model$pred$mtry==model$bestTune$mtry

# Check and plot the training model ROC curve
my_model_eval <- evalm(model)
my_model_eval$roc
my_model_eval$stdres


#Prediction time :D
my_pred_class = predict(model,testing)
my_pred = predict(model,testing,type="prob")
confusionMatrix(data=my_pred_class, testing$ibd_Diagnosis)
my_roc=pROC::roc(testing$ibd_Diagnosis,my_pred[["CD"]], auc=T, ci=T, smooth=F, levels=c("UC", "CD"))

my_roc$auc
pROC::ggroc(my_roc, col="orange", size=1.5, ci=T) + theme_bw() + geom_abline(intercept = 1, col="black", size=1, linetype="dashed")

```




14. Explore rare metabolites
===

```{r}

#Relation to bacteria

rare_mtb=all_new_ID_raw [,colnames(all_new_ID_raw) %in% to_remove$Metabolite]
annot_rare=annot[rownames(annot)%in%colnames(rare_mtb),]
rare_mtb_coded=rare_mtb
rare_mtb_coded[!is.na(rare_mtb_coded)]=1
rare_mtb_coded[is.na(rare_mtb_coded)]=0
t_rare=merge(cc_pheno3,rare_mtb_coded,by="row.names")

rare_mtb2=merge(cc_pheno3,rare_mtb,by="row.names")

t_rare=t_rare[,c(1,207:ncol(t_rare))]


#t_res=matrix(nrow=34444,ncol = 4)
#v=1
#for (i in 2:110){
  for (a in 111:ncol(t_rare)){
    t_res[v,1]=colnames(t_rare)[i]
    t_res[v,2]=colnames(t_rare)[a]
    t_res[v,3]=sum(t_rare[,a] !=0)
    if(t_res[v,3]!=0){
      t_res[v,4]=wilcox.test(t_rare[,i]~as.factor(t_rare[,a]))$p.value
      v=v+1
    } else {
      t_res[v,4]=1
      v=v+1
    }
  }
}

#t_res=as.data.frame(t_res)
#t_res$bnf=p.adjust(as.numeric(as.character(t_res[,4])), method = "bonferroni")

#Relation to drugs

t_rare=merge(cc_pheno3,rare_mtb_coded,by="row.names")
t_rare=t_rare[,c(1,16:59,316:ncol(t_rare))]
m_res=matrix(nrow=14175,ncol = 4)
v=1
for (i in 2:46){
  for (a in 47:ncol(t_rare)){
    m_res[v,1]=colnames(t_rare)[i]
    m_res[v,2]=colnames(t_rare)[a]
    m_res[v,3]=sum(t_rare[,a] !=0)
    if(m_res[v,3]!=0){
      m_res[v,4]=chisq.test(t_rare[,i],as.factor(t_rare[,a]))$p.value
      v=v+1
    } else {
      m_res[v,4]=1
      v=v+1
    }
  }
}

m_res=as.data.frame(m_res)
m_res$bnf=p.adjust(as.numeric(as.character(m_res[,4])), method = "bonferroni")
m_res$FDR=p.adjust(as.numeric(as.character(m_res[,4])), method = "BH")
colnames(m_res)=c("Drug", "Metabolite", "Number_samples_metabolite_detected","BNF","FDR")
write.table(m_res, "~/Desktop/Metabolomics_for_cluster/3.Preliminary_results_v2/rare_metabolites_medication.txt", sep = "\t", quote = F)

#Only for drugs within IBD

m_meds=ibd_test_phenos[,c(44:92)]
m_meds2=merge(m_meds,rare_mtb_coded, by="row.names")

m_res_ibd=matrix(nrow=15484,ncol = 4)
v=1
for (i in 2:49){
  for (a in 50:ncol(m_meds2)){
    m_res_ibd[v,1]=colnames(m_meds2)[i]
    m_res_ibd[v,2]=colnames(m_meds2)[a]
    m_res_ibd[v,3]=sum(m_meds2[,a] !=0)
    if(m_res_ibd[v,3]!=0){
      m_res_ibd[v,4]=chisq.test(m_meds2[,i],as.factor(m_meds2[,a]))$p.value
      v=v+1
    } else {
      m_res_ibd[v,4]=1
      v=v+1
    }
  }
}

m_res_ibd=as.data.frame(m_res_ibd)
m_res_ibd$bnf=p.adjust(as.numeric(as.character(m_res_ibd[,4])), method = "bonferroni")
m_res_ibd$FDR=p.adjust(as.numeric(as.character(m_res_ibd[,4])), method = "BH")
write.table(m_res_ibd, "~/Desktop/Metabolomics_for_cluster/3.Preliminary_results_v2/rare_metabolites_medication_IBD.txt", sep = "\t", quote = F)
```



15. Relation metabolite-microbiota
===

```{r}

#1: Regress age, sex, BMI, etc.  => Test correlation IBD / non-IBD
#2: Regress age,sex, BMI etc. + IBD => test correlation all samples 

#cc_pheno4=merge(SCFA2,cc_pheno3, by="row.names")
#rownames(cc_pheno4)=cc_pheno4$Row.names
#cc_pheno4$Row.names=NULL
#cc_pheno5=merge(my_batch_id,cc_pheno4, by="row.names")
cc_rd=read.table("~/Documents/Project_Metabolome/1.Input_files/IDs_read_depth.txt", header = T, row.names = "PID")
cc_rd$row.names=NULL
cc_rd2=merge(cc_rd,cc_pheno5,by="row.names", all.y = T)


my_regress_phenos=cc_rd2[,c("Row.names","run_day_cat", "Amount_sample_gram", "LC.COLUMN" , "metabolon_Month_in_freezer" , "host_Age_sampling", "host_BMI","host_Sex", "PF_RD", "clinical_BowelMovementADayDef", "ibd_Diagnosis")]

levels(my_regress_phenos$ibd_Diagnosis)=c(levels(my_regress_phenos$ibd_Diagnosis), "IBD")
my_regress_phenos$ibd_Diagnosis[my_regress_phenos$ibd_Diagnosis!="Control"]="IBD"
my_regress_phenos$ibd_Diagnosis=droplevels(my_regress_phenos$ibd_Diagnosis)

my_regress_phenos$clinical_BowelMovementADayDef[is.na(my_regress_phenos$clinical_BowelMovementADayDef)]=median(my_regress_phenos$clinical_BowelMovementADayDef,na.rm = T)

rownames(my_regress_phenos)=my_regress_phenos$Row.names
my_regress_phenos$Row.names=NULL

#clus_mtb
rgs_mb=merge(my_regress_phenos,clus_mtb, by="row.names")

rgs_mb2=matrix(nrow=nrow(my_regress_phenos), ncol = ncol(clus_mtb))
rgs_mb3=matrix(nrow=nrow(my_regress_phenos), ncol = ncol(clus_mtb))
n=1
for (i in 12:ncol(rgs_mb)){
  
  #my_lm=lm(rgs_mb[,i] ~ Amount_sample_gram+LC.COLUMN+metabolon_Month_in_freezer+host_Age_sampling+ host_BMI+ host_Sex +clinical_BowelMovementADayDef+ ibd_Diagnosis,data = rgs_mb)
  
  #rgs_mb2[,n] = my_lm$residuals
  
  my_lm2=lm(rgs_mb[,i] ~ Amount_sample_gram+LC.COLUMN+metabolon_Month_in_freezer+host_Age_sampling+ host_BMI+ host_Sex+ clinical_BowelMovementADayDef,data = rgs_mb)
  
  rgs_mb3[,n] = my_lm2$residuals
  n=n+1
}
#rgs_mb2=as.data.frame(rgs_mb2)
#rownames(rgs_mb2)=rgs_mb$Row.names
#colnames(rgs_mb2)=colnames(rgs_mb)[12:ncol(rgs_mb)]

rgs_mb3=as.data.frame(rgs_mb3)
rownames(rgs_mb3)=rgs_mb$Row.names
colnames(rgs_mb3)=colnames(rgs_mb)[12:ncol(rgs_mb)]


#clus_taxa
my_regress_phenos$PF_RD[is.na(my_regress_phenos$PF_RD)]=median(my_regress_phenos$PF_RD,na.rm = T)
rgs_tx=merge(my_regress_phenos,clus_taxa, by="row.names")
rgs_tx2=matrix(nrow=nrow(my_regress_phenos), ncol = ncol(clus_taxa))
rgs_tx3=matrix(nrow=nrow(my_regress_phenos), ncol = ncol(clus_taxa))
n=1
for (i in 12:ncol(rgs_tx)){

  #my_lm=lm(rgs_tx[,i] ~ host_Age_sampling+ host_BMI+ host_Sex +PF_RD+clinical_BowelMovementADayDef+ ibd_Diagnosis,data = rgs_tx)
  
  #rgs_tx2[,n] = my_lm$residuals
  
  my_lm2=lm(rgs_tx[,i] ~host_Age_sampling+ host_BMI+ host_Sex+clinical_BowelMovementADayDef+ PF_RD,data = rgs_tx)
  
  rgs_tx3[,n] = my_lm2$residuals
  n=n+1
}
#rgs_tx2=as.data.frame(rgs_tx2)
#rownames(rgs_tx2)=rgs_tx$Row.names
#colnames(rgs_tx2)=colnames(rgs_tx)[12:ncol(rgs_tx)]

rgs_tx3=as.data.frame(rgs_tx3)
rownames(rgs_tx3)=rgs_tx$Row.names
colnames(rgs_tx3)=colnames(rgs_tx)[12:ncol(rgs_tx)]

##test correlation between bacteri and metabolites##

##IBD regressed values##

corr_mtb_taxa=merge(rgs_mb2,rgs_tx2, by="row.names")
row.names(corr_mtb_taxa)=corr_mtb_taxa$Row.names
corr_mtb_taxa$Row.names=NULL
#corr_mtb_taxa_vals=rcorr(as.matrix(corr_mtb_taxa), type = "spearman")

#test_corr_coef=corr_mtb_taxa_vals$r
#test_corr_coef_sel=test_corr_coef[(ncol(rgs_mb2)+1):nrow(test_corr_coef), 1:ncol(rgs_mb2)]
#test_corr_pval=corr_mtb_taxa_vals$P
#test_corr_pval_sel=test_corr_pval[(ncol(rgs_mb2)+1):nrow(test_corr_pval), 1:ncol(rgs_mb2)]
#corr_coef=melt(test_corr_coef_sel)
#corr_pval=melt(test_corr_pval_sel)

#table(corr_coef$Var1 == corr_pval$Var1)
#table(corr_coef$Var2 == corr_pval$Var2)

#colnames(corr_coef)=c("Taxa","Metabolites","Coeff")
#colnames(corr_pval)=c("Taxa","Metabolites2","P_val")
#corr_taxa_metabolites_IBD=cbind(corr_coef, corr_pval)
#corr_taxa_metabolites_IBD$Metabolites2=NULL
#corr_taxa_metabolites_IBD[,4]=NULL

########################
##Split IBD / Controls##
########################

corr_mtb_taxa_IBD=subset(corr_mtb_taxa,row.names(corr_mtb_taxa) %in% my_IBD)
corr_mtb_taxa_CNT=subset(corr_mtb_taxa,row.names(corr_mtb_taxa) %in% my_CNT)


corr_mtb_taxa_vals=rcorr(as.matrix(corr_mtb_taxa_IBD), type = "spearman")

test_corr_coef=corr_mtb_taxa_vals$r
test_corr_coef_sel=test_corr_coef[(ncol(rgs_mb2)+1):nrow(test_corr_coef), 1:ncol(rgs_mb2)]
test_corr_pval=corr_mtb_taxa_vals$P
test_corr_pval_sel=test_corr_pval[(ncol(rgs_mb2)+1):nrow(test_corr_pval), 1:ncol(rgs_mb2)]
corr_coef=melt(test_corr_coef_sel)
corr_pval=melt(test_corr_pval_sel)

table(corr_coef$Var1 == corr_pval$Var1)
table(corr_coef$Var2 == corr_pval$Var2)

colnames(corr_coef)=c("Taxa","Metabolites","Coeff")
colnames(corr_pval)=c("Taxa","Metabolites2","P_val")
corr_taxa_metabolites_oIBD=cbind(corr_coef, corr_pval)


corr_mtb_taxa_vals=rcorr(as.matrix(corr_mtb_taxa_CNT), type = "spearman")

test_corr_coef=corr_mtb_taxa_vals$r
test_corr_coef_sel=test_corr_coef[(ncol(rgs_mb2)+1):nrow(test_corr_coef), 1:ncol(rgs_mb2)]
test_corr_pval=corr_mtb_taxa_vals$P
test_corr_pval_sel=test_corr_pval[(ncol(rgs_mb2)+1):nrow(test_corr_pval), 1:ncol(rgs_mb2)]
corr_coef=melt(test_corr_coef_sel)
corr_pval=melt(test_corr_pval_sel)

table(corr_coef$Var1 == corr_pval$Var1)
table(corr_coef$Var2 == corr_pval$Var2)

colnames(corr_coef)=c("Taxa","Metabolites","Coeff")
colnames(corr_pval)=c("Taxa","Metabolites2","P_val")
corr_taxa_metabolites_CNT=cbind(corr_coef, corr_pval)

corr_taxa_metabolites_oIBD$Metabolites2=NULL
corr_taxa_metabolites_oIBD[,4]=NULL
corr_taxa_metabolites_CNT$Metabolites2=NULL
corr_taxa_metabolites_CNT[,4]=NULL

corr_taxa_metabolites_oIBD$Bonferroni=p.adjust(corr_taxa_metabolites_oIBD$P_val, method = "bonferroni")
corr_taxa_metabolites_CNT$Bonferroni=p.adjust(corr_taxa_metabolites_CNT$P_val, method = "bonferroni")
corr_taxa_metabolites_oIBD$FDR=p.adjust(corr_taxa_metabolites_oIBD$P_val, method = "BH")
corr_taxa_metabolites_CNT$FDR=p.adjust(corr_taxa_metabolites_CNT$P_val, method = "BH")

corr_taxa_metabolites_f=cbind(corr_taxa_metabolites_oIBD, corr_taxa_metabolites_CNT)
colnames(corr_taxa_metabolites_f)=c("Taxa", "Metabolite", "Coef_IBD", "P-val_IBD","BFN_IBD","FDR_IBD", "Taxa", "Metabolites","Coeff_Controls", "P-val_Controls", "BFN_Controls", "FDR_Controls")

write.table(corr_taxa_metabolites_f, "~/Desktop/Metabolomics_for_cluster/3.Preliminary_results_v2/correlation_metabolites_taxa_spearman.txt", sep = "\t", quote = F)

sig_corr=subset(corr_taxa_metabolites_f, corr_taxa_metabolites_f$FDR_IBD<0.05 | corr_taxa_metabolites_f$FDR_Controls<0.05 )
sig_corr$replicated="No"
sig_corr$replicated[sig_corr$FDR_IBD<0.05 & sig_corr$FDR_Controls<0.05 ]="Yes"
write.table(sig_corr, "~/Desktop/Metabolomics_for_cluster/3.Preliminary_results_v2/correlation_metabolites_taxa_spearman_significant.txt", sep = "\t", quote = F)



res_mtb_tx=corr_taxa_metabolites_f[,c(1,2,3,4,5)]
res_mtb_tx$Coef2=res_mtb_tx$Coef_IBD
res_mtb_tx$Coef2[res_mtb_tx$`P-val_IBD`>0.05]=0
my_test2=dcast(res_mtb_tx, formula=Metabolite~Taxa, value.var="Coef2")
rownames(my_test2)=my_test2$Metabolite
my_test2$Metabolite=NULL


pheatmap(t(my_test2), cluster_cols = T, cluster_rows = T,show_rownames = T, show_colnames = T, annotation_col = annot_v2, scale = "none",fontsize = 5, color = colorRampPalette(c("navy", "white", "firebrick3"))(50))

#Select metabolites based on the discriminant power in iClust analysis
my_test3=my_test2[rownames(my_test2) %in% sigfeatures[[1]], ]
#pheatmap(t(my_test3), cluster_cols = T, cluster_rows = T,show_rownames = T, show_colnames = T, annotation_col = annot_v2, scale = "none", color = inferno(20),fontsize = 5)
pheatmap(t(my_test3), cluster_cols = T, cluster_rows = T,show_rownames = T, show_colnames = T, annotation_col = annot_v2, scale = "none",fontsize = 5, color = colorRampPalette(c("navy", "white", "firebrick3"))(50))



res_mtb_tx=corr_taxa_metabolites_f[,c(6,7,8,9,10)]
res_mtb_tx$Coef2=res_mtb_tx$Coeff_Controls
res_mtb_tx$Coef2[res_mtb_tx$BFN_Controls>0.05]=0
my_test2=dcast(res_mtb_tx, formula=Metabolites~Taxa, value.var="Coef2")
rownames(my_test2)=my_test2$Metabolites
my_test2$Metabolites=NULL


pheatmap(t(my_test2), cluster_cols = T, cluster_rows = T,show_rownames = T, show_colnames = T, annotation_col = annot_v2, scale = "none",fontsize = 5, color = colorRampPalette(c("navy", "white", "firebrick3"))(50))

#Select metabolites based on the discriminant power in iClust analysis
my_test3=my_test2[rownames(my_test2) %in% sigfeatures[[1]], ]
#pheatmap(t(my_test3), cluster_cols = T, cluster_rows = T,show_rownames = T, show_colnames = T, annotation_col = annot_v2, scale = "none", color = inferno(20),fontsize = 5)
pheatmap(t(my_test3), cluster_cols = T, cluster_rows = T,show_rownames = T, show_colnames = T, annotation_col = annot_v2, scale = "none",fontsize = 5, color = colorRampPalette(c("navy", "white", "firebrick3"))(50))


```



16.Biosynthetic gene clusters
===

```{r}

##############IMPLEMENT: Check covarage #####################



#Calculated using BiG-MAP: https://github.com/medema-group/BiG-MAP & https://github.com/victoriapascal/gutsmash/tree/gutsmash

#Hannah database: gutSMASH on a combined collection of 4.083 genomes from the Culturable Genome Reference (CGR), Human Microbiome Project (HMP) and complete genomes belonging to Clostridiales species. This resulted in 11.509 predicted metabolic gene clusters which we used as input for BiG-MAP.family

#conda activate /home/umcg-avich/miniconda3/envs/BiG-MAP_process
#ml Bowtie2
#python3 /groups/umcg-gastrocol/tmp01/Arnau/BiG-MAP/src/BiG-MAP.map.py -th 10 -I1 /groups/umcg-gastrocol/tmp01/Arnau/BiG-MAP/batch8/*_1.fastq -I2 /groups/umcg-gastrocol/tmp01/Arnau/BiG-MAP/batch8/*_2.fastq -O /groups/umcg-gastrocol/tmp01/Arnau/BiG-MAP/result_b8/ -P ./BiG-MAP.GCF_HGF.pickle 
cc_rd=read.table("~/Documents/Project_Metabolome/1.Input_files/IDs_read_depth.txt", header = T, row.names = "PID")
cc_rd$row.names=NULL

setwd("~/Documents/Project_Metabolome/7.Bacterial_genes_and_pathways/all2/")
 
file_list <- list.files()
flag=1
for (file in file_list){
       
  # if the merged dataset doesn't exist, create it
  if (flag==1){
    bgc <- read.table(file, header=TRUE,  check.names = F, sep=",", row.names = 1, stringsAsFactors = F)
    flag=5
  }
   
  # if the merged dataset does exist, append to it
  else {
    temp_bgc <-read.table(file, header=TRUE,  check.names = F, sep=",", row.names = 1, stringsAsFactors = F)
    bgc<-merge(bgc, temp_bgc, by="row.names", all.x=T)
    row.names(bgc)=bgc$Row.names
    bgc$Row.names=NULL
    rm(temp_bgc)
  }
 
}

bgc_rpkm=select(bgc, contains(".RPKM"))
colnames(bgc_rpkm)=gsub(".RPKM","",colnames(bgc_rpkm))
cc_rd$id2=rownames(cc_rd)
row.names(cc_rd)=cc_rd$IDs_MGS
cc_rd$row.names=NULL

bgc_rpkm2=merge(cc_rd,t(bgc_rpkm), by="row.names")
rownames(bgc_rpkm2)=bgc_rpkm2$id2
bgc_rpkm2$id2=NULL
bgc_rpkm2$Row.names=NULL
bgc_rpkm2$IDs_MGS=NULL
bgc_rpkm2$PF_RD=NULL
bgc_rpkm3=subset(bgc_rpkm2, row.names(bgc_rpkm2) %in% row.names(cc_pheno2))
bgc_rpkm3.1=bgc_rpkm3[,colnames(bgc_rpkm3) %in% colnames(bcg_fil_trans)]
select2=subset(select,select$ibd_Diagnosis!="IBDU")
summary_bgc=summary_stats(bgc_rpkm3.1,select2, missing_vals = 0)


bgc_coverage=select(bgc, contains(".cov"))
colnames(bgc_coverage)=gsub(".cov","",colnames(bgc_coverage))

bgc_coverage2=merge(cc_rd,t(bgc_coverage), by="row.names")
rownames(bgc_coverage2)=bgc_coverage2$id2
bgc_coverage2$id2=NULL
bgc_coverage2$Row.names=NULL
bgc_coverage2$IDs_MGS=NULL
bgc_coverage2$PF_RD=NULL
bgc_coverage3=subset(bgc_coverage2, row.names(bgc_coverage2) %in% row.names(cc_pheno2))

#h_genes=read.table("~/Documents/Project_Metabolome/7.Bacterial_genes_and_pathways/test_data_Hannah/RPKM_BiG-MAP_names.tsv", header = T)

#Select samples in our study

features_per_samples=data.frame(Sample=rownames(bgc_rpkm3), Counts=rowSums(bgc_rpkm3!=0))
samples_per_feature=data.frame(BCG=colnames(bgc_rpkm3), Counts=colSums(bgc_rpkm3!=0))


ggplot(features_per_samples, aes(reorder(Sample,-Counts), Counts )) + geom_bar(stat = "identity", fill="lightblue3") + theme_classic() + theme(axis.text.x=element_blank(), axis.ticks.x=element_blank()) + ylab ("Number of Biosynthetic genes clusters")  + xlab ("Samples")

boxss=merge(cc_pheno, features_per_samples, by="row.names")

ggplot(boxss, aes (ibd_Diagnosis, Counts, fill=ibd_Diagnosis)) + geom_violin() + geom_boxplot(width=0.1,position=position_dodge(1), alpha=0.2)  + theme_bw() +scale_fill_manual(values=c("firebrick2","white","gray80", "blue2")) + ylab ("Number of biosynthetic gene clusters detected") + xlab ("") + scale_x_discrete(labels=c("n=239","n=255", "n=12", "n=175"))

ggplot(samples_per_feature, aes(reorder(BCG,-Counts), Counts )) + geom_bar(stat = "identity", fill="gray50") + theme_classic() + theme(axis.text.x=element_blank(), axis.ticks.x=element_blank()) + ylab ("Number of samples")  + xlab ("Biosynthetic genes clusters")

#Filtering based on the region coverage (10%)
bgc_coverage4=bgc_coverage3 [,colnames(bgc_coverage3) %in% colnames(bcg_fil_trans)]
bgc_mean_coverage=data.frame(BCG=colnames(bgc_coverage4), Mean=colMeans(bgc_coverage4))
ggplot(bgc_mean_coverage, aes(reorder(BCG,-Mean), Mean )) + geom_bar(stat = "identity", fill="red4") + theme_classic() + theme(axis.text.x=element_blank(), axis.ticks.x=element_blank()) + ylab ("Mean coverage")  + xlab ("Biosynthetic genes clusters")

select_coverage=droplevels(bgc_mean_coverage$BCG[bgc_mean_coverage$Mean>0.1])

bcg_fil_trans=bgc_rpkm3[,colnames(bgc_rpkm3) %in% select_coverage]

#Filtering prevalence in 20% => 568 metabolites
bcg_fil_trans2=transform_and_filter_taxa(bcg_fil_trans, samples_row = T, method = "clr",missing_filter = 20)


cc_ibd=cc_pheno6[,"ibd_Diagnosis", drop=F]
bgc_ibd=merge(cc_ibd,bcg_fil_trans2, by="row.names")

flag=1
for (a in 3:ncol(bgc_ibd)){
  my_name=colnames(bgc_ibd)[a]
  yy= pairwise.wilcox.test(bgc_ibd[,a],bgc_ibd$ibd_Diagnosis, p.adjust.method = "none")$p.value
  ccb=data.frame("Feature"=my_name,"CD_Control"=yy[1,1], "CD_IBDU"=yy[2,1],"CD_UC"=yy[3,1],"Control_IBDU"=yy[2,2], "Control_UC"=yy[3,2],"IBDU_UC"=yy[3,3] )
  if(flag==1){
    bgc_res=ccb
    flag=5
  }else{
    bgc_res=rbind(bgc_res,ccb)
  }
}
bgc_res$CD_IBDU=NULL
bgc_res$Control_IBDU=NULL
bgc_res$IBDU_UC=NULL
bgc_res_2=melt(bgc_res)
bgc_res_2$FDR=p.adjust(bgc_res_2$value, method = "BH")
bgc_res_2=dcast(bgc_res_2, formula=Feature~variable, value.var="FDR")
a_report_bgc=merge(summary_bgc,bgc_res_2, by.x="Metabolite", by.y="Feature")


write.table(bgc_ibd, "~/Desktop/Metabolomics_for_cluster/1.Raw_data_Metabolon/BGC_with_IBD", sep = "\t", quote = F)
write.table(summary_bgc, "~/Desktop/Metabolomics_for_cluster/3.Preliminary_results_v2/BGC_summary_stats.txt", sep = "\t", quote = F)
write.table(a_report_bgc, "~/Desktop/Metabolomics_for_cluster/3.Preliminary_results_v2/BGC_summary_stats_significant.txt", sep = "\t", quote = F)



###################################
## correlation BGC & Taxa        ##
###################################


#Regress covariates to BGC

#clus_taxa

rgs_bgc=merge(my_regress_phenos,bcg_fil_trans2, by="row.names")

rgs_bgc3=matrix(nrow=nrow(my_regress_phenos), ncol = ncol(bcg_fil_trans2))
n=1
for (i in 12:ncol(rgs_bgc)){

  #my_lm=lm(rgs_tx[,i] ~ host_Age_sampling+ host_BMI+ host_Sex +PF_RD+clinical_BowelMovementADayDef+ ibd_Diagnosis,data = rgs_tx)
  
  #rgs_tx2[,n] = my_lm$residuals
  
  my_lm2=lm(rgs_bgc[,i] ~host_Age_sampling+ host_BMI+ host_Sex+clinical_BowelMovementADayDef+ PF_RD,data = rgs_bgc)
  
  rgs_bgc3[,n] = my_lm2$residuals
  n=n+1
}
rgs_bgc3=as.data.frame(rgs_bgc3)
rownames(rgs_bgc3)=rgs_bgc$Row.names
colnames(rgs_bgc3)=colnames(rgs_bgc)[12:ncol(rgs_bgc)]

rgs_bgc_tx=merge(rgs_bgc3,rgs_tx3, by="row.names")
rownames(rgs_bgc_tx)=rgs_bgc_tx$Row.names
rgs_bgc_tx$Row.names=NULL

corr_bgc_tx_IBD=subset(rgs_bgc_tx,row.names(rgs_bgc_tx) %in% my_IBD)
corr_bgc_tx_CNT=subset(rgs_bgc_tx,row.names(rgs_bgc_tx) %in% my_CNT)


corr_bgc_tx_IBD_val=rcorr(as.matrix(corr_bgc_tx_IBD), type = "spearman")

test_corr_coef=corr_bgc_tx_IBD_val$r
test_corr_coef_sel=test_corr_coef[(ncol(rgs_bgc3)+1):nrow(test_corr_coef), 1:ncol(rgs_bgc3)]
test_corr_pval=corr_bgc_tx_IBD_val$P
test_corr_pval_sel=test_corr_pval[(ncol(rgs_bgc3)+1):nrow(test_corr_pval), 1:ncol(rgs_bgc3)]
corr_coef=melt(test_corr_coef_sel)
corr_pval=melt(test_corr_pval_sel)

table(corr_coef$Var1 == corr_pval$Var1)
table(corr_coef$Var2 == corr_pval$Var2)

colnames(corr_coef)=c("Taxa","BGC","Coeff")
colnames(corr_pval)=c("Taxa","BGC","P_val")
corr_taxa_bgc_oIBD=cbind(corr_coef, corr_pval)


corr_bgc_tx_vals=rcorr(as.matrix(corr_bgc_tx_CNT), type = "spearman")

test_corr_coef=corr_bgc_tx_vals$r
test_corr_coef_sel=test_corr_coef[(ncol(rgs_bgc3)+1):nrow(test_corr_coef), 1:ncol(rgs_bgc3)]
test_corr_pval=corr_bgc_tx_vals$P
test_corr_pval_sel=test_corr_pval[(ncol(rgs_bgc3)+1):nrow(test_corr_pval), 1:ncol(rgs_bgc3)]
corr_coef=melt(test_corr_coef_sel)
corr_pval=melt(test_corr_pval_sel)

table(corr_coef$Var1 == corr_pval$Var1)
table(corr_coef$Var2 == corr_pval$Var2)

colnames(corr_coef)=c("Taxa","BGC","Coeff")
colnames(corr_pval)=c("Taxa","BGC","P_val")
corr_bgc_tx_CNT=cbind(corr_coef, corr_pval)

corr_taxa_bgc_oIBD$Metabolites2=NULL
corr_taxa_bgc_oIBD[,4]=NULL
corr_bgc_tx_CNT$Metabolites2=NULL
corr_bgc_tx_CNT[,4]=NULL

corr_taxa_bgc_oIBD$Bonferroni=p.adjust(corr_taxa_bgc_oIBD$P_val, method = "bonferroni")
corr_bgc_tx_CNT$Bonferroni=p.adjust(corr_bgc_tx_CNT$P_val, method = "bonferroni")
corr_taxa_bgc_oIBD$FDR=p.adjust(corr_taxa_bgc_oIBD$P_val, method = "BH")
corr_bgc_tx_CNT$FDR=p.adjust(corr_bgc_tx_CNT$P_val, method = "BH")

corr_taxa_bgc_f=cbind(corr_taxa_bgc_oIBD, corr_bgc_tx_CNT)
colnames(corr_taxa_bgc_f)=c("Taxa", "BGC", "Coef_IBD", "BGC","P-val_IBD","BFN_IBD","FDR_IBD", "Taxa", "BGC","Coeff_Controls","BGC", "P-val_Controls", "BFN_Controls", "FDR_Controls")


write.table(corr_taxa_bgc_f, "~/Desktop/Metabolomics_for_cluster/3.Preliminary_results_v2/correlation_BGC_taxa_spearman.txt", sep = "\t", quote = F)

sig_corr=subset(corr_taxa_bgc_f, corr_taxa_bgc_f$FDR_IBD<0.05 | corr_taxa_bgc_f$FDR_Controls<0.05 )
sig_corr$replicated="No"
sig_corr$replicated[sig_corr$FDR_IBD<0.05 & sig_corr$FDR_Controls<0.05 ]="Yes"
write.table(sig_corr, "~/Desktop/Metabolomics_for_cluster/3.Preliminary_results_v2/correlation_BGC_taxa_spearman_significant.txt", sep = "\t", quote = F)



###################################
## correlation BGC & Metabolites ##
###################################

rgs_bgc_mtb=merge(rgs_bgc3,rgs_mb3, by="row.names")
rownames(rgs_bgc_mtb)=rgs_bgc_mtb$Row.names
rgs_bgc_mtb$Row.names=NULL

rgs_bgc_mtb_IBD=subset(rgs_bgc_mtb,row.names(rgs_bgc_mtb) %in% my_IBD)
rgs_bgc_mtb_CNT=subset(rgs_bgc_mtb,row.names(rgs_bgc_mtb) %in% my_CNT)


rgs_bgc_mtb_IBD_val=rcorr(as.matrix(rgs_bgc_mtb_IBD), type = "spearman")

test_corr_coef=rgs_bgc_mtb_IBD_val$r
test_corr_coef_sel=test_corr_coef[(ncol(rgs_bgc3)+1):nrow(test_corr_coef), 1:ncol(rgs_bgc3)]
test_corr_pval=rgs_bgc_mtb_IBD_val$P
test_corr_pval_sel=test_corr_pval[(ncol(rgs_bgc3)+1):nrow(test_corr_pval), 1:ncol(rgs_bgc3)]
corr_coef=melt(test_corr_coef_sel)
corr_pval=melt(test_corr_pval_sel)

table(corr_coef$Var1 == corr_pval$Var1)
table(corr_coef$Var2 == corr_pval$Var2)

colnames(corr_coef)=c("Metabolite","BGC","Coeff")
colnames(corr_pval)=c("Metabolite","BGC","P_val")
corr_mtb_bgc_oIBD=cbind(corr_coef, corr_pval)





corr_bgc_mtb_vals=rcorr(as.matrix(rgs_bgc_mtb_CNT), type = "spearman")
#here:
test_corr_coef=corr_bgc_mtb_vals$r
test_corr_coef_sel=test_corr_coef[(ncol(rgs_bgc3)+1):nrow(test_corr_coef), 1:ncol(rgs_bgc3)]
test_corr_pval=corr_bgc_mtb_vals$P
test_corr_pval_sel=test_corr_pval[(ncol(rgs_bgc3)+1):nrow(test_corr_pval), 1:ncol(rgs_bgc3)]
corr_coef=melt(test_corr_coef_sel)
corr_pval=melt(test_corr_pval_sel)

table(corr_coef$Var1 == corr_pval$Var1)
table(corr_coef$Var2 == corr_pval$Var2)

colnames(corr_coef)=c("Metabolites","BGC","Coeff")
colnames(corr_pval)=c("Metavolites","BGC","P_val")
corr_bgc_mtb_CNT=cbind(corr_coef, corr_pval)

corr_mtb_bgc_oIBD$Bonferroni=p.adjust(corr_mtb_bgc_oIBD$P_val, method = "bonferroni")
corr_bgc_mtb_CNT$Bonferroni=p.adjust(corr_bgc_mtb_CNT$P_val, method = "bonferroni")
corr_mtb_bgc_oIBD$FDR=p.adjust(corr_mtb_bgc_oIBD$P_val, method = "BH")
corr_bgc_mtb_CNT$FDR=p.adjust(corr_bgc_mtb_CNT$P_val, method = "BH")

corr_mtb_bgc_f=cbind(corr_mtb_bgc_oIBD, corr_bgc_mtb_CNT)


colnames(corr_mtb_bgc_f)=c("Metabolites", "BGC", "Coef_IBD", "Metabolites","BGC","P-val_IBD","BFN_IBD","FDR_IBD", "Metabolites", "BGC","Coeff_Controls","Metabolites","BGC", "P-val_Controls", "BFN_Controls", "FDR_Controls")


write.table(corr_mtb_bgc_f, "~/Desktop/Metabolomics_for_cluster/3.Preliminary_results_v2/correlation_BGC_metabolites_spearman.txt", sep = "\t", quote = F)

sig_corr=subset(corr_mtb_bgc_f, corr_mtb_bgc_f$FDR_IBD<0.05 | corr_mtb_bgc_f$FDR_Controls<0.05 )
sig_corr$replicated="No"
sig_corr$replicated[sig_corr$FDR_IBD<0.05 & sig_corr$FDR_Controls<0.05 ]="Yes"
write.table(sig_corr, "~/Desktop/Metabolomics_for_cluster/3.Preliminary_results_v2/correlation_BGC_metabolites_spearman_significant.txt", sep = "\t", quote = F)

###############

##Make plots##

##############

colnames(corr_coef_tx)=c("BGC","Metabolites","Coeff")
colnames(corr_pval_tx)=c("BGC2","Metabolites2","P_val")
bcg_tx=cbind(corr_coef_tx, corr_pval_tx)
bcg_tx$BGC2=NULL
bcg_tx$Metabolites2=NULL
bcg_tx$Coeff2=bcg_tx$Coeff
bcg_tx$FDR=p.adjust(bcg_tx$P_val, method = "BH")
bcg_tx[bcg_tx$FDR>0.05,]$Coeff2=0

bcg_tx_plotlo=dcast(bcg_tx, formula=Metabolites~BGC, value.var="Coeff2")
row.names(bcg_tx_plot)=bcg_tx_plot$Metabolites
bcg_tx_plot$Metabolites=NULL

pheatmap(bcg_tx_plot, cluster_cols = T, cluster_rows = T,show_rownames = T, show_colnames = F, breaks = seq(-1,1,length.out=60),scale = "none",fontsize = 5, color = colorRampPalette(c("navy", "white", "firebrick3"))(60))


#####################################################
### TEST within IBD BGC - Metabolite associations ###
#####################################################

ibd_test_bgc=read.table("~/Desktop/Metabolomics_for_cluster/2.Input/intermed/ibd_BGC.txt", sep = "\t", header = T, row.names = 1)

flag=1
for ( i in 1:322) {
  my_pheno=colnames(ibd_test_bgc)[i]
  if (my_pheno=="clinical_BowelMovementADayDef" | my_pheno=="LC.COLUMN" | my_pheno=="host_BMI"| my_pheno=="Amount_sample_gram"| my_pheno=="metabolon_Month_in_freezer" | my_pheno=="host_Sex" | my_pheno=="host_Age") {
    print (paste("Skipping phenotype:",my_pheno))
  } else {
    print (paste("Testing:",my_pheno))
    for (a in 323:ncol(ibd_test_bgc)){
      my_trait=colnames(ibd_test_bgc)[a]
      my_uni_test=ibd_test_bgc[,c("host_Sex","host_Age","clinical_BowelMovementADayDef","host_BMI","LC.COLUMN","Amount_sample_gram","metabolon_Month_in_freezer",my_pheno,my_trait)]
      my_preds=c("host_Sex","host_Age","clinical_BowelMovementADayDef", "host_BMI","LC.COLUMN","Amount_sample_gram","metabolon_Month_in_freezer",my_pheno)
      my_f=as.formula(paste(my_trait, paste(my_preds, collapse = " + "), sep = " ~ "))
      #my_uni_test[,8]=as.numeric(as.character(my_uni_test[,8]))
      my_samples=nrow(my_uni_test)
      my_lm=summary(lm(my_f,data = my_uni_test ))
      my_lm_coef=as.data.frame(my_lm$coefficients)
      my_lm_pheno=try(my_lm_coef[grep(my_pheno, rownames(my_lm_coef)), ])
      my_lm_pheno$metabolite=my_trait
      my_lm_pheno$phenotype=my_pheno
      my_lm_pheno$factor=rownames(my_lm_pheno)
      if (flag!=1){
          my_univariate_results=rbind(my_univariate_results,my_lm_pheno)
      }else{
          my_univariate_results=my_lm_pheno
          flag=5
      }
    }
  }
}

associations_bgc_wIBD=my_univariate_results
#associations_controls=subset(associations_controls, associations_controls$phenotype!="BioMK_ChromograninA")
associations_bgc_wIBD$FDR=p.adjust(associations_bgc_wIBD$`Pr(>|t|)`,method = "BH")
associations_bgc_wIBD$Bonferroni=p.adjust(associations_bgc_wIBD$`Pr(>|t|)`,method = "bonferroni")

write.table(associations_bgc_wIBD, "~/Desktop/Metabolomics_for_cluster/3.Preliminary_results_v2/within_IBD_associations_BGC_quantitative.txt", sep = "\t", quote = F)


##########################################################
### TEST within Controls BGC - Metabolite associations ###
##########################################################

cnt_test_bgc=read.table("~/Desktop/Metabolomics_for_cluster/2.Input/intermed/controls_BGC.txt", sep = "\t", header = T, row.names = 1)

flag=1
for ( i in 1:322) {
  my_pheno=colnames(cnt_test_bgc)[i]
  if (my_pheno=="clinical_BowelMovementADayDef" | my_pheno=="LC.COLUMN" | my_pheno=="host_BMI"| my_pheno=="Amount_sample_gram"| my_pheno=="metabolon_Month_in_freezer" | my_pheno=="host_Sex" | my_pheno=="host_Age") {
    print (paste("Skipping phenotype:",my_pheno))
  } else {
    print (paste("Testing:",my_pheno))
    for (a in 323:ncol(cnt_test_bgc)){
      my_trait=colnames(cnt_test_bgc)[a]
      my_uni_test=cnt_test_bgc[,c("host_Sex","host_Age","clinical_BowelMovementADayDef","host_BMI","LC.COLUMN","Amount_sample_gram","metabolon_Month_in_freezer",my_pheno,my_trait)]
      my_preds=c("host_Sex","host_Age","clinical_BowelMovementADayDef", "host_BMI","LC.COLUMN","Amount_sample_gram","metabolon_Month_in_freezer",my_pheno)
      my_f=as.formula(paste(my_trait, paste(my_preds, collapse = " + "), sep = " ~ "))
      #my_uni_test[,8]=as.numeric(as.character(my_uni_test[,8]))
      my_samples=nrow(my_uni_test)
      my_lm=summary(lm(my_f,data = my_uni_test ))
      my_lm_coef=as.data.frame(my_lm$coefficients)
      my_lm_pheno=try(my_lm_coef[grep(my_pheno, rownames(my_lm_coef)), ])
      my_lm_pheno$metabolite=my_trait
      my_lm_pheno$phenotype=my_pheno
      my_lm_pheno$factor=rownames(my_lm_pheno)
      if (flag!=1){
          my_univariate_results=rbind(my_univariate_results,my_lm_pheno)
      }else{
          my_univariate_results=my_lm_pheno
          flag=5
      }
    }
  }
}

associations_bgc_CNT=my_univariate_results
#associations_controls=subset(associations_controls, associations_controls$phenotype!="BioMK_ChromograninA")
associations_bgc_CNT$FDR=p.adjust(associations_bgc_CNT$`Pr(>|t|)`,method = "BH")
associations_bgc_CNT$Bonferroni=p.adjust(associations_bgc_CNT$`Pr(>|t|)`,method = "bonferroni")

write.table(associations_bgc_CNT, "~/Desktop/Metabolomics_for_cluster/3.Preliminary_results_v2/Controls_associations_BGC_quantitative.txt", sep = "\t", quote = F)



ibd_test_phenos2=read.table("~/Desktop/Metabolomics_for_cluster/2.Input/intermed/controls_BGC_prev2.txt", sep = "\t", header = T, row.names = 1)

ibd_test_phenos2$clinical_BowelMovementADayDef.1=NULL

flag=1
for ( i in 1:323) {
  my_pheno=colnames(ibd_test_phenos2)[i]
  if (my_pheno=="run_day_cat" | my_pheno=="clinical_BowelMovementADayDef" | my_pheno=="LC.COLUMN"| my_pheno=="Amount_sample_gram"| my_pheno=="metabolon_Month_in_freezer" | my_pheno=="host_Sex" | my_pheno=="host_Age" | my_pheno=="host_BMI") {
    print (paste("Skipping phenotype:",my_pheno))
  } else {
    print (paste("Testing:",my_pheno))
    for (a in 324:ncol(ibd_test_phenos2)){
      my_trait=colnames(ibd_test_phenos2)[a]
      my_uni_test=ibd_test_phenos2[,c("host_Sex","host_Age","run_day_cat","LC.COLUMN","Amount_sample_gram","metabolon_Month_in_freezer","clinical_BowelMovementADayDef", "host_BMI",my_pheno,my_trait)]
      my_preds=c("host_Sex","host_Age","host_BMI", "run_day_cat","LC.COLUMN","Amount_sample_gram","metabolon_Month_in_freezer","clinical_BowelMovementADayDef",my_pheno)
      my_f=as.formula(paste(my_trait, paste(my_preds, collapse = " + "), sep = " ~ "))
      my_uni_test[,10]=as.numeric(as.character(my_uni_test[,10]))
      my_samples=nrow(my_uni_test)
      #my_lm=summary(lm(my_f,data = my_uni_test ))
      my_lm=summary(glm(my_f, family = binomial(link="logit"), data =my_uni_test))
      my_lm_coef=as.data.frame(my_lm$coefficients)
      my_lm_pheno=try(my_lm_coef[grep(my_pheno, rownames(my_lm_coef)), ])
      my_lm_pheno$metabolite=my_trait
      my_lm_pheno$phenotype=my_pheno
      my_lm_pheno$factor=rownames(my_lm_pheno)
      if (flag!=1){
        my_univariate_results=rbind(my_univariate_results,my_lm_pheno)
      }else{
        my_univariate_results=my_lm_pheno
        flag=5
      }
    }
  }
}


#change this below and save!

associations_BGC_control_prev=my_univariate_results
associations_BGC_control_prev$FDR=p.adjust(associations_BGC_control_prev$`Pr(>|z|)`,method = "BH")
associations_BGC_control_prev$Bonferroni=p.adjust(associations_BGC_control_prev$`Pr(>|z|)`,method = "bonferroni")
write.table(associations_BGC_control_prev, "~/Desktop/Metabolomics_for_cluster/3.Preliminary_results_v2/controls_associations_BGC_prevalence.txt", sep = "\t", quote = F)

```



17.Enzime commission gene abundance
===

```{r}

enzymes=read.table("~/Desktop/Metabolomics_for_cluster/2.Input/Merged_EC_filtered.txt", header = T, sep="\t", row.names = 1, check.names = F)

colnames(enzymes)=gsub("_EC_translated","",colnames(enzymes))

enzymes2=merge(cc_rd,t(enzymes), by="row.names")
rownames(enzymes2)=enzymes2$id2
enzymes2$id2=NULL
enzymes2$Row.names=NULL
enzymes2$IDs_MGS=NULL
enzymes2$PF_RD=NULL
enzymes3=subset(enzymes2, row.names(enzymes2) %in% row.names(cc_pheno2))
#summary_enzymes=summary_stats(enzymes3,select2, missing_vals = 0)

features_per_samples=data.frame(Sample=rownames(enzymes3), Counts=rowSums(enzymes3!=0))
samples_per_feature=data.frame(BCG=colnames(enzymes3), Counts=colSums(enzymes3!=0))

ggplot(features_per_samples, aes(reorder(Sample,-Counts), Counts )) + geom_bar(stat = "identity", fill="lightblue3") + theme_classic() + theme(axis.text.x=element_blank(), axis.ticks.x=element_blank()) + ylab ("Number of Enzymes per samples")  + xlab ("Samples")

boxss=merge(cc_pheno, features_per_samples, by="row.names")

ggplot(boxss, aes (ibd_Diagnosis, Counts, fill=ibd_Diagnosis)) + geom_violin() + geom_boxplot(width=0.1,position=position_dodge(1), alpha=0.2)  + theme_bw() +scale_fill_manual(values=c("firebrick2","white","gray80", "blue2")) + ylab ("Number of enzymes detected") + xlab ("") + scale_x_discrete(labels=c("n=239","n=255", "n=12", "n=175"))

ggplot(samples_per_feature, aes(reorder(BCG,-Counts), Counts )) + geom_bar(stat = "identity", fill="gray50") + theme_classic() + theme(axis.text.x=element_blank(), axis.ticks.x=element_blank()) + ylab ("Number of samples")  + xlab ("Enzymes")


enzymes3$UNMAPPED=NULL
enzymes3$UNGROUPED=NULL

#enzymes4=enzymes3[,((colSums(enzymes3 !=0) / nrow(enzymes3)) *100 )>20]


#Filtering prevalence in 20% 
enzymes_trans=transform_and_filter_taxa(enzymes3, samples_row = T, method = "clr",missing_filter = 20)
enz_corr=cor(enzymes_trans, method = "spearman")
corr_enz <- findCorrelation(enz_corr, cutoff=0.85, names = T)
enzymes_trans2=enzymes_trans[,colnames(enzymes_trans) %ni% corr_enz]

#write.table(enzymes_trans2, "~/Desktop/Metabolomics_for_cluster/2.Input/EC_filtered.txt", sep = "\t", quote = F)

enzy_ibd=merge(cc_ibd,enzymes_trans2, by="row.names")

flag=1
for (a in 3:ncol(enzy_ibd)){
  my_name=colnames(enzy_ibd)[a]
  yy= pairwise.wilcox.test(enzy_ibd[,a],enzy_ibd$ibd_Diagnosis, p.adjust.method = "none")$p.value
  ccb=data.frame("Feature"=my_name,"CD_Control"=yy[1,1], "CD_IBDU"=yy[2,1],"CD_UC"=yy[3,1],"Control_IBDU"=yy[2,2], "Control_UC"=yy[3,2],"IBDU_UC"=yy[3,3] )
  if(flag==1){
    enzy_res=ccb
    flag=5
  }else{
    enzy_res=rbind(enzy_res,ccb)
  }
}
enzy_res$CD_IBDU=NULL
enzy_res$Control_IBDU=NULL
enzy_res$IBDU_UC=NULL
enzy_res_2=melt(enzy_res)
enzy_res_2$FDR=p.adjust(enzy_res_2$value, method = "BH")
enzy_res_2=dcast(enzy_res_2, formula=Feature~variable, value.var="FDR")


#a_report_bgc=merge(summary_bgc,bgc_res_2, by.x="Metabolite", by.y="Feature")

#######################################
## correlation enzymes & metabolites ##
#######################################


rgs_enzy=merge(my_regress_phenos,enzymes_trans2, by="row.names")

rgs_enzy3=matrix(nrow=nrow(my_regress_phenos), ncol = ncol(enzymes_trans2))
n=1
for (i in 12:ncol(rgs_bgc)){

  #my_lm=lm(rgs_tx[,i] ~ host_Age_sampling+ host_BMI+ host_Sex +PF_RD+clinical_BowelMovementADayDef+ ibd_Diagnosis,data = rgs_tx)
  
  #rgs_tx2[,n] = my_lm$residuals
  
  my_lm2=lm(rgs_enzy[,i] ~host_Age_sampling+ host_BMI+ host_Sex+clinical_BowelMovementADayDef+ PF_RD,data = rgs_enzy)
  
  rgs_enzy3[,n] = my_lm2$residuals
  n=n+1
}
rgs_enzy3=as.data.frame(rgs_enzy3)
rownames(rgs_enzy3)=rgs_enzy$Row.names
colnames(rgs_enzy3)=colnames(rgs_enzy)[12:ncol(rgs_enzy)]



enzy_mtb=merge(rgs_enzy3,rgs_mb3, by="row.names")
rownames(enzy_mtb)=enzy_mtb$Row.names
enzy_mtb$Row.names=NULL

enzy_mtb_IBD=subset(enzy_mtb,row.names(enzy_mtb) %in% my_IBD)
enzy_mtb_CNT=subset(enzy_mtb,row.names(enzy_mtb) %in% my_CNT)


enzy_mtb_IBD_val=rcorr(as.matrix(enzy_mtb_IBD), type = "spearman")

test_corr_coef=enzy_mtb_IBD_val$r
test_corr_coef_sel=test_corr_coef[(ncol(rgs_enzy3)+1):nrow(test_corr_coef), 1:ncol(rgs_enzy3)]
test_corr_pval=enzy_mtb_IBD_val$P
test_corr_pval_sel=test_corr_pval[(ncol(rgs_enzy3)+1):nrow(test_corr_pval), 1:ncol(rgs_enzy3)]
corr_coef=melt(test_corr_coef_sel)
corr_pval=melt(test_corr_pval_sel)

table(corr_coef$Var1 == corr_pval$Var1)
table(corr_coef$Var2 == corr_pval$Var2)

colnames(corr_coef)=c("Metabolite","Enzyme","Coeff")
colnames(corr_pval)=c("Metabolite","Enzyme","P_val")
corr_mtb_enz_oIBD=cbind(corr_coef, corr_pval)


enzy_mtb_vals=rcorr(as.matrix(enzy_mtb_CNT), type = "spearman")
#here:
test_corr_coef=enzy_mtb_vals$r
test_corr_coef_sel=test_corr_coef[(ncol(rgs_enzy3)+1):nrow(test_corr_coef), 1:ncol(rgs_enzy3)]
test_corr_pval=enzy_mtb_vals$P
test_corr_pval_sel=test_corr_pval[(ncol(rgs_enzy3)+1):nrow(test_corr_pval), 1:ncol(rgs_enzy3)]
corr_coef=melt(test_corr_coef_sel)
corr_pval=melt(test_corr_pval_sel)

table(corr_coef$Var1 == corr_pval$Var1)
table(corr_coef$Var2 == corr_pval$Var2)

colnames(corr_coef)=c("Metabolites","Enzyme","Coeff")
colnames(corr_pval)=c("Metabolites","Enzyme","P_val")
corr_enzy_mtb_CNT=cbind(corr_coef, corr_pval)

corr_mtb_enz_oIBD$Bonferroni=p.adjust(corr_mtb_enz_oIBD$P_val, method = "bonferroni")
corr_enzy_mtb_CNT$Bonferroni=p.adjust(corr_enzy_mtb_CNT$P_val, method = "bonferroni")
corr_mtb_enz_oIBD$FDR=p.adjust(corr_mtb_enz_oIBD$P_val, method = "BH")
corr_enzy_mtb_CNT$FDR=p.adjust(corr_enzy_mtb_CNT$P_val, method = "BH")

corr_mtb_enzy_f=cbind(corr_mtb_enz_oIBD, corr_enzy_mtb_CNT)


colnames(corr_mtb_enzy_f)=c("Metabolites", "Enzyme", "Coef_IBD", "Metabolites","Enzyme","P-val_IBD","BFN_IBD","FDR_IBD", "Metabolites", "Enzyme","Coeff_Controls","Metabolites","Enzyme", "P-val_Controls", "BFN_Controls", "FDR_Controls")


write.table(corr_mtb_enzy_f, "~/Desktop/Metabolomics_for_cluster/3.Preliminary_results_v2/correlation_enzyme_metabolites_spearman.txt", sep = "\t", quote = F)

sig_corr=subset(corr_mtb_enzy_f, corr_mtb_enzy_f$FDR_IBD<0.05 | corr_mtb_enzy_f$FDR_Controls<0.05 )
sig_corr$replicated="No"
sig_corr$replicated[sig_corr$FDR_IBD<0.05 & sig_corr$FDR_Controls<0.05 ]="Yes"
write.table(sig_corr, "~/Desktop/Metabolomics_for_cluster/3.Preliminary_results_v2/correlation_enzyme_metabolites_spearman_significant.txt", sep = "\t", quote = F)


```









